{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46bf0fc8-b44c-4814-b726-0f7cd1ac8e7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 30956 descriptors\n"
     ]
    }
   ],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "from difflib import SequenceMatcher, get_close_matches\n",
    "\n",
    "def parse_mesh_descriptors(xml_path):\n",
    "    \"\"\"Parse the MeSH XML and return a list of descriptors.\"\"\"\n",
    "    tree = ET.parse(xml_path)\n",
    "    root = tree.getroot()\n",
    "    descriptors = []\n",
    "    for dr in root.findall('DescriptorRecord'):\n",
    "        ui_el   = dr.find('DescriptorUI')\n",
    "        name_el = dr.find('DescriptorName/String')\n",
    "        if ui_el is None or name_el is None:\n",
    "            continue\n",
    "        ui   = ui_el.text\n",
    "        name = name_el.text\n",
    "        tree_nums = [tn.text for tn in dr.findall('TreeNumberList/TreeNumber') if tn.text]\n",
    "        descriptors.append({'ui':ui, 'name':name, 'tree_numbers':tree_nums})\n",
    "    return descriptors\n",
    "\n",
    "MESH_XML = 'desc2025.xml' #from the folder\n",
    "descriptors = parse_mesh_descriptors(MESH_XML)\n",
    "print(f\"Parsed {len(descriptors)} descriptors\")\n",
    "#print(descriptors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "95543660-d21e-49a9-b9cc-fe6952f092d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bacteria_taxonomy(descriptors):\n",
    "    \"\"\"\n",
    "    Find the descriptor with name 'Bacteria', \n",
    "    then collect every descriptor under that tree-number prefix.\n",
    "    \"\"\"\n",
    "    prefix = None #bacteria is B03\n",
    "    for d in descriptors:\n",
    "        if d['name'].lower() == 'bacteria' and d['tree_numbers']:\n",
    "            prefix = d['tree_numbers'][0]\n",
    "            break\n",
    "    if not prefix:\n",
    "        raise RuntimeError(\"Could not find 'Bacteria' in descriptors\")\n",
    "    tax = {}\n",
    "    for d in descriptors:\n",
    "        for tn in d['tree_numbers']:\n",
    "            if tn == prefix or tn.startswith(prefix + '.'):\n",
    "                tax[tn] = {'ui':d['ui'], 'name':d['name']}\n",
    "                break\n",
    "    return tax\n",
    "\n",
    "bacteria_tax = get_bacteria_taxonomy(descriptors)\n",
    "#print(f\"{len(bacteria_tax)} bacterial MeSH nodes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8decb4e7-6384-4ba5-a4dc-70eec8ad9d67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexed 859 bacterial names\n"
     ]
    }
   ],
   "source": [
    "def build_name_index(taxonomy):\n",
    "    \"\"\"\n",
    "    Build a dict: lower-case name -> list of (tree#, ui, canonical name)\n",
    "    \"\"\"\n",
    "    idx = {}\n",
    "    for tree_num, info in taxonomy.items():\n",
    "        key = info['name'].lower()\n",
    "        idx.setdefault(key, []).append((tree_num, info['ui'], info['name']))\n",
    "    return idx\n",
    "\n",
    "name_index = build_name_index(bacteria_tax)\n",
    "print(f\"Indexed {len(name_index)} bacterial names\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56b33d82-0ea2-4473-8cd7-1f462e832da5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B03         Bacteria                        (D001419)\n",
      "B03.026     Acidobacteria                   (D061271)\n",
      "B03.054     Agricultural Inoculants         (D059827)\n",
      "B03.110     Atypical Bacterial Forms        (D001295)\n",
      "B03.110.422  L Forms                         (D007740)\n",
      "B03.110.761  Spheroplasts                    (D013104)\n",
      "B03.120     Bacteria, Aerobic               (D001420)\n",
      "B03.130     Bacteria, Anaerobic             (D001421)\n",
      "B03.135     Bacteria, Thermoduric           (D000072280)\n",
      "B03.250     Chlorobi                        (D019414)\n",
      "B03.250.140  Chlorobium                      (D041883)\n",
      "B03.275     Chloroflexi                     (D041862)\n",
      "B03.275.150  Chloroflexus                    (D041861)\n",
      "B03.275.575  Dehalococcoides                 (D000082942)\n",
      "B03.280     Cyanobacteria                   (D000458)\n",
      "B03.280.100  Anabaena                        (D017033)\n",
      "B03.280.100.150  Anabaena cylindrica             (D046868)\n",
      "B03.280.100.900  Anabaena variabilis             (D046870)\n",
      "B03.280.110  Aphanizomenon                   (D046890)\n",
      "B03.280.200  Cyanothece                      (D046889)\n",
      "B03.280.210  Cylindrospermopsis              (D046891)\n",
      "B03.280.283  Dolichospermum flos-aquae       (D046869)\n",
      "B03.280.355  Lyngbya                         (D000083582)\n",
      "B03.280.500  Microcystis                     (D046931)\n",
      "B03.280.550  Nodularia                       (D046932)\n",
      "B03.280.575  Nostoc                          (D046937)\n",
      "B03.280.575.150  Nostoc commune                  (D050501)\n",
      "B03.280.575.500  Nostoc muscorum                 (D050502)\n",
      "B03.280.612  Oscillatoria                    (D054300)\n",
      "B03.280.622  Pannus                          (D000086282)\n",
      "B03.280.631  Phormidium                      (D000084242)\n",
      "B03.280.641  Planktothrix                    (D000086302)\n",
      "B03.280.650  Plectonema                      (D046938)\n",
      "B03.280.697  Spirulina                       (D053545)\n",
      "B03.280.745  Synechococcus                   (D046940)\n",
      "B03.280.750  Synechocystis                   (D046939)\n",
      "B03.280.813  Thermosynechococcus             (D000086325)\n",
      "B03.280.875  Trichodesmium                   (D000072359)\n",
      "B03.300     Endospore-Forming Bacteria      (D004725)\n",
      "B03.300.390  Gram-Positive Endospore-Forming Bacteria  (D016936)\n",
      "B03.300.390.400  Gram-Positive Endospore-Forming Rods  (D016937)\n",
      "B03.300.390.400.158  Bacillaceae                     (D001406)\n",
      "B03.300.390.400.158.108  Anoxybacillus                   (D058677)\n",
      "B03.300.390.400.158.218  Bacillus                        (D001407)\n",
      "B03.300.390.400.158.218.076  Bacillus amyloliquefaciens      (D000069977)\n",
      "B03.300.390.400.158.218.151  Bacillus anthracis              (D001408)\n",
      "B03.300.390.400.158.218.252  Bacillus cereus                 (D001409)\n",
      "B03.300.390.400.158.218.283  Bacillus clausii                (D000070019)\n",
      "B03.300.390.400.158.218.314  Bacillus coagulans              (D000070000)\n",
      "B03.300.390.400.158.218.345  Bacillus firmus                 (D000070098)\n",
      "B03.300.390.400.158.218.376  Bacillus licheniformis          (D000069976)\n",
      "B03.300.390.400.158.218.500  Bacillus megaterium             (D001410)\n",
      "B03.300.390.400.158.218.613  Bacillus pumilus                (D000069979)\n",
      "B03.300.390.400.158.218.725  Bacillus subtilis               (D001412)\n",
      "B03.300.390.400.158.218.800  Bacillus thuringiensis          (D001413)\n",
      "B03.300.390.400.158.400  Geobacillus                     (D056506)\n",
      "B03.300.390.400.158.400.400  Geobacillus stearothermophilus  (D001411)\n",
      "B03.300.390.400.158.600  Halobacillus                    (D058678)\n",
      "B03.300.390.400.158.900  Virgibacillus                   (D058679)\n",
      "B03.300.390.400.160  Brevibacillus                   (D058675)\n",
      "B03.300.390.400.200  Clostridium                     (D003013)\n",
      "B03.300.390.400.200.025  Clostridium acetobutylicum      (D046969)\n",
      "B03.300.390.400.200.100  Clostridium beijerinckii        (D046970)\n",
      "B03.300.390.400.200.130  Clostridium bifermentans        (D046971)\n",
      "B03.300.390.400.200.160  Clostridium botulinum           (D003014)\n",
      "B03.300.390.400.200.160.050  Clostridium botulinum type A    (D047013)\n",
      "B03.300.390.400.200.160.100  Clostridium botulinum type B    (D047014)\n",
      "B03.300.390.400.200.160.150  Clostridium botulinum type C    (D047968)\n",
      "B03.300.390.400.200.160.200  Clostridium botulinum type D    (D047969)\n",
      "B03.300.390.400.200.160.250  Clostridium botulinum type E    (D047970)\n",
      "B03.300.390.400.200.160.300  Clostridium botulinum type F    (D047971)\n",
      "B03.300.390.400.200.160.350  Clostridium botulinum type G    (D048568)\n",
      "B03.300.390.400.200.180  Clostridium butyricum           (D046972)\n",
      "B03.300.390.400.200.200  Clostridium cellulolyticum      (D046973)\n",
      "B03.300.390.400.200.205  Clostridium cellulovorans       (D046974)\n",
      "B03.300.390.400.200.215  Clostridium chauvoei            (D046975)\n",
      "B03.300.390.400.200.412  Clostridium histolyticum        (D047009)\n",
      "B03.300.390.400.200.493  Clostridium kluyveri            (D047010)\n",
      "B03.300.390.400.200.575  Clostridium perfringens         (D003016)\n",
      "B03.300.390.400.200.690  Clostridium septicum            (D052557)\n",
      "B03.300.390.400.200.700  Clostridium sordellii           (D047989)\n",
      "B03.300.390.400.200.710  Clostridium sticklandii         (D047991)\n",
      "B03.300.390.400.200.713  Clostridium symbiosum           (D048009)\n",
      "B03.300.390.400.200.722  Clostridium tertium             (D048010)\n",
      "B03.300.390.400.200.725  Clostridium tetani              (D003017)\n",
      "B03.300.390.400.200.740  Clostridium tetanomorphum       (D048012)\n",
      "B03.300.390.400.200.770  Clostridium thermocellum        (D048013)\n",
      "B03.300.390.400.200.800  Clostridium tyrobutyricum       (D048014)\n",
      "B03.300.390.400.500  Micromonosporaceae              (D008848)\n",
      "B03.300.390.400.500.250  Actinoplanes                    (D000079584)\n",
      "B03.300.390.400.500.500  Micromonospora                  (D008847)\n",
      "B03.300.390.400.525  Moorella                        (D058845)\n",
      "B03.300.390.400.645  Paenibacillus                   (D056507)\n",
      "B03.300.390.400.645.500  Paenibacillus larvae            (D000070059)\n",
      "B03.300.390.400.645.750  Paenibacillus polymyxa          (D000070057)\n",
      "B03.300.390.400.790  Saccharopolyspora               (D016930)\n",
      "B03.300.390.400.800  Staphylococcaceae               (D040141)\n",
      "B03.300.390.400.800.750  Staphylococcus                  (D013210)\n",
      "B03.300.390.400.800.750.100  Staphylococcus aureus           (D013211)\n",
      "B03.300.390.400.800.750.100.500  Methicillin-Resistant Staphylococcus aureus  (D055624)\n",
      "B03.300.390.400.800.750.100.750  Vancomycin-Resistant Staphylococcus aureus  (D000083123)\n",
      "B03.300.390.400.800.750.222  Staphylococcus capitis          (D000074039)\n",
      "B03.300.390.400.800.750.343  Staphylococcus epidermidis      (D013212)\n",
      "B03.300.390.400.800.750.400  Staphylococcus haemolyticus     (D041162)\n",
      "B03.300.390.400.800.750.425  Staphylococcus hominis          (D041181)\n",
      "B03.300.390.400.800.750.435  Staphylococcus hyicus           (D057788)\n",
      "B03.300.390.400.800.750.445  Staphylococcus intermedius      (D057789)\n",
      "B03.300.390.400.800.750.500  Staphylococcus lugdunensis      (D057787)\n",
      "B03.300.390.400.800.750.750  Staphylococcus saprophyticus    (D057790)\n",
      "B03.300.390.400.810  Streptomycetaceae               (D013306)\n",
      "B03.300.390.400.810.768  Streptomyces                    (D013302)\n",
      "B03.300.390.400.810.768.100  Streptomyces antibioticus       (D013303)\n",
      "B03.300.390.400.810.768.125  Streptomyces aureofaciens       (D013304)\n",
      "B03.300.390.400.810.768.200  Streptomyces coelicolor         (D048372)\n",
      "B03.300.390.400.810.768.375  Streptomyces griseus            (D013305)\n",
      "B03.300.390.400.810.768.500  Streptomyces lividans           (D048170)\n",
      "B03.300.390.400.810.768.750  Streptomyces rimosus            (D064411)\n",
      "B03.300.390.650  Pasteuria                       (D058676)\n",
      "B03.300.390.825  Sporosarcina                    (D058644)\n",
      "B03.353     Firmicutes                      (D000068536)\n",
      "B03.353.250  Acidaminococcus                 (D045850)\n",
      "B03.353.500  Bacillales                      (D056449)\n",
      "B03.353.500.050  Alicyclobacillus                (D056450)\n",
      "B03.353.500.100.309  Exiguobacterium                 (D000086183)\n",
      "B03.353.500.120  Brochothrix                     (D058640)\n",
      "B03.353.500.310  Gemella                         (D058834)\n",
      "B03.353.500.500  Listeria                        (D008087)\n",
      "B03.353.500.500.500  Listeria monocytogenes          (D008089)\n",
      "B03.353.500.700  Planococcaceae                  (D058641)\n",
      "B03.353.500.875  Thermoactinomyces               (D056509)\n",
      "B03.353.563  Caldicellulosiruptor            (D000082623)\n",
      "B03.353.625  Clostridiales                   (D000068016)\n",
      "B03.353.625.250  Butyrivibrio                    (D045856)\n",
      "B03.353.625.250.500  Butyrivibrio fibrisolvens       (D000070097)\n",
      "B03.353.625.375  Clostridiaceae                  (D000073561)\n",
      "B03.353.625.375.750  Sarcina                         (D012505)\n",
      "B03.353.625.657  Clostridioides                  (D000082682)\n",
      "B03.353.625.657.500  Clostridioides difficile        (D016360)\n",
      "B03.353.625.750  Eubacterium                     (D005051)\n",
      "B03.353.625.766  Faecalibacterium                (D000070036)\n",
      "B03.353.625.766.500  Faecalibacterium prausnitzii    (D000070037)\n",
      "B03.353.625.774  Neglecta                        (D000086242)\n",
      "B03.353.625.782  Peptococcaceae                  (D010459)\n",
      "B03.353.625.782.149  Desulfitobacterium              (D045858)\n",
      "B03.353.625.782.297  Desulfotomaculum                (D045859)\n",
      "B03.353.625.782.593  Peptococcus                     (D010460)\n",
      "B03.353.625.798  Peptostreptococcus              (D010462)\n",
      "B03.353.625.806  Ruminococcus                    (D045857)\n",
      "B03.353.688  Erysipelothrix                  (D004888)\n",
      "B03.353.750  Lactobacillales                 (D056546)\n",
      "B03.353.750.030  Aerococcaceae                   (D056567)\n",
      "B03.353.750.030.014  Abiotrophia                     (D058835)\n",
      "B03.353.750.030.030  Aerococcus                      (D056570)\n",
      "B03.353.750.130  Carnobacteriaceae               (D056568)\n",
      "B03.353.750.130.150  Carnobacterium                  (D056573)\n",
      "B03.353.750.250  Enterococcaceae                 (D056569)\n",
      "B03.353.750.250.250  Enterococcus                    (D016983)\n",
      "B03.353.750.250.250.280  Enterococcus faecalis           (D013293)\n",
      "B03.353.750.250.250.300  Enterococcus faecium            (D016984)\n",
      "B03.353.750.250.250.475  Enterococcus hirae              (D000070076)\n",
      "B03.353.750.250.250.650  Vancomycin-Resistant Enterococci  (D065507)\n",
      "B03.353.750.450  Lactobacillaceae                (D007777)\n",
      "B03.353.750.450.475  Lactobacillus                   (D007778)\n",
      "B03.353.750.450.475.100  Lactobacillus acidophilus       (D007779)\n",
      "B03.353.750.450.475.238  Lactobacillus crispatus         (D000070003)\n",
      "B03.353.750.450.475.250  Lactobacillus delbrueckii       (D048668)\n",
      "B03.353.750.450.475.363  Lactobacillus gasseri           (D000069998)\n",
      "B03.353.750.450.475.400  Lactobacillus helveticus        (D052198)\n",
      "B03.353.750.450.475.453  Lactobacillus johnsonii         (D000069999)\n",
      "B03.353.750.450.475.506  Lactobacillus leichmannii       (D052199)\n",
      "B03.353.750.450.475.586  Lactobacillus pentosus          (D000070018)\n",
      "B03.353.750.450.475.612  Lactobacillus plantarum         (D048191)\n",
      "B03.353.750.450.606  Lacticaseibacillus              (D000093622)\n",
      "B03.353.750.450.606.500  Lacticaseibacillus casei        (D007780)\n",
      "B03.353.750.450.606.750  Lacticaseibacillus paracasei    (D000069980)\n",
      "B03.353.750.450.606.875  Lacticaseibacillus rhamnosus    (D052201)\n",
      "B03.353.750.450.623  Latilactobacillus sakei         (D000070256)\n",
      "B03.353.750.450.639  Levilactobacillus brevis        (D052196)\n",
      "B03.353.750.450.656  Ligilactobacillus salivarius    (D000069996)\n",
      "B03.353.750.450.672  Limosilactobacillus fermentum   (D052197)\n",
      "B03.353.750.450.705  Limosilactobacillus reuteri     (D052200)\n",
      "B03.353.750.450.737  Pediococcus                     (D010376)\n",
      "B03.353.750.450.737.500  Pediococcus acidilactici        (D000070016)\n",
      "B03.353.750.450.737.750  Pediococcus pentosaceus         (D000070006)\n",
      "B03.353.750.475  Leuconostocaceae                (D056584)\n",
      "B03.353.750.475.450  Leuconostoc                     (D007934)\n",
      "B03.353.750.475.450.500  Leuconostoc mesenteroides       (D000070007)\n",
      "B03.353.750.475.600  Oenococcus                      (D056585)\n",
      "B03.353.750.475.900  Weissella                       (D058836)\n",
      "B03.353.750.737  Streptococcaceae                (D013289)\n",
      "B03.353.750.737.500  Lactococcus                     (D017045)\n",
      "B03.353.750.737.500.400  Lactococcus lactis              (D013294)\n",
      "B03.353.750.737.872  Streptococcus                   (D013291)\n",
      "B03.353.750.737.872.100  Streptococcus agalactiae        (D013292)\n",
      "B03.353.750.737.872.150  Streptococcus bovis             (D016985)\n",
      "B03.353.750.737.872.225  Streptococcus equi              (D018502)\n",
      "B03.353.750.737.872.243  Streptococcus gallolyticus      (D000070099)\n",
      "B03.353.750.737.872.243.500  Streptococcus gallolyticus subspecies gallolyticus  (D000070197)\n",
      "B03.353.750.737.872.260  Streptococcus gordonii          (D054773)\n",
      "B03.353.750.737.872.405  Streptococcus iniae             (D000070058)\n",
      "B03.353.750.737.872.550  Streptococcus pneumoniae        (D013296)\n",
      "B03.353.750.737.872.575  Streptococcus pyogenes          (D013297)\n",
      "B03.353.750.737.872.663  Streptococcus salivarius        (D000070001)\n",
      "B03.353.750.737.872.750  Streptococcus suis              (D017011)\n",
      "B03.353.750.737.872.800  Streptococcus thermophilus      (D048249)\n",
      "B03.353.750.737.872.875  Viridans Streptococci           (D034363)\n",
      "B03.353.750.737.872.875.475  Streptococcus milleri Group     (D034364)\n",
      "B03.353.750.737.872.875.475.080  Streptococcus anginosus         (D034366)\n",
      "B03.353.750.737.872.875.475.150  Streptococcus constellatus      (D034365)\n",
      "B03.353.750.737.872.875.475.400  Streptococcus intermedius       (D034367)\n",
      "B03.353.750.737.872.875.500  Streptococcus mitis             (D034361)\n",
      "B03.353.750.737.872.875.520  Streptococcus mutans            (D013295)\n",
      "B03.353.750.737.872.875.600  Streptococcus oralis            (D019533)\n",
      "B03.353.750.737.872.875.700  Streptococcus sanguis           (D013298)\n",
      "B03.353.750.737.872.875.750  Streptococcus sobrinus          (D016986)\n",
      "B03.353.937  Thermoanaerobacter              (D046968)\n",
      "B03.353.968  Thermoanaerobacterium           (D046950)\n",
      "B03.353.984  Veillonellaceae                 (D045854)\n",
      "B03.370     Fusobacteria                    (D045182)\n",
      "B03.370.250  Fusobacterium                   (D005673)\n",
      "B03.370.250.400  Fusobacterium necrophorum       (D005675)\n",
      "B03.370.250.500  Fusobacterium nucleatum         (D016967)\n",
      "B03.370.437  Leptotrichia                    (D045183)\n",
      "B03.370.600  Propionigenium                  (D045184)\n",
      "B03.370.700  Streptobacillus                 (D013288)\n",
      "B03.440     Gram-Negative Bacteria          (D006090)\n",
      "B03.440.025  Aquifex                         (D000086142)\n",
      "B03.440.050  Arcobacter                      (D020640)\n",
      "B03.440.080  Bacteroidetes                   (D041963)\n",
      "B03.440.080.094  Bacteroidaceae                  (D001438)\n",
      "B03.440.080.094.152  Bacteroides                     (D001439)\n",
      "B03.440.080.094.152.400  Bacteroides fragilis            (D001441)\n",
      "B03.440.080.094.152.700  Bacteroides thetaiotaomicron    (D000070077)\n",
      "B03.440.080.094.389  Mediterranea                    (D000086222)\n",
      "B03.440.080.094.625  Porphyromonas                   (D017861)\n",
      "B03.440.080.094.625.350  Porphyromonas endodontalis      (D045202)\n",
      "B03.440.080.094.625.515  Porphyromonas gingivalis        (D016966)\n",
      "B03.440.080.094.640  Prevotella                      (D018720)\n",
      "B03.440.080.094.640.375  Prevotella intermedia           (D018713)\n",
      "B03.440.080.094.640.500  Prevotella melaninogenica       (D001443)\n",
      "B03.440.080.094.640.550  Prevotella nigrescens           (D045242)\n",
      "B03.440.080.094.640.700  Prevotella ruminicola           (D045243)\n",
      "B03.440.080.130  Cytophagaceae                   (D041964)\n",
      "B03.440.080.130.150  Cytophaga                       (D003589)\n",
      "B03.440.080.130.200  Flexibacter                     (D045248)\n",
      "B03.440.080.190  Flavobacteriaceae               (D041965)\n",
      "B03.440.080.190.100  Capnocytophaga                  (D002206)\n",
      "B03.440.080.190.120  Chryseobacterium                (D045247)\n",
      "B03.440.080.190.250  Flavobacterium                  (D005417)\n",
      "B03.440.080.190.600  Ornithobacterium                (D045246)\n",
      "B03.440.080.190.700  Riemerella                      (D060494)\n",
      "B03.440.080.190.800  Tenacibaculum                   (D060493)\n",
      "B03.440.080.600  Pedobacter                      (D060485)\n",
      "B03.440.080.750  Rhodothermus                    (D045245)\n",
      "B03.440.080.800  Sphingobacterium                (D045249)\n",
      "B03.440.080.850  Tannerella                      (D000084422)\n",
      "B03.440.080.850.500  Tannerella forsythia            (D000070436)\n",
      "B03.440.090  Bartonellaceae                  (D001475)\n",
      "B03.440.090.100  Bartonella                      (D001473)\n",
      "B03.440.090.100.080  Bartonella bacilliformis        (D045503)\n",
      "B03.440.090.100.350  Bartonella henselae             (D018416)\n",
      "B03.440.090.100.650  Bartonella quintana             (D018413)\n",
      "B03.440.097  Brachyspira                     (D017400)\n",
      "B03.440.097.100  Brachyspira hyodysenteriae      (D017401)\n",
      "B03.440.100  Buchnera                        (D020635)\n",
      "B03.440.180  Campylobacter                   (D002167)\n",
      "B03.440.180.200  Campylobacter coli              (D017000)\n",
      "B03.440.180.325  Campylobacter fetus             (D002168)\n",
      "B03.440.180.375  Campylobacter hyointestinalis   (D044862)\n",
      "B03.440.180.425  Campylobacter jejuni            (D016123)\n",
      "B03.440.180.500  Campylobacter lari              (D044863)\n",
      "B03.440.180.650  Campylobacter rectus            (D044883)\n",
      "B03.440.180.700  Campylobacter sputorum          (D044884)\n",
      "B03.440.180.800  Campylobacter upsaliensis       (D044885)\n",
      "B03.440.190  Chlamydiales                    (D002695)\n",
      "B03.440.190.190  Chlamydiaceae                   (D002693)\n",
      "B03.440.190.190.190  Chlamydia                       (D002689)\n",
      "B03.440.190.190.190.500  Chlamydia muridarum             (D023502)\n",
      "B03.440.190.190.190.750  Chlamydia trachomatis           (D002692)\n",
      "B03.440.190.190.230  Chlamydophila                   (D023501)\n",
      "B03.440.190.190.230.249  Chlamydophila pneumoniae        (D016993)\n",
      "B03.440.190.190.230.500  Chlamydophila psittaci          (D002691)\n",
      "B03.440.305  Fibrobacteres                   (D061306)\n",
      "B03.440.305.500  Fibrobacter                     (D045848)\n",
      "B03.440.400  Gram-Negative Aerobic Bacteria  (D006088)\n",
      "B03.440.400.280  Caulobacter                     (D016934)\n",
      "B03.440.400.280.280  Caulobacter crescentus          (D016935)\n",
      "B03.440.400.400  Gallionellaceae                 (D043372)\n",
      "B03.440.400.425  Gram-Negative Aerobic Rods and Cocci  (D016946)\n",
      "B03.440.400.425.100  Acetobacteraceae                (D016947)\n",
      "B03.440.400.425.100.100  Acetobacter                     (D000091)\n",
      "B03.440.400.425.100.110  Acidiphilium                    (D041801)\n",
      "B03.440.400.425.100.550  Gluconobacter                   (D020576)\n",
      "B03.440.400.425.100.550.550  Gluconobacter oxydans           (D020600)\n",
      "B03.440.400.425.103  Acidithiobacillus               (D042763)\n",
      "B03.440.400.425.103.800  Acidithiobacillus thiooxidans   (D013856)\n",
      "B03.440.400.425.115  Alcaligenaceae                  (D042401)\n",
      "B03.440.400.425.115.030  Achromobacter                   (D042402)\n",
      "B03.440.400.425.115.030.100  Achromobacter cycloclastes      (D042422)\n",
      "B03.440.400.425.115.030.120  Achromobacter denitrificans     (D042441)\n",
      "B03.440.400.425.115.050  Alcaligenes                     (D000421)\n",
      "B03.440.400.425.115.050.200  Alcaligenes faecalis            (D042481)\n",
      "B03.440.400.425.115.425  Bordetella                      (D001884)\n",
      "B03.440.400.425.115.425.024  Bordetella avium                (D042482)\n",
      "B03.440.400.425.115.425.050  Bordetella bronchiseptica       (D016950)\n",
      "B03.440.400.425.115.425.590  Bordetella parapertussis        (D042483)\n",
      "B03.440.400.425.115.425.600  Bordetella pertussis            (D001886)\n",
      "B03.440.400.425.115.800  Taylorella                      (D042502)\n",
      "B03.440.400.425.115.800.200  Taylorella equigenitalis        (D020625)\n",
      "B03.440.400.425.126  Alteromonas                     (D020591)\n",
      "B03.440.400.425.127  Azorhizobium                    (D020583)\n",
      "B03.440.400.425.127.100  Azorhizobium caulinodans        (D020599)\n",
      "B03.440.400.425.180  Bdellovibrio                    (D001501)\n",
      "B03.440.400.425.180.500  Bdellovibrio bacteriovorus      (D000070060)\n",
      "B03.440.400.425.200  Bradyrhizobiaceae               (D041841)\n",
      "B03.440.400.425.200.040  Afipia                          (D020622)\n",
      "B03.440.400.425.200.090  Bradyrhizobium                  (D020369)\n",
      "B03.440.400.425.200.520  Nitrobacter                     (D009576)\n",
      "B03.440.400.425.200.700  Rhodopseudomonas                (D012241)\n",
      "B03.440.400.425.215  Brucellaceae                    (D042021)\n",
      "B03.440.400.425.215.500  Brucella                        (D002002)\n",
      "B03.440.400.425.215.500.100  Brucella abortus                (D002003)\n",
      "B03.440.400.425.215.500.150  Brucella canis                  (D041002)\n",
      "B03.440.400.425.215.500.500  Brucella melitensis             (D017347)\n",
      "B03.440.400.425.215.500.625  Brucella ovis                   (D042061)\n",
      "B03.440.400.425.215.500.750  Brucella suis                   (D041021)\n",
      "B03.440.400.425.251  Burkholderiaceae                (D042521)\n",
      "B03.440.400.425.251.200  Cupriavidus                     (D049921)\n",
      "B03.440.400.425.251.200.200  Cupriavidus necator             (D020118)\n",
      "B03.440.400.425.288  Caulobacteraceae                (D042063)\n",
      "B03.440.400.425.293  Comamonadaceae                  (D042621)\n",
      "B03.440.400.425.293.150  Comamonas                       (D020585)\n",
      "B03.440.400.425.293.150.750  Comamonas testosteroni          (D020604)\n",
      "B03.440.400.425.293.170  Delftia                         (D042742)\n",
      "B03.440.400.425.293.170.030  Delftia acidovorans             (D020605)\n",
      "B03.440.400.425.293.750  Sphaerotilus                    (D042744)\n",
      "B03.440.400.425.297  Coxiellaceae                    (D044125)\n",
      "B03.440.400.425.297.150  Coxiella                        (D003381)\n",
      "B03.440.400.425.297.150.100  Coxiella burnetii               (D016997)\n",
      "B03.440.400.425.340  Francisella                     (D005603)\n",
      "B03.440.400.425.340.590  Francisella tularensis          (D005604)\n",
      "B03.440.400.425.365  Gluconacetobacter               (D041821)\n",
      "B03.440.400.425.365.500  Gluconacetobacter xylinus       (D041822)\n",
      "B03.440.400.425.377  Halomonadaceae                  (D044129)\n",
      "B03.440.400.425.377.249  Chromohalobacter                (D055353)\n",
      "B03.440.400.425.377.500  Halomonas                       (D020632)\n",
      "B03.440.400.425.395  Halothiobacillus                (D044222)\n",
      "B03.440.400.425.450  Legionellaceae                  (D016951)\n",
      "B03.440.400.425.450.450  Legionella                      (D007875)\n",
      "B03.440.400.425.450.450.400  Legionella longbeachae          (D044128)\n",
      "B03.440.400.425.450.450.500  Legionella pneumophila          (D016952)\n",
      "B03.440.400.425.475  Leptospiraceae                  (D016953)\n",
      "B03.440.400.425.475.475  Leptospira                      (D007919)\n",
      "B03.440.400.425.475.475.400  Leptospira interrogans          (D007921)\n",
      "B03.440.400.425.475.475.400.050  Leptospira interrogans serovar australis  (D041281)\n",
      "B03.440.400.425.475.475.400.055  Leptospira interrogans serovar autumnalis  (D041301)\n",
      "B03.440.400.425.475.475.400.150  Leptospira interrogans serovar canicola  (D041341)\n",
      "B03.440.400.425.475.475.400.400  Leptospira interrogans serovar hebdomadis  (D041343)\n",
      "B03.440.400.425.475.475.400.420  Leptospira interrogans serovar icterohaemorrhagiae  (D041383)\n",
      "B03.440.400.425.475.475.400.650  Leptospira interrogans serovar pomona  (D041401)\n",
      "B03.440.400.425.481  Leptothrix                      (D042743)\n",
      "B03.440.400.425.487  Methylobacteriaceae             (D042082)\n",
      "B03.440.400.425.487.500  Methylobacterium                (D020580)\n",
      "B03.440.400.425.487.500.150  Methylobacterium extorquens     (D020601)\n",
      "B03.440.400.425.500  Methylococcaceae                (D008772)\n",
      "B03.440.400.425.500.500  Methylococcus                   (D020618)\n",
      "B03.440.400.425.500.500.100  Methylococcus capsulatus        (D020619)\n",
      "B03.440.400.425.500.505  Methylomonas                    (D020620)\n",
      "B03.440.400.425.525  Methylophilaceae                (D042762)\n",
      "B03.440.400.425.525.500  Methylobacillus                 (D020631)\n",
      "B03.440.400.425.525.520  Methylophilus                   (D020629)\n",
      "B03.440.400.425.525.520.500  Methylophilus methylotrophus    (D020630)\n",
      "B03.440.400.425.537  Moraxellaceae                   (D044007)\n",
      "B03.440.400.425.537.050  Acinetobacter                   (D000150)\n",
      "B03.440.400.425.537.050.099  Acinetobacter baumannii         (D040981)\n",
      "B03.440.400.425.537.050.200  Acinetobacter calcoaceticus     (D016954)\n",
      "B03.440.400.425.537.525  Moraxella                       (D009016)\n",
      "B03.440.400.425.537.525.200  Moraxella catarrhalis           (D001936)\n",
      "B03.440.400.425.537.525.500  Moraxella bovis                 (D016955)\n",
      "B03.440.400.425.537.650  Psychrobacter                   (D044096)\n",
      "B03.440.400.425.550  Neisseriaceae                   (D009346)\n",
      "B03.440.400.425.550.400  Kingella                        (D017881)\n",
      "B03.440.400.425.550.400.375  Kingella kingae                 (D017882)\n",
      "B03.440.400.425.550.550  Neisseria                       (D009343)\n",
      "B03.440.400.425.550.550.125  Neisseria cinerea               (D042804)\n",
      "B03.440.400.425.550.550.299  Neisseria elongata              (D042805)\n",
      "B03.440.400.425.550.550.474  Neisseria gonorrhoeae           (D009344)\n",
      "B03.440.400.425.550.550.600  Neisseria lactamica             (D042806)\n",
      "B03.440.400.425.550.550.641  Neisseria meningitidis          (D009345)\n",
      "B03.440.400.425.550.550.641.700  Neisseria meningitidis, Serogroup A  (D038381)\n",
      "B03.440.400.425.550.550.641.710  Neisseria meningitidis, Serogroup B  (D038541)\n",
      "B03.440.400.425.550.550.641.720  Neisseria meningitidis, Serogroup C  (D038542)\n",
      "B03.440.400.425.550.550.641.800  Neisseria meningitidis, Serogroup W-135  (D038582)\n",
      "B03.440.400.425.550.550.641.820  Neisseria meningitidis, Serogroup Y  (D038581)\n",
      "B03.440.400.425.550.550.660  Neisseria mucosa                (D042807)\n",
      "B03.440.400.425.550.550.800  Neisseria sicca                 (D042808)\n",
      "B03.440.400.425.562  Nitrosomonadaceae               (D043369)\n",
      "B03.440.400.425.562.500  Nitrosomonas                    (D009606)\n",
      "B03.440.400.425.562.500.200  Nitrosomonas europaea           (D043370)\n",
      "B03.440.400.425.575  Ochrobactrum                    (D042041)\n",
      "B03.440.400.425.575.650  Ochrobactrum anthropi           (D020621)\n",
      "B03.440.400.425.587  Oxalobacteraceae                (D042802)\n",
      "B03.440.400.425.587.400  Herbaspirillum                  (D042803)\n",
      "B03.440.400.425.600  Paracoccus                      (D010230)\n",
      "B03.440.400.425.600.380  Paracoccus denitrificans        (D010231)\n",
      "B03.440.400.425.600.690  Paracoccus pantotrophus         (D042084)\n",
      "B03.440.400.425.622  Pseudoalteromonas               (D031161)\n",
      "B03.440.400.425.625  Pseudomonadaceae                (D011548)\n",
      "B03.440.400.425.625.050  Azotobacter                     (D001395)\n",
      "B03.440.400.425.625.050.850  Azotobacter vinelandii          (D016948)\n",
      "B03.440.400.425.625.150  Cellvibrio                      (D020633)\n",
      "B03.440.400.425.625.625  Pseudomonas                     (D011549)\n",
      "B03.440.400.425.625.625.100  Pseudomonas aeruginosa          (D011550)\n",
      "B03.440.400.425.625.625.120  Pseudomonas alcaligenes         (D044202)\n",
      "B03.440.400.425.625.625.223  Pseudomonas chlororaphis        (D000070096)\n",
      "B03.440.400.425.625.625.325  Pseudomonas fluorescens         (D011551)\n",
      "B03.440.400.425.625.625.350  Pseudomonas fragi               (D044206)\n",
      "B03.440.400.425.625.625.500  Pseudomonas mendocina           (D044203)\n",
      "B03.440.400.425.625.625.650  Pseudomonas oleovorans          (D044204)\n",
      "B03.440.400.425.625.625.662  Pseudomonas pseudoalcaligenes   (D044205)\n",
      "B03.440.400.425.625.625.675  Pseudomonas putida              (D016958)\n",
      "B03.440.400.425.625.625.750  Pseudomonas stutzeri            (D044223)\n",
      "B03.440.400.425.625.625.770  Pseudomonas syringae            (D044224)\n",
      "B03.440.400.425.700  Rhizobiaceae                    (D012230)\n",
      "B03.440.400.425.700.024  Agrobacterium                   (D060054)\n",
      "B03.440.400.425.700.024.050  Agrobacterium tumefaciens       (D016960)\n",
      "B03.440.400.425.700.800  Rhizobium                       (D012231)\n",
      "B03.440.400.425.700.800.337  Rhizobium etli                  (D042163)\n",
      "B03.440.400.425.700.800.450  Rhizobium leguminosarum         (D016961)\n",
      "B03.440.400.425.700.800.600  Rhizobium phaseoli              (D042161)\n",
      "B03.440.400.425.700.800.800  Rhizobium tropici               (D042162)\n",
      "B03.440.400.425.700.887  Sinorhizobium                   (D020579)\n",
      "B03.440.400.425.700.887.249  Sinorhizobium fredii            (D042181)\n",
      "B03.440.400.425.700.887.500  Sinorhizobium meliloti          (D016962)\n",
      "B03.440.400.425.708  Rhodospirillaceae               (D012244)\n",
      "B03.440.400.425.708.100  Azospirillum                    (D016949)\n",
      "B03.440.400.425.708.100.020  Azospirillum brasilense         (D015806)\n",
      "B03.440.400.425.708.100.510  Azospirillum lipoferum          (D042081)\n",
      "B03.440.400.425.708.500  Magnetospirillum                (D042201)\n",
      "B03.440.400.425.708.733  Rhodospirillum                  (D012246)\n",
      "B03.440.400.425.708.733.150  Rhodospirillum centenum         (D042182)\n",
      "B03.440.400.425.708.733.650  Rhodospirillum rubrum           (D012247)\n",
      "B03.440.400.425.750  Sphingomonas                    (D020578)\n",
      "B03.440.400.425.875  Thermus                         (D013824)\n",
      "B03.440.400.425.875.875  Thermus thermophilus            (D016963)\n",
      "B03.440.400.425.935  Xanthobacter                    (D020584)\n",
      "B03.440.400.425.967  Xanthomonadaceae                (D044166)\n",
      "B03.440.400.425.967.500  Lysobacter                      (D053528)\n",
      "B03.440.400.425.967.750  Stenotrophomonas                (D020588)\n",
      "B03.440.400.425.967.750.500  Stenotrophomonas maltophilia    (D020615)\n",
      "B03.440.400.425.967.930  Xanthomonas                     (D014974)\n",
      "B03.440.400.425.967.930.100  Xanthomonas campestris          (D016959)\n",
      "B03.440.400.425.967.930.850  Xanthomonas vesicatoria         (D044184)\n",
      "B03.440.400.425.967.950  Xylella                         (D044167)\n",
      "B03.440.400.425.983  Zoogloea                        (D015045)\n",
      "B03.440.400.450  Gram-Negative Chemolithotrophic Bacteria  (D006091)\n",
      "B03.440.400.450.800  Thiobacillus                    (D013855)\n",
      "B03.440.400.645  Thiotrichaceae                  (D041962)\n",
      "B03.440.400.645.100  Beggiatoa                       (D053524)\n",
      "B03.440.400.645.800  Thiothrix                       (D053525)\n",
      "B03.440.400.840  Vitreoscilla                    (D020589)\n",
      "B03.440.425  Gram-Negative Anaerobic Bacteria  (D006089)\n",
      "B03.440.425.200  Desulfovibrionaceae             (D000073564)\n",
      "B03.440.425.400  Gram-Negative Anaerobic Cocci   (D016964)\n",
      "B03.440.425.400.500  Megasphaera                     (D045851)\n",
      "B03.440.425.400.500.500  Megasphaera elsdenii            (D000070101)\n",
      "B03.440.425.400.750  Thiocapsa                       (D020586)\n",
      "B03.440.425.400.750.660  Thiocapsa roseopersicina        (D020616)\n",
      "B03.440.425.410  Gram-Negative Anaerobic Straight, Curved, and Helical Rods  (D016965)\n",
      "B03.440.425.410.145  Anaerobiospirillum              (D044064)\n",
      "B03.440.425.410.200  Bilophila                       (D045282)\n",
      "B03.440.425.410.290  Chromatium                      (D002844)\n",
      "B03.440.425.410.350  Desulfovibrio                   (D003901)\n",
      "B03.440.425.410.350.040  Desulfovibrio africanus         (D045284)\n",
      "B03.440.425.410.350.150  Desulfovibrio desulfuricans     (D045285)\n",
      "B03.440.425.410.350.350  Desulfovibrio gigas             (D045286)\n",
      "B03.440.425.410.350.875  Desulfovibrio vulgaris          (D016969)\n",
      "B03.440.425.410.360  Desulfuromonas                  (D045287)\n",
      "B03.440.425.410.400  Dichelobacter nodosus           (D020376)\n",
      "B03.440.425.410.405  Ectothiorhodospiraceae          (D042001)\n",
      "B03.440.425.410.405.500  Ectothiorhodospira              (D020587)\n",
      "B03.440.425.410.405.500.720  Ectothiorhodospira shaposhnikovii  (D020608)\n",
      "B03.440.425.410.405.750  Halorhodospira halophila        (D020607)\n",
      "B03.440.425.410.430  Geobacter                       (D045302)\n",
      "B03.440.425.410.580  Oxalobacter formigenes          (D020624)\n",
      "B03.440.425.410.711  Spirochaetaceae                 (D013143)\n",
      "B03.440.425.410.711.193  Borrelia                        (D001898)\n",
      "B03.440.425.410.711.193.150  Borrelia burgdorferi Group      (D015748)\n",
      "B03.440.425.410.711.193.150.125  Borrelia burgdorferi            (D025065)\n",
      "B03.440.425.410.711.679  Spirochaeta                     (D013142)\n",
      "B03.440.425.410.711.795  Treponema                       (D014208)\n",
      "B03.440.425.410.711.795.700  Treponema denticola             (D048373)\n",
      "B03.440.425.410.711.795.840  Treponema pallidum              (D014210)\n",
      "B03.440.425.410.722  Succinivibrionaceae             (D044063)\n",
      "B03.440.425.410.745  Thauera                         (D020628)\n",
      "B03.440.425.410.748  Thermotoga                      (D000084444)\n",
      "B03.440.425.410.748.500  Thermotoga maritima             (D020124)\n",
      "B03.440.425.410.748.750  Thermotoga neapolitana          (D046949)\n",
      "B03.440.425.410.800.755  Pectinatus                      (D045852)\n",
      "B03.440.425.410.800.877  Selenomonas                     (D020357)\n",
      "B03.440.425.410.800.938  Veillonella                     (D014678)\n",
      "B03.440.425.410.875  Wolinella                       (D016968)\n",
      "B03.440.450  Gram-Negative Facultatively Anaerobic Rods  (D006092)\n",
      "B03.440.450.009  Actinobacillus                  (D000188)\n",
      "B03.440.450.009.200  Actinobacillus equuli           (D044131)\n",
      "B03.440.450.009.580  Actinobacillus pleuropneumoniae  (D016977)\n",
      "B03.440.450.009.675  Actinobacillus seminis          (D044133)\n",
      "B03.440.450.009.700  Actinobacillus suis             (D044132)\n",
      "B03.440.450.019  Aeromonadaceae                  (D044044)\n",
      "B03.440.450.019.025  Aeromonas                       (D000333)\n",
      "B03.440.450.019.025.150  Aeromonas caviae                (D057585)\n",
      "B03.440.450.019.025.380  Aeromonas hydrophila            (D016980)\n",
      "B03.440.450.019.025.690  Aeromonas salmonicida           (D048409)\n",
      "B03.440.450.019.025.845  Aeromonas veronii               (D000070079)\n",
      "B03.440.450.040  Azoarcus                        (D020627)\n",
      "B03.440.450.342  Cardiobacteriaceae              (D044065)\n",
      "B03.440.450.342.100  Cardiobacterium                 (D044066)\n",
      "B03.440.450.360  Chromobacterium                 (D002861)\n",
      "B03.440.450.400  Eikenella                       (D016970)\n",
      "B03.440.450.400.275  Eikenella corrodens             (D016124)\n",
      "B03.440.450.425  Enterobacteriaceae              (D004755)\n",
      "B03.440.450.425.180  Calymmatobacterium              (D002157)\n",
      "B03.440.450.425.189  Carbapenem-Resistant Enterobacteriaceae  (D000073182)\n",
      "B03.440.450.425.200  Citrobacter                     (D002954)\n",
      "B03.440.450.425.200.275  Citrobacter freundii            (D016971)\n",
      "B03.440.450.425.200.475  Citrobacter koseri              (D041041)\n",
      "B03.440.450.425.200.737  Citrobacter rodentium           (D044082)\n",
      "B03.440.450.425.215  Cronobacter                     (D059126)\n",
      "B03.440.450.425.215.150  Cronobacter sakazakii           (D044083)\n",
      "B03.440.450.425.260  Edwardsiella                    (D020594)\n",
      "B03.440.450.425.260.340  Edwardsiella ictaluri           (D020610)\n",
      "B03.440.450.425.260.750  Edwardsiella tarda              (D020609)\n",
      "B03.440.450.425.275  Enterobacter                    (D004754)\n",
      "B03.440.450.425.275.099  Enterobacter aerogenes          (D021902)\n",
      "B03.440.450.425.275.200  Enterobacter cloacae            (D016972)\n",
      "B03.440.450.425.300  Erwinia                         (D004885)\n",
      "B03.440.450.425.300.050  Erwinia amylovora               (D044084)\n",
      "B03.440.450.425.325  Escherichia                     (D004925)\n",
      "B03.440.450.425.325.300  Escherichia coli                (D004926)\n",
      "B03.440.450.425.325.300.330  Enteropathogenic Escherichia coli  (D054308)\n",
      "B03.440.450.425.325.300.340  Enterotoxigenic Escherichia coli  (D054307)\n",
      "B03.440.450.425.325.300.360  Escherichia coli K12            (D048168)\n",
      "B03.440.450.425.325.300.580  Extraintestinal Pathogenic Escherichia coli  (D000072282)\n",
      "B03.440.450.425.325.300.580.500  Uropathogenic Escherichia coli  (D056689)\n",
      "B03.440.450.425.325.300.800  Shiga-Toxigenic Escherichia coli  (D054323)\n",
      "B03.440.450.425.325.300.800.250  Enterohemorrhagic Escherichia coli  (D054324)\n",
      "B03.440.450.425.325.300.800.250.250  Escherichia coli O104           (D000069981)\n",
      "B03.440.450.425.325.300.800.250.500  Escherichia coli O157           (D019453)\n",
      "B03.440.450.425.375  Hafnia                          (D020595)\n",
      "B03.440.450.425.375.050  Hafnia alvei                    (D020611)\n",
      "B03.440.450.425.425  Klebsiella                      (D007709)\n",
      "B03.440.450.425.425.580  Klebsiella oxytoca              (D041121)\n",
      "B03.440.450.425.425.600  Klebsiella pneumoniae           (D007711)\n",
      "B03.440.450.425.435  Kluyvera                        (D044086)\n",
      "B03.440.450.425.500  Morganella                      (D020596)\n",
      "B03.440.450.425.500.500  Morganella morganii             (D020613)\n",
      "B03.440.450.425.580  Pantoea                         (D020636)\n",
      "B03.440.450.425.585  Pectobacterium                  (D044043)\n",
      "B03.440.450.425.585.120  Pectobacterium carotovorum      (D016973)\n",
      "B03.440.450.425.590  Photorhabdus                    (D020637)\n",
      "B03.440.450.425.595  Plesiomonas                     (D016981)\n",
      "B03.440.450.425.600  Proteus                         (D011511)\n",
      "B03.440.450.425.600.501  Proteus mirabilis               (D011513)\n",
      "B03.440.450.425.600.650  Proteus penneri                 (D044095)\n",
      "B03.440.450.425.600.800  Proteus vulgaris                (D011514)\n",
      "B03.440.450.425.670  Providencia                     (D011532)\n",
      "B03.440.450.425.800  Salmonella                      (D012475)\n",
      "B03.440.450.425.800.100  Salmonella arizonae             (D012476)\n",
      "B03.440.450.425.800.200  Salmonella enterica             (D019779)\n",
      "B03.440.450.425.800.200.300  Salmonella enteritidis          (D012477)\n",
      "B03.440.450.425.800.200.600  Salmonella paratyphi A          (D012482)\n",
      "B03.440.450.425.800.200.700  Salmonella paratyphi B          (D012484)\n",
      "B03.440.450.425.800.200.750  Salmonella paratyphi C          (D012479)\n",
      "B03.440.450.425.800.200.800  Salmonella typhi                (D012485)\n",
      "B03.440.450.425.800.200.825  Salmonella typhimurium          (D012486)\n",
      "B03.440.450.425.814  Serratia                        (D012705)\n",
      "B03.440.450.425.814.620  Serratia liquefaciens           (D044090)\n",
      "B03.440.450.425.814.664  Serratia marcescens             (D012706)\n",
      "B03.440.450.425.850  Shigella                        (D012760)\n",
      "B03.440.450.425.850.175  Shigella boydii                 (D012761)\n",
      "B03.440.450.425.850.350  Shigella dysenteriae            (D012762)\n",
      "B03.440.450.425.850.450  Shigella flexneri               (D012763)\n",
      "B03.440.450.425.850.800  Shigella sonnei                 (D012764)\n",
      "B03.440.450.425.887  Wigglesworthia                  (D044106)\n",
      "B03.440.450.425.890  Xenorhabdus                     (D020593)\n",
      "B03.440.450.425.900  Yersinia                        (D015007)\n",
      "B03.440.450.425.900.300  Yersinia enterocolitica         (D015008)\n",
      "B03.440.450.425.900.600  Yersinia pestis                 (D015010)\n",
      "B03.440.450.425.900.615  Yersinia pseudotuberculosis     (D015011)\n",
      "B03.440.450.425.900.650  Yersinia ruckeri                (D044124)\n",
      "B03.440.450.510  Moritella                       (D034221)\n",
      "B03.440.450.600  Pasteurellaceae                 (D015661)\n",
      "B03.440.450.600.224  Aggregatibacter                 (D064206)\n",
      "B03.440.450.600.224.500  Aggregatibacter actinomycetemcomitans  (D016976)\n",
      "B03.440.450.600.224.750  Aggregatibacter aphrophilus     (D064207)\n",
      "B03.440.450.600.224.875  Aggregatibacter segnis          (D064208)\n",
      "B03.440.450.600.450  Haemophilus                     (D006190)\n",
      "B03.440.450.600.450.125  Haemophilus ducreyi             (D006191)\n",
      "B03.440.450.600.450.330  Haemophilus influenzae          (D006193)\n",
      "B03.440.450.600.450.330.150  Haemophilus influenzae type b   (D019741)\n",
      "B03.440.450.600.450.665  Haemophilus paragallinarum      (D044134)\n",
      "B03.440.450.600.450.675  Haemophilus parainfluenzae      (D044135)\n",
      "B03.440.450.600.450.690  Haemophilus paraphrophilus      (D044136)\n",
      "B03.440.450.600.450.700  Haemophilus parasuis            (D044137)\n",
      "B03.440.450.600.450.750  Haemophilus somnus              (D044138)\n",
      "B03.440.450.600.500  Mannheimia                      (D031121)\n",
      "B03.440.450.600.500.500  Mannheimia haemolytica          (D016978)\n",
      "B03.440.450.600.600  Pasteurella                     (D010325)\n",
      "B03.440.450.600.600.500  Pasteurella multocida           (D016979)\n",
      "B03.440.450.600.600.650  Pasteurella pneumotropica       (D044142)\n",
      "B03.440.450.650  Rahnella                        (D020638)\n",
      "B03.440.450.690  Shewanella                      (D020592)\n",
      "B03.440.450.690.590  Shewanella putrefaciens         (D020614)\n",
      "B03.440.450.900  Vibrionaceae                    (D014737)\n",
      "B03.440.450.900.050  Aliivibrio                      (D059706)\n",
      "B03.440.450.900.050.860  Aliivibrio fischeri             (D048908)\n",
      "B03.440.450.900.050.930  Aliivibrio salmonicida          (D044165)\n",
      "B03.440.450.900.500  Listonella                      (D053526)\n",
      "B03.440.450.900.604  Photobacterium                  (D010776)\n",
      "B03.440.450.900.859  Vibrio                          (D014733)\n",
      "B03.440.450.900.859.030  Vibrio alginolyticus            (D044149)\n",
      "B03.440.450.900.859.225  Vibrio cholerae                 (D014734)\n",
      "B03.440.450.900.859.225.075  Vibrio cholerae non-O1          (D044163)\n",
      "B03.440.450.900.859.225.151  Vibrio cholerae O1              (D041241)\n",
      "B03.440.450.900.859.225.200  Vibrio cholerae O139            (D041242)\n",
      "B03.440.450.900.859.500  Vibrio mimicus                  (D044164)\n",
      "B03.440.450.900.859.550  Vibrio parahaemolyticus         (D014736)\n",
      "B03.440.450.900.859.900  Vibrio vulnificus               (D041261)\n",
      "B03.440.450.980  Zymomonas                       (D017355)\n",
      "B03.440.475  Gram-Negative Oxygenic Photosynthetic Bacteria  (D019688)\n",
      "B03.440.475.100.655  Prochlorophytes                 (D048928)\n",
      "B03.440.475.100.655.600  Prochlorococcus                 (D048929)\n",
      "B03.440.475.100.655.620  Prochloron                      (D019691)\n",
      "B03.440.475.100.655.640  Prochlorothrix                  (D019692)\n",
      "B03.440.500  Helicobacter                    (D016998)\n",
      "B03.440.500.250  Helicobacter felis              (D044922)\n",
      "B03.440.500.350  Helicobacter heilmannii         (D018247)\n",
      "B03.440.500.362  Helicobacter hepaticus          (D044923)\n",
      "B03.440.500.450  Helicobacter mustelae           (D045502)\n",
      "B03.440.500.550  Helicobacter pylori             (D016480)\n",
      "B03.440.520  Lawsonia Bacteria               (D020639)\n",
      "B03.440.540  Methylosinus                    (D020581)\n",
      "B03.440.540.500  Methylosinus trichosporium      (D020602)\n",
      "B03.440.595  Oceanospirillaceae              (D044130)\n",
      "B03.440.612  Piscirickettsiaceae             (D044147)\n",
      "B03.440.619  Planctomycetes                  (D000090643)\n",
      "B03.440.619.500  Planctomycetales                (D061328)\n",
      "B03.440.623  Rhodobacter                     (D016992)\n",
      "B03.440.623.225  Rhodobacter capsulatus          (D016354)\n",
      "B03.440.623.700  Rhodobacter sphaeroides         (D012242)\n",
      "B03.440.645  Rhodomicrobium                  (D020623)\n",
      "B03.440.646  Rhodovulum                      (D041903)\n",
      "B03.440.664  Rickettsiales                   (D000076944)\n",
      "B03.440.664.750  Anaplasmataceae                 (D000710)\n",
      "B03.440.664.750.050  Anaplasma                       (D000709)\n",
      "B03.440.664.750.050.100  Anaplasma centrale              (D042321)\n",
      "B03.440.664.750.050.500  Anaplasma marginale             (D042322)\n",
      "B03.440.664.750.050.575  Anaplasma ovis                  (D042323)\n",
      "B03.440.664.750.050.600  Anaplasma phagocytophilum       (D041081)\n",
      "B03.440.664.750.287  Ehrlichia                       (D004536)\n",
      "B03.440.664.750.287.090  Ehrlichia canis                 (D041061)\n",
      "B03.440.664.750.287.150  Ehrlichia chaffeensis           (D018469)\n",
      "B03.440.664.750.287.700  Ehrlichia ruminantium           (D016995)\n",
      "B03.440.664.750.525  Neorickettsia                   (D041102)\n",
      "B03.440.664.750.525.700  Neorickettsia risticii          (D041103)\n",
      "B03.440.664.750.525.720  Neorickettsia sennetsu          (D041101)\n",
      "B03.440.680  Roseobacter                     (D041904)\n",
      "B03.440.840  Spirillaceae                    (D043167)\n",
      "B03.440.840.750  Spirillum                       (D013139)\n",
      "B03.440.860  Tenericutes                     (D008973)\n",
      "B03.440.860.074  Acholeplasmataceae              (D000129)\n",
      "B03.440.860.074.150  Acholeplasma                    (D000127)\n",
      "B03.440.860.074.150.500  Acholeplasma laidlawii          (D000128)\n",
      "B03.440.860.074.575  Phytoplasma                     (D045582)\n",
      "B03.440.860.300  Entomoplasmatales               (D060366)\n",
      "B03.440.860.300.250  Entomoplasmataceae              (D045622)\n",
      "B03.440.860.300.750  Spiroplasmataceae               (D016991)\n",
      "B03.440.860.300.750.625  Spiroplasma                     (D013149)\n",
      "B03.440.860.300.750.625.500  Spiroplasma citri               (D045642)\n",
      "B03.440.860.580  Mycoplasmatales                 (D009179)\n",
      "B03.440.860.580.553  Mycoplasmataceae                (D009178)\n",
      "B03.440.860.580.553.553  Mycoplasma                      (D009174)\n",
      "B03.440.860.580.553.553.050  Mycoplasma agalactiae           (D045644)\n",
      "B03.440.860.580.553.553.075  Mycoplasma arthritidis          (D045645)\n",
      "B03.440.860.580.553.553.195  Mycoplasma bovigenitalium       (D045646)\n",
      "B03.440.860.580.553.553.200  Mycoplasma bovis                (D045682)\n",
      "B03.440.860.580.553.553.210  Mycoplasma capricolum           (D045623)\n",
      "B03.440.860.580.553.553.220  Mycoplasma conjunctivae         (D045684)\n",
      "B03.440.860.580.553.553.267  Mycoplasma dispar               (D045685)\n",
      "B03.440.860.580.553.553.315  Mycoplasma fermentans           (D016858)\n",
      "B03.440.860.580.553.553.345  Mycoplasma gallisepticum        (D045686)\n",
      "B03.440.860.580.553.553.355  Mycoplasma genitalium           (D045704)\n",
      "B03.440.860.580.553.553.400  Mycoplasma hominis              (D019535)\n",
      "B03.440.860.580.553.553.420  Mycoplasma hyopneumoniae        (D045705)\n",
      "B03.440.860.580.553.553.425  Mycoplasma hyorhinis            (D045790)\n",
      "B03.440.860.580.553.553.430  Mycoplasma hyosynoviae          (D045791)\n",
      "B03.440.860.580.553.553.450  Mycoplasma iowae                (D045793)\n",
      "B03.440.860.580.553.553.570  Mycoplasma meleagridis          (D045794)\n",
      "B03.440.860.580.553.553.585  Mycoplasma mycoides             (D009176)\n",
      "B03.440.860.580.553.553.630  Mycoplasma orale                (D045795)\n",
      "B03.440.860.580.553.553.635  Mycoplasma ovipneumoniae        (D045802)\n",
      "B03.440.860.580.553.553.645  Mycoplasma penetrans            (D018661)\n",
      "B03.440.860.580.553.553.650  Mycoplasma pneumoniae           (D009177)\n",
      "B03.440.860.580.553.553.670  Mycoplasma pulmonis             (D045803)\n",
      "B03.440.860.580.553.553.720  Mycoplasma salivarium           (D045804)\n",
      "B03.440.860.580.553.553.770  Mycoplasma synoviae             (D045805)\n",
      "B03.440.860.580.553.900  Ureaplasma                      (D014509)\n",
      "B03.440.860.580.553.900.800  Ureaplasma urealyticum          (D016990)\n",
      "B03.440.930  Verrucomicrobia                 (D061327)\n",
      "B03.440.930.500  Akkermansia                     (D000086102)\n",
      "B03.510     Gram-Positive Bacteria          (D006094)\n",
      "B03.510.024  Actinobacteria                  (D039903)\n",
      "B03.510.024.049  Actinomycetales                 (D000192)\n",
      "B03.510.024.049.025  Actinomadura                    (D000082403)\n",
      "B03.510.024.049.050  Actinomycetaceae                (D000191)\n",
      "B03.510.024.049.050.050  Actinomyces                     (D000190)\n",
      "B03.510.024.049.050.050.050  Actinomyces viscosus            (D016924)\n",
      "B03.510.024.049.050.287  Arcanobacterium                 (D055667)\n",
      "B03.510.024.049.050.525  Mobiluncus                      (D017904)\n",
      "B03.510.024.049.075  Amycolatopsis                   (D000082502)\n",
      "B03.510.024.100  Bifidobacterium                 (D001644)\n",
      "B03.510.024.100.125  Bifidobacterium adolescentis    (D000070005)\n",
      "B03.510.024.100.250  Bifidobacterium animalis        (D000069982)\n",
      "B03.510.024.100.313  Bifidobacterium bifidum         (D000069985)\n",
      "B03.510.024.100.375  Bifidobacterium breve           (D000069984)\n",
      "B03.510.024.100.500  Bifidobacterium longum          (D000069978)\n",
      "B03.510.024.100.500.500  Bifidobacterium longum subspecies infantis  (D000070236)\n",
      "B03.510.024.100.750  Bifidobacterium pseudocatenulatum  (D000070257)\n",
      "B03.510.024.175  Brevibacterium                  (D001951)\n",
      "B03.510.024.213  Cellulomonas                    (D040143)\n",
      "B03.510.024.232  Clavibacter                     (D000086162)\n",
      "B03.510.024.250  Corynebacterium                 (D003352)\n",
      "B03.510.024.250.100  Brevibacterium flavum           (D040142)\n",
      "B03.510.024.250.150  Corynebacterium diphtheriae     (D003353)\n",
      "B03.510.024.250.300  Corynebacterium glutamicum      (D048230)\n",
      "B03.510.024.250.600  Corynebacterium pseudotuberculosis  (D016925)\n",
      "B03.510.024.250.625  Corynebacterium pyogenes        (D003355)\n",
      "B03.510.024.288  Dermatophilus                   (D000086182)\n",
      "B03.510.024.325  Frankia                         (D040161)\n",
      "B03.510.024.400  Gardnerella                     (D016975)\n",
      "B03.510.024.400.800  Gardnerella vaginalis           (D006194)\n",
      "B03.510.024.700  Gordonia Bacterium              (D033601)\n",
      "B03.510.024.775  Microbacterium                  (D000083822)\n",
      "B03.510.024.850  Micrococcaceae                  (D008835)\n",
      "B03.510.024.850.050  Arthrobacter                    (D001173)\n",
      "B03.510.024.850.500  Micrococcus                     (D008837)\n",
      "B03.510.024.850.500.500  Micrococcus luteus              (D016982)\n",
      "B03.510.024.850.750  Renibacterium                   (D000084402)\n",
      "B03.510.024.962  Mycobacteriaceae                (D009159)\n",
      "B03.510.024.962.500  Mycobacterium                   (D009161)\n",
      "B03.510.024.962.500.300  Mycobacterium avium             (D009162)\n",
      "B03.510.024.962.500.300.600  Mycobacterium avium subsp. paratuberculosis  (D016927)\n",
      "B03.510.024.962.500.402  Mycobacterium bovis             (D009163)\n",
      "B03.510.024.962.500.480  Mycobacterium haemophilum       (D018393)\n",
      "B03.510.024.962.500.502  Mycobacterium leprae            (D009166)\n",
      "B03.510.024.962.500.602  Mycobacterium lepraemurium      (D009167)\n",
      "B03.510.024.962.500.640  Mycobacterium phlei             (D009168)\n",
      "B03.510.024.962.500.702  Mycobacterium tuberculosis      (D009169)\n",
      "B03.510.024.962.500.720  Nontuberculous Mycobacteria     (D009170)\n",
      "B03.510.024.962.500.720.050  Mycobacterium abscessus         (D000073358)\n",
      "B03.510.024.962.500.720.100  Mycobacterium avium Complex     (D015269)\n",
      "B03.510.024.962.500.720.225  Mycobacterium chelonae          (D016926)\n",
      "B03.510.024.962.500.720.325  Mycobacterium fortuitum         (D019907)\n",
      "B03.510.024.962.500.720.400  Mycobacterium kansasii          (D019909)\n",
      "B03.510.024.962.500.720.500  Mycobacterium marinum           (D019910)\n",
      "B03.510.024.962.500.720.625  Mycobacterium scrofulaceum      (D015359)\n",
      "B03.510.024.962.500.720.662  Mycobacterium smegmatis         (D020102)\n",
      "B03.510.024.962.500.720.700  Mycobacterium ulcerans          (D019911)\n",
      "B03.510.024.962.500.720.950  Mycobacterium xenopi            (D019912)\n",
      "B03.510.024.981  Nocardiaceae                    (D009618)\n",
      "B03.510.024.981.550  Nocardia                        (D009615)\n",
      "B03.510.024.981.550.550  Nocardia asteroides             (D009616)\n",
      "B03.510.024.981.775  Rhodococcus                     (D012240)\n",
      "B03.510.024.981.775.700  Rhodococcus equi                (D016929)\n",
      "B03.510.024.986  Nocardioides                    (D000083962)\n",
      "B03.510.024.988  Nocardiopsis                    (D000086262)\n",
      "B03.510.024.990  Propionibacteriaceae            (D011423)\n",
      "B03.510.024.990.600  Propionibacterium               (D011424)\n",
      "B03.510.024.990.600.600  Propionibacterium acnes         (D011425)\n",
      "B03.510.024.990.600.800  Propionibacterium freudenreichii  (D000070002)\n",
      "B03.510.024.993  Pseudonocardia                  (D000084382)\n",
      "B03.510.024.998  Tropheryma                      (D054851)\n",
      "B03.510.400  Gram-Positive Cocci             (D006095)\n",
      "B03.510.400.200  Deinococcus                     (D034301)\n",
      "B03.510.400.630.600  Planococcus Bacteria            (D058642)\n",
      "B03.510.460  Gram-Positive Rods              (D016987)\n",
      "B03.510.460.400  Gram-Positive Asporogenous Rods  (D006093)\n",
      "B03.510.460.400.400  Gram-Positive Asporogenous Rods, Irregular  (D016988)\n",
      "B03.510.460.400.400.024  Acetobacterium                  (D045855)\n",
      "B03.510.460.400.400.049.957  Thermobifida                    (D000086324)\n",
      "B03.510.460.400.400.049.971  Thermomonospora                 (D000084442)\n",
      "B03.510.460.400.410  Gram-Positive Asporogenous Rods, Regular  (D016989)\n",
      "B03.585     Nitrogen-Fixing Bacteria        (D000072738)\n",
      "B03.585.200  Beijerinckiaceae                (D042062)\n",
      "B03.660     Proteobacteria                  (D020560)\n",
      "B03.660.050  Alphaproteobacteria             (D020561)\n",
      "B03.660.050.340  Holosporaceae                   (D042281)\n",
      "B03.660.050.350  Hyphomicrobiaceae               (D041921)\n",
      "B03.660.050.350.400  Hyphomicrobium                  (D020582)\n",
      "B03.660.050.512  Methylocystaceae                (D042083)\n",
      "B03.660.050.600  Phyllobacteriaceae              (D059127)\n",
      "B03.660.050.600.500  Mesorhizobium                   (D059128)\n",
      "B03.660.050.662.347  Liberibacter                    (D000086202)\n",
      "B03.660.050.750  Rhodobacteraceae                (D041902)\n",
      "B03.660.050.755  Rhodospirillales                (D012245)\n",
      "B03.660.050.783.500.762  Wolbachia                       (D020577)\n",
      "B03.660.050.783.875  Rickettsiaceae                  (D012287)\n",
      "B03.660.050.783.875.650  Rickettsieae                    (D016996)\n",
      "B03.660.050.783.875.650.450  Orientia                        (D000084162)\n",
      "B03.660.050.783.875.650.450.500  Orientia tsutsugamushi          (D012285)\n",
      "B03.660.050.783.875.650.650  Rickettsia                      (D012281)\n",
      "B03.660.050.783.875.650.650.040  Rickettsia akari                (D041141)\n",
      "B03.660.050.783.875.650.650.180  Rickettsia conorii              (D020114)\n",
      "B03.660.050.783.875.650.650.320  Rickettsia felis                (D041161)\n",
      "B03.660.050.783.875.650.650.600  Rickettsia prowazekii           (D012283)\n",
      "B03.660.050.783.875.650.650.650  Rickettsia rickettsii           (D012284)\n",
      "B03.660.050.783.875.650.650.810  Rickettsia typhi                (D012286)\n",
      "B03.660.050.800  Sphingomonadaceae               (D042301)\n",
      "B03.660.075  Betaproteobacteria              (D020562)\n",
      "B03.660.075.090  Burkholderiales                 (D000073556)\n",
      "B03.660.075.090.688.100  Burkholderia                    (D019117)\n",
      "B03.660.075.090.688.100.110  Burkholderia cepacia complex    (D042602)\n",
      "B03.660.075.090.688.100.110.490  Burkholderia cenocepacia        (D057508)\n",
      "B03.660.075.090.688.100.110.500  Burkholderia cepacia            (D016956)\n",
      "B03.660.075.090.688.100.355  Burkholderia gladioli           (D042725)\n",
      "B03.660.075.090.688.100.477  Burkholderia mallei             (D042726)\n",
      "B03.660.075.090.688.100.600  Burkholderia pseudomallei       (D016957)\n",
      "B03.660.075.090.688.600  Ralstonia                       (D043367)\n",
      "B03.660.075.090.688.600.600  Ralstonia pickettii             (D049920)\n",
      "B03.660.075.090.688.600.750  Ralstonia solanacearum          (D043368)\n",
      "B03.660.075.090.941  Xylophilus                      (D000086326)\n",
      "B03.660.075.400  Hydrogenophilaceae              (D043363)\n",
      "B03.660.075.655  Rhodocyclaceae                  (D043362)\n",
      "B03.660.125  Deltaproteobacteria             (D020564)\n",
      "B03.660.125.144  Desulfovibrionales              (D000076642)\n",
      "B03.660.125.500  Myxococcales                    (D009231)\n",
      "B03.660.125.500.605  Myxococcus                      (D016940)\n",
      "B03.660.125.500.605.900  Myxococcus xanthus              (D016941)\n",
      "B03.660.125.500.658  Sorangium                       (D000086323)\n",
      "B03.660.125.500.710  Stigmatella                     (D020597)\n",
      "B03.660.125.500.710.050  Stigmatella aurantiaca          (D020606)\n",
      "B03.660.150  Epsilonproteobacteria           (D020565)\n",
      "B03.660.150.235  Campylobacterales               (D000074707)\n",
      "B03.660.150.235.250  Campylobacteraceae              (D000073557)\n",
      "B03.660.150.235.500  Helicobacteraceae               (D000073569)\n",
      "B03.660.250  Gammaproteobacteria             (D020563)\n",
      "B03.660.250.018  Aeromonadales                   (D000074762)\n",
      "B03.660.250.019  Alcanivoraceae                  (D053521)\n",
      "B03.660.250.021  Alteromonadaceae                (D034241)\n",
      "B03.660.250.021.480  Marinobacter                    (D053520)\n",
      "B03.660.250.080.775  Rappaport                       (D000086322)\n",
      "B03.660.250.110  Chromatiaceae                   (D002841)\n",
      "B03.660.250.139  Dickeya                         (D000083042)\n",
      "B03.660.250.139.500  Dickeya chrysanthemi            (D016974)\n",
      "B03.660.250.540.500  Marinomonas                     (D053544)\n",
      "B03.660.250.565.600  Piscirickettsia                 (D053522)\n",
      "B03.660.250.915.930.049  Xanthomonas axonopodis          (D053532)\n",
      "B03.851     Spirochaetales                  (D013144)\n",
      "B03.900     Sulfur-Reducing Bacteria        (D016999)\n"
     ]
    }
   ],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def parse_mesh_descriptors(xml_path):\n",
    "    \"\"\"\n",
    "    Parse the MeSH XML and return a list of dicts:\n",
    "      { 'ui': DescriptorUI,\n",
    "        'name': DescriptorName,\n",
    "        'tree_numbers': [treeNum1, treeNum2, ...]\n",
    "      }\n",
    "    \"\"\"\n",
    "    tree = ET.parse(xml_path)\n",
    "    root = tree.getroot()\n",
    "    descriptors = []\n",
    "    for dr in root.findall('DescriptorRecord'):\n",
    "        ui_el = dr.find('DescriptorUI')\n",
    "        name_el = dr.find('DescriptorName/String')\n",
    "        if ui_el is None or name_el is None:\n",
    "            continue\n",
    "        ui = ui_el.text\n",
    "        name = name_el.text\n",
    "        tree_numbers = [tn.text for tn in dr.findall('TreeNumberList/TreeNumber') if tn.text]\n",
    "        descriptors.append({\n",
    "            'ui': ui,\n",
    "            'name': name,\n",
    "            'tree_numbers': tree_numbers\n",
    "        })\n",
    "    return descriptors\n",
    "\n",
    "def get_bacteria_taxonomy(xml_path):\n",
    "    \"\"\"\n",
    "    From the full MeSH descriptors file, extract all descriptors\n",
    "    under the 'Bacteria' branch (tree number B03).\n",
    "    Returns a dict: { tree_number: { 'ui': ..., 'name': ... }, ... }\n",
    "    \"\"\"\n",
    "    descriptors = parse_mesh_descriptors(xml_path)\n",
    "\n",
    "    bacteria_prefix = None\n",
    "    for d in descriptors:\n",
    "        if d['name'] == 'Bacteria':\n",
    "            bacteria_prefix = d['tree_numbers'][0]\n",
    "            break\n",
    "\n",
    "    if not bacteria_prefix:\n",
    "        raise RuntimeError(\"Couldn't find a descriptor named 'Bacteria' in the file.\")\n",
    "\n",
    "    taxonomy = {}\n",
    "    for d in descriptors:\n",
    "        for tn in d['tree_numbers']:\n",
    "            if tn == bacteria_prefix or tn.startswith(bacteria_prefix + '.'):\n",
    "                taxonomy[tn] = {\n",
    "                    'ui': d['ui'],\n",
    "                    'name': d['name']\n",
    "                }\n",
    "                break\n",
    "\n",
    "    return taxonomy\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    xml_file = 'desc2025.xml'\n",
    "    bacteria_tax = get_bacteria_taxonomy(xml_file)\n",
    "\n",
    "    for tree_num in sorted(bacteria_tax):\n",
    "        info = bacteria_tax[tree_num]\n",
    "        print(f\"{tree_num:10s}  {info['name']:30s}  ({info['ui']})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8bb2a787-fcdf-419f-bea1-5a1bf60cda60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 30956 descriptors\n"
     ]
    }
   ],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "from difflib import SequenceMatcher, get_close_matches\n",
    "\n",
    "def parse_mesh_descriptors(xml_path):\n",
    "    \"\"\"Parse the MeSH XML and return a list of descriptors.\"\"\"\n",
    "    tree = ET.parse(xml_path)\n",
    "    root = tree.getroot()\n",
    "    descriptors = []\n",
    "    for dr in root.findall('DescriptorRecord'):\n",
    "        ui_el   = dr.find('DescriptorUI')\n",
    "        name_el = dr.find('DescriptorName/String')\n",
    "        if ui_el is None or name_el is None:\n",
    "            continue\n",
    "        ui   = ui_el.text\n",
    "        name = name_el.text\n",
    "        tree_nums = [tn.text for tn in dr.findall('TreeNumberList/TreeNumber') if tn.text]\n",
    "        descriptors.append({'ui':ui, 'name':name, 'tree_numbers':tree_nums})\n",
    "    return descriptors\n",
    "\n",
    "MESH_XML = 'desc2025.xml' #from the folder\n",
    "descriptors = parse_mesh_descriptors(MESH_XML)\n",
    "print(f\"Parsed {len(descriptors)} descriptors\")\n",
    "#print(descriptors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df027a14-5327-4bd0-b6ab-2b2c6ba45cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "from difflib import SequenceMatcher\n",
    "\n",
    "def find_mesh_match(input_name, name_index, n=5, cutoff=0.6):\n",
    "    \"\"\"\n",
    "    Return matches for input_name among the MeSH bacterial names.\n",
    "    Exact matches come first with score=1.0.\n",
    "    Then fuzzy matches (score computed via SequenceMatcher.ratio).\n",
    "    \"\"\"\n",
    "    key = input_name.lower()\n",
    "    results = []\n",
    "    if key in name_index:\n",
    "        for tree_num, ui, name in name_index[key]:\n",
    "            results.append((name, ui, tree_num, 1.0))\n",
    "        return results\n",
    "\n",
    "    all_names = list(name_index.keys())\n",
    "    #print(all_names)\n",
    "    \n",
    "    close = get_close_matches(key, all_names, n=n, cutoff=cutoff)\n",
    "    for cname in close:\n",
    "        score = SequenceMatcher(None, key, cname).ratio()\n",
    "        for tree_num, ui, name in name_index[cname]:\n",
    "            results.append((name, ui, tree_num, score))\n",
    "    results.sort(key=lambda x: x[3], reverse=True)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7d5c6fc4-63fd-4884-9a9d-b703e07594e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 5 entries to dietary_supplements_taxonomy.txt\n"
     ]
    }
   ],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "MESH_XML = 'desc2025.xml'      # your MeSH download\n",
    "OUTPUT_TXT = 'dietary_supplements_taxonomy.txt'\n",
    "\n",
    "def parse_mesh_descriptors(xml_path):\n",
    "    tree = ET.parse(xml_path)\n",
    "    root = tree.getroot()\n",
    "    descriptors = []\n",
    "    for dr in root.findall('DescriptorRecord'):\n",
    "        ui_el   = dr.find('DescriptorUI')\n",
    "        name_el = dr.find('DescriptorName/String')\n",
    "        if ui_el is None or name_el is None:\n",
    "            continue\n",
    "        ui   = ui_el.text\n",
    "        name = name_el.text\n",
    "        tree_nums = [tn.text for tn in dr.findall('TreeNumberList/TreeNumber') if tn.text]\n",
    "        descriptors.append({'ui':ui, 'name':name, 'tree_numbers':tree_nums})\n",
    "    return descriptors\n",
    "\n",
    "def extract_subtree(descriptors, *, root_name=None, root_ui=None):\n",
    "    # 1) find the tree-number prefixes for our root term\n",
    "    prefixes = []\n",
    "    for d in descriptors:\n",
    "        if root_ui and d['ui'] == root_ui:\n",
    "            prefixes = d['tree_numbers']; break\n",
    "        if root_name and d['name'].lower() == root_name.lower():\n",
    "            prefixes = d['tree_numbers']; break\n",
    "    if not prefixes:\n",
    "        raise ValueError(f\"Couldnt find '{root_name or root_ui}' in descriptors\")\n",
    "    # 2) collect everything whose tree-number equals or starts with any prefix\n",
    "    subtree = {}\n",
    "    for d in descriptors:\n",
    "        for tn in d['tree_numbers']:\n",
    "            if any(tn == p or tn.startswith(p + '.') for p in prefixes):\n",
    "                subtree[tn] = {'ui':d['ui'], 'name':d['name']}\n",
    "                break\n",
    "    return subtree\n",
    "\n",
    "def sort_mesh_keys(keys):\n",
    "    # sort by letter then numeric components\n",
    "    def keyfunc(tn):\n",
    "        letter = tn[0]\n",
    "        nums   = list(map(int, tn[1:].split('.')))\n",
    "        return (letter, nums)\n",
    "    return sorted(keys, key=keyfunc)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    descs = parse_mesh_descriptors(MESH_XML)\n",
    "    # you can also pass root_ui='D019587' instead of root_name\n",
    "    tree = extract_subtree(descs, root_name='Dietary Supplements')\n",
    "\n",
    "    # figure out padding\n",
    "    all_tns  = sort_mesh_keys(tree.keys())\n",
    "    max_tn   = max(len(tn) for tn in all_tns)\n",
    "    max_name = max(len(info['name']) for info in tree.values())\n",
    "\n",
    "    with open(OUTPUT_TXT, 'w') as f:\n",
    "        for tn in all_tns:\n",
    "            name = tree[tn]['name']\n",
    "            ui   = tree[tn]['ui']\n",
    "            # two spaces after each column for readability\n",
    "            f.write(f\"{tn.ljust(max_tn+2)}{name.ljust(max_name+2)}({ui})\\n\")\n",
    "\n",
    "    print(f\"Wrote {len(all_tns)} entries to {OUTPUT_TXT}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "52f3363c-0329-413b-841c-9d17a3616e99",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bacteria_taxonomy(descriptors):\n",
    "    \"\"\"\n",
    "    Find the descriptor with name 'Bacteria', \n",
    "    then collect every descriptor under that tree-number prefix.\n",
    "    \"\"\"\n",
    "    prefix = None #bacteria is B03\n",
    "    for d in descriptors:\n",
    "        if d['name'].lower() == 'bacteria' and d['tree_numbers']:\n",
    "            prefix = d['tree_numbers'][0]\n",
    "            break\n",
    "    if not prefix:\n",
    "        raise RuntimeError(\"Could not find 'Bacteria' in descriptors\")\n",
    "    tax = {}\n",
    "    for d in descriptors:\n",
    "        for tn in d['tree_numbers']:\n",
    "            if tn == prefix or tn.startswith(prefix + '.'):\n",
    "                tax[tn] = {'ui':d['ui'], 'name':d['name']}\n",
    "                break\n",
    "    return tax\n",
    "\n",
    "bacteria_tax = get_bacteria_taxonomy(descriptors)\n",
    "#print(f\"{len(bacteria_tax)} bacterial MeSH nodes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "51eaed77-caff-4100-930b-6afb0e22943e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexed 859 bacterial names\n"
     ]
    }
   ],
   "source": [
    "def build_name_index(taxonomy):\n",
    "    \"\"\"\n",
    "    Build a dict: lower-case name -> list of (tree#, ui, canonical name)\n",
    "    \"\"\"\n",
    "    idx = {}\n",
    "    for tree_num, info in taxonomy.items():\n",
    "        key = info['name'].lower()\n",
    "        idx.setdefault(key, []).append((tree_num, info['ui'], info['name']))\n",
    "    return idx\n",
    "\n",
    "name_index = build_name_index(bacteria_tax)\n",
    "print(f\"Indexed {len(name_index)} bacterial names\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5458d974-96ef-426c-903f-edc703ebcf88",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, json, numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "def load_taxonomy_tree(path):\n",
    "    row_re = re.compile(r\"^\\s*(.*?)\\s+\\[([^\\]]+)\\]\\s*$\")\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        for ln in fh:\n",
    "            m = row_re.match(ln)\n",
    "            if m:\n",
    "                label, tid = m.groups()\n",
    "                depth      = len(ln) - len(ln.lstrip())\n",
    "                rows.append((label, tid, depth))\n",
    "    return rows\n",
    "\n",
    "TAX_FILE = \"bacteria_tree1.txt\"\n",
    "rows     = load_taxonomy_tree(TAX_FILE)\n",
    "\n",
    "exact_ix  = defaultdict(list)\n",
    "for lbl, tid, depth in rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, tid, depth))\n",
    "\n",
    "labels_only  = [r[0] for r in rows]\n",
    "vec          = TfidfVectorizer(stop_words=\"english\")\n",
    "mat          = vec.fit_transform(labels_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v   = vec.transform([term])\n",
    "    sc  = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr: break\n",
    "        lbl, tid, d = rows[i]\n",
    "        out.append((lbl, tid, d, sc[i]))\n",
    "    return out\n",
    "\n",
    "abbr_re = re.compile(r\"^([A-Z])\\.\\s+([A-Za-z_-]+)$\")\n",
    "def preprocess(term):\n",
    "    return term.replace('_', ' ').strip()\n",
    "\n",
    "def genus_abbrev_lookup(term):\n",
    "    m = abbr_re.match(term)\n",
    "    if not m:\n",
    "        return []\n",
    "    initial, species = m.groups()\n",
    "    species = species.lower()\n",
    "    hits = []\n",
    "    for lbl, tid, d in rows:\n",
    "        if lbl.lower().endswith(' ' + species) and lbl[0].upper() == initial:\n",
    "            hits.append((lbl, tid, d, 1.00))\n",
    "    return hits\n",
    "\n",
    "def strip_html(text):\n",
    "    return re.sub(r\"<[^>]+>\", \"\", text)\n",
    "\n",
    "with open(r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\data\\train_platinum\\train_platinum.json\", encoding=\"utf-8\") as fh:\n",
    "    js = json.load(fh)\n",
    "\n",
    "queries = { strip_html(e[\"text_span\"]).strip()\n",
    "            for rec in js.values()\n",
    "            for e in rec.get(\"entities\", [])\n",
    "            if e.get(\"label\",\"\").lower() == \"bacteria\" }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a25cf28d-cd18-4612-9ee6-dd4d5db1e838",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: Veillonella\n",
      "   Veillonella                              ID=NCBITaxon_29465 depth=12 (exact)\n",
      "   Veillonella                              ID=NCBITaxon_29465 depth=12 score=1.00\n",
      "\n",
      "Query: Roseburia\n",
      "   Roseburia                                ID=NCBITaxon_841   depth=12 (exact)\n",
      "   Roseburia                                ID=NCBITaxon_841   depth=12 score=1.00\n",
      "\n",
      "Query: Christensenellaceae R-7 group\n",
      "   Christensenellaceae                      ID=NCBITaxon_990719 depth=10 score=0.87\n",
      "\n",
      "Query: Subdoligranulum\n",
      "   Subdoligranulum                          ID=NCBITaxon_292632 depth=12 (exact)\n",
      "   Subdoligranulum                          ID=NCBITaxon_292632 depth=12 score=1.00\n",
      "\n",
      "Query: Oscillibacter\n",
      "   Oscillibacter                            ID=NCBITaxon_459786 depth=12 (exact)\n",
      "   Oscillibacter                            ID=NCBITaxon_459786 depth=12 score=1.00\n",
      "\n",
      "Query: UCG-005\n",
      "   Ruminococcaceae bacterium UCG-005        ID=NCBITaxon_3068309 depth=14 score=0.87\n",
      "\n",
      "Query: acetate-producing bacteria\n",
      "   Actinobacteria                 UI=D039903  Tree=B03.510.024   scoreMESH=0.65\n",
      "Query: acetate-producing bacteria\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: bacteria\n",
      "   Bacteria                                 ID=NCBITaxon_2     depth=0  (exact)\n",
      "   Bacteria                                 ID=NCBITaxon_2     depth=0  score=1.00\n",
      "\n",
      "Query: oral bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_2\n",
      "\n",
      "Query: Lactobacillus plantarum CCFM8661\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.76\n",
      "\n",
      "Query: live and inactivated Lactobacillus plantarum CCFM8661\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.76\n",
      "\n",
      "Query: live L  plantarum CCFM8661\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.89\n",
      "\n",
      "Query: live L  plantarum CCFM8661\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: inactivated L  plantarum CCFM8661\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.89\n",
      "\n",
      "Query: live and inactivated L  plantarum CCFM8661\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.89\n",
      "\n",
      "Query: Staphylococcus\n",
      "   Staphylococcus                           ID=NCBITaxon_1279  depth=12 (exact)\n",
      "   Staphylococcus                           ID=NCBITaxon_1279  depth=12 score=1.00\n",
      "\n",
      "Query: L  plantarum CCFM8661\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.89\n",
      "\n",
      "Query: Lactobacillaceae\n",
      "   Lactobacillaceae                         ID=NCBITaxon_33958 depth=10 (exact)\n",
      "   Lactobacillaceae                         ID=NCBITaxon_33958 depth=10 score=1.00\n",
      "\n",
      "Query: Lachnospiraceae\n",
      "   Lachnospiraceae                          ID=NCBITaxon_186803 depth=10 (exact)\n",
      "   Lachnospiraceae                          ID=NCBITaxon_186803 depth=10 score=1.00\n",
      "\n",
      "Query: Lachnospiraceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_186803\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Ruminococcaceae bacterium enrichment culture ID=NCBITaxon_1776336 depth=14 score=0.80\n",
      "\n",
      "Query: Levilactobacillus brevis\n",
      "   Levilactobacillus brevis                 ID=NCBITaxon_1580  depth=14 (exact)\n",
      "   Levilactobacillus brevis                 ID=NCBITaxon_1580  depth=14 score=1.00\n",
      "\n",
      "Query: Lactiplantibacillus plantarum\n",
      "   Lactiplantibacillus plantarum            ID=NCBITaxon_1590  depth=14 (exact)\n",
      "   Lactiplantibacillus plantarum            ID=NCBITaxon_1590  depth=14 score=1.00\n",
      "\n",
      "Query: Lacticaseibacillus paracasei\n",
      "   Lacticaseibacillus paracasei             ID=NCBITaxon_1597  depth=14 (exact)\n",
      "   Lacticaseibacillus paracasei             ID=NCBITaxon_1597  depth=14 score=1.00\n",
      "\n",
      "Query: Ligilactobacillus salivarius\n",
      "   Ligilactobacillus salivarius             ID=NCBITaxon_1624  depth=14 (exact)\n",
      "   Ligilactobacillus salivarius             ID=NCBITaxon_1624  depth=14 score=1.00\n",
      "\n",
      "Query: Streptococcus thermophilus\n",
      "   Streptococcus thermophilus               ID=NCBITaxon_1308  depth=14 (exact)\n",
      "   Streptococcus thermophilus               ID=NCBITaxon_1308  depth=14 score=1.00\n",
      "\n",
      "Query: L  brevis P30021\n",
      "   Paenibacillus brevis                     ID=NCBITaxon_2841508 depth=14 score=0.88\n",
      "\n",
      "Query: L  plantarum P30025\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.89\n",
      "\n",
      "Query: Veillonellaceae\n",
      "   Veillonellaceae                          ID=NCBITaxon_31977 depth=10 (exact)\n",
      "   Veillonellaceae                          ID=NCBITaxon_31977 depth=10 score=1.00\n",
      "\n",
      "Query: Bacteroides\n",
      "   Bacteroides                              ID=NCBITaxon_816   depth=16 (exact)\n",
      "   Bacteroides                              ID=NCBITaxon_816   depth=16 score=1.00\n",
      "\n",
      "Query: Proteobacteria\n",
      "   Proteobacteria                 UI=D020560  Tree=B03.660       scoreMESH=1.00\n",
      "Query: Succinivibrio\n",
      "   Succinivibrio                            ID=NCBITaxon_83770 depth=12 (exact)\n",
      "   Succinivibrio                            ID=NCBITaxon_83770 depth=12 score=1.00\n",
      "\n",
      "Query: Megasphaera\n",
      "   Megasphaera                              ID=NCBITaxon_906   depth=12 (exact)\n",
      "   Megasphaera                              ID=NCBITaxon_906   depth=12 score=1.00\n",
      "\n",
      "Query: Collinsella\n",
      "   Collinsella                              ID=NCBITaxon_102106 depth=12 (exact)\n",
      "   Collinsella                              ID=NCBITaxon_102106 depth=12 score=1.00\n",
      "\n",
      "Query: Clostridium\n",
      "   Clostridium                              ID=NCBITaxon_1485  depth=12 (exact)\n",
      "   Clostridium                              ID=NCBITaxon_1485  depth=12 score=1.00\n",
      "\n",
      "Query: Klebsiella\n",
      "   Klebsiella                               ID=NCBITaxon_570   depth=14 (exact)\n",
      "   Klebsiella                               ID=NCBITaxon_570   depth=14 score=1.00\n",
      "\n",
      "Query: Methanobrevibacter\n",
      "   Brevibacterium                 UI=D001951  Tree=B03.510.024.175  scoreMESH=0.69\n",
      "Query: Blautia\n",
      "   Blautia                                  ID=NCBITaxon_572511 depth=12 (exact)\n",
      "   Blautia                                  ID=NCBITaxon_572511 depth=12 score=1.00\n",
      "\n",
      "Query: Coprococcus\n",
      "   Coprococcus                              ID=NCBITaxon_33042 depth=12 (exact)\n",
      "   Coprococcus                              ID=NCBITaxon_33042 depth=12 score=1.00\n",
      "\n",
      "Query: Roseburia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_841\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Lactobacillus                            ID=NCBITaxon_1578  depth=12 (exact)\n",
      "   Lactobacillus                            ID=NCBITaxon_1578  depth=12 score=1.00\n",
      "\n",
      "Query: L\n",
      "Query: plantarum P-8\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.89\n",
      "\n",
      "Query: Bifidobacterium adolescentis\n",
      "   Bifidobacterium adolescentis             ID=NCBITaxon_1680  depth=14 (exact)\n",
      "   Bifidobacterium adolescentis             ID=NCBITaxon_1680  depth=14 score=1.00\n",
      "\n",
      "Query: Bifidobacterium longum\n",
      "   Bifidobacterium longum                   ID=NCBITaxon_216816 depth=14 (exact)\n",
      "   Bifidobacterium longum                   ID=NCBITaxon_216816 depth=14 score=1.00\n",
      "\n",
      "Query: Fecalibacterium prausnitzii\n",
      "   Faecalibacterium prausnitzii             ID=NCBITaxon_853   depth=14 score=0.79\n",
      "\n",
      "Query: Roseburia faecis\n",
      "   Roseburia faecis                         ID=NCBITaxon_301302 depth=14 (exact)\n",
      "   Roseburia faecis                         ID=NCBITaxon_301302 depth=14 score=1.00\n",
      "\n",
      "Query: Fusicatenibacter saccharivorans\n",
      "   Fusicatenibacter saccharivorans          ID=NCBITaxon_1150298 depth=14 (exact)\n",
      "   Fusicatenibacter saccharivorans          ID=NCBITaxon_1150298 depth=14 score=1.00\n",
      "\n",
      "Query: Bacteroidaceae and Porphyromonadaceae families\n",
      "   Bacteroidaceae                           ID=NCBITaxon_815   depth=14 score=0.76\n",
      "\n",
      "Query: Rikencellaceae\n",
      "   Brucellaceae                   UI=D042021  Tree=B03.440.400.425.215  scoreMESH=0.77\n",
      "Query: S24-7\n",
      "   bacterium S24                            ID=NCBITaxon_1315627 depth=4  score=0.97\n",
      "\n",
      "Query: bacteria\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_2\n",
      "\n",
      "Query: Lactococcus\n",
      "   Lactococcus                              ID=NCBITaxon_1357  depth=12 (exact)\n",
      "   Lactococcus                              ID=NCBITaxon_1357  depth=12 score=1.00\n",
      "\n",
      "Query: Biffdobacterium\n",
      "   Bifidobacterium                UI=D001644  Tree=B03.510.024.100  scoreMESH=0.93\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Biffdobacterium\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D001644\n",
      "\n",
      "Query: Lactobacillus helveticus R0052\n",
      "   Lactobacillus helveticus R0052           ID=NCBITaxon_880633 depth=16 (exact)\n",
      "   Lactobacillus helveticus R0052           ID=NCBITaxon_880633 depth=16 score=1.00\n",
      "\n",
      "Query: Lactobacillus plantarum R1012\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.76\n",
      "\n",
      "Query: Bifidobacterium longum R0175\n",
      "   Bifidobacterium longum                   ID=NCBITaxon_216816 depth=14 score=1.00\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Lactobacillus species\n",
      "   Actinobacillus suis            UI=D044132  Tree=B03.440.450.009.700  scoreMESH=0.80\n",
      "Query: Gut Bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Escherichia coli\n",
      "   Escherichia coli                         ID=NCBITaxon_562   depth=14 (exact)\n",
      "   Escherichia coli W                       ID=NCBITaxon_566546 depth=16 score=1.00\n",
      "\n",
      "Query: Enterobacter cloacae\n",
      "   Enterobacter cloacae                     ID=NCBITaxon_550   depth=16 (exact)\n",
      "   Enterobacter cloacae                     ID=NCBITaxon_550   depth=16 score=1.00\n",
      "\n",
      "Query: maternal vaginal microbes\n",
      "   Gardnerella vaginalis          UI=D006194  Tree=B03.510.024.400.800  scoreMESH=0.61\n",
      "Query: Bifidobacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D001644\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Bifidobacterium                          ID=NCBITaxon_1678  depth=12 (exact)\n",
      "   Bifidobacterium                          ID=NCBITaxon_1678  depth=12 score=1.00\n",
      "\n",
      "Query: Verrucomicrobiaceae\n",
      "   Verrucomicrobiaceae                      ID=NCBITaxon_203557 depth=12 (exact)\n",
      "   Verrucomicrobiaceae                      ID=NCBITaxon_203557 depth=12 score=1.00\n",
      "\n",
      "Query: Akkermansia\n",
      "   Akkermansia                              ID=NCBITaxon_239934 depth=14 (exact)\n",
      "   Akkermansia                              ID=NCBITaxon_239934 depth=14 score=1.00\n",
      "\n",
      "Query: Faecalibacterium\n",
      "   Faecalibacterium                         ID=NCBITaxon_216851 depth=12 (exact)\n",
      "   Faecalibacterium                         ID=NCBITaxon_216851 depth=12 score=1.00\n",
      "\n",
      "Query: Coprococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33042\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Prevotella\n",
      "   Prevotella                               ID=NCBITaxon_838   depth=16 (exact)\n",
      "   Prevotella                               ID=NCBITaxon_838   depth=16 score=1.00\n",
      "\n",
      "Query: Prevotellaceae\n",
      "   Prevotellaceae                           ID=NCBITaxon_171552 depth=14 (exact)\n",
      "   Prevotellaceae                           ID=NCBITaxon_171552 depth=14 score=1.00\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Bacteroidetes bacterium 5-4              ID=NCBITaxon_226197 depth=12 score=0.93\n",
      "\n",
      "Query: Proteobacteria\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D020560\n",
      "\n",
      "Query: Helicobacter pylori\n",
      "   Helicobacter pylori                      ID=NCBITaxon_210   depth=14 (exact)\n",
      "   Helicobacter pylori                      ID=NCBITaxon_210   depth=14 score=1.00\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Actinobacteria\n",
      "   Actinobacteria bacterium HGW-Actinobacteria-9 ID=NCBITaxon_2013654 depth=8  score=0.87\n",
      "\n",
      "Query: Firmicutes\n",
      "   Bacillus <firmicutes>                    ID=NCBITaxon_1386  depth=12 score=0.87\n",
      "\n",
      "Query: microbes\n",
      "   Microbacterium                 UI=D000083822 Tree=B03.510.024.775  scoreMESH=0.64\n",
      "Query: firmicutes\n",
      "   Bacillus <firmicutes>                    ID=NCBITaxon_1386  depth=12 score=0.87\n",
      "\n",
      "Query: coliform\n",
      "   coliform bacterium S1/7                  ID=NCBITaxon_464824 depth=4  score=0.78\n",
      "\n",
      "Query: bacteroides\n",
      "   Bacteroides                              ID=NCBITaxon_816   depth=16 (exact)\n",
      "   Bacteroides                              ID=NCBITaxon_816   depth=16 score=1.00\n",
      "\n",
      "Query: firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: gut bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Toxoplasma gondii\n",
      "   Mycoplasma genitalium          UI=D045704  Tree=B03.440.860.580.553.553.355  scoreMESH=0.63\n",
      "Query: Lachnospiraceae family\n",
      "   Leptospiraceae                 UI=D016953  Tree=B03.440.400.425.475  scoreMESH=0.61\n",
      "Query: bifidobacterial populations\n",
      "   Bifidobacterium pseudocatenulatum UI=D000070257 Tree=B03.510.024.100.750  scoreMESH=0.67\n",
      "Query: gut Bifidobacterium\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Akkermansias\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Desulfovibrio\n",
      "   Desulfovibrio                            ID=NCBITaxon_872   depth=12 (exact)\n",
      "   Desulfovibrio                            ID=NCBITaxon_872   depth=12 score=1.00\n",
      "\n",
      "Query: Klebsiella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_570\n",
      "\n",
      "Query: Akkermansia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Lactobacillus plantarum\n",
      "   Acinetobacter plantarum                  ID=NCBITaxon_1591090 depth=14 score=0.76\n",
      "\n",
      "Query: GABA-producing bacteria\n",
      "   Sulfur-Reducing Bacteria       UI=D016999  Tree=B03.900       scoreMESH=0.72\n",
      "Query: Lactobacillus plantarum\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Odoribacter\n",
      "   Odoribacter                              ID=NCBITaxon_283168 depth=16 (exact)\n",
      "   Odoribacter                              ID=NCBITaxon_283168 depth=16 score=1.00\n",
      "\n",
      "Query: Oscillibacter\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_459786\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Bifidobacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Clostridium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1485\n",
      "\n",
      "Query: Lachnospiraceae NK4A136 group\n",
      "   Lachnospiraceae bacterium NK4A136        ID=NCBITaxon_877420 depth=14 score=0.90\n",
      "\n",
      "Query: Subdoligranulum\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_292632\n",
      "\n",
      "Query: Faecalibacterium prausnitzii\n",
      "   Faecalibacterium prausnitzii             ID=NCBITaxon_853   depth=14 (exact)\n",
      "   Faecalibacterium prausnitzii             ID=NCBITaxon_853   depth=14 score=1.00\n",
      "\n",
      "Query: Corynebacterium\n",
      "   Corynebacterium                          ID=NCBITaxon_1716  depth=12 (exact)\n",
      "   Corynebacterium                          ID=NCBITaxon_1716  depth=12 score=1.00\n",
      "\n",
      "Query: Ruminococcaceae Incertae Sedis\n",
      "   uncultured incertae sedis 5 bacterium    ID=NCBITaxon_412050 depth=4  score=0.77\n",
      "\n",
      "Query: Fusobacterium\n",
      "   Fusobacterium                            ID=NCBITaxon_848   depth=12 (exact)\n",
      "   Fusobacterium                            ID=NCBITaxon_848   depth=12 score=1.00\n",
      "\n",
      "Query: Corynebacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1716\n",
      "\n",
      "Query: Ruminococcaceae Incertae Sedis\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_412050\n",
      "\n",
      "Query: microorganisms\n",
      "Query: intestinal bacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: Verrucomicrobia\n",
      "   Verrucomicrobia bacterium B-1-8          ID=NCBITaxon_747582 depth=10 score=0.94\n",
      "\n",
      "Query: Mucispirillum\n",
      "   Mucispirillum                            ID=NCBITaxon_248038 depth=12 (exact)\n",
      "   Mucispirillum                            ID=NCBITaxon_248038 depth=12 score=1.00\n",
      "\n",
      "Query: Porphyromonas\n",
      "   Porphyromonas                            ID=NCBITaxon_836   depth=16 (exact)\n",
      "   Porphyromonas                            ID=NCBITaxon_836   depth=16 score=1.00\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Parabacteroides\n",
      "   Parabacteroides                          ID=NCBITaxon_375288 depth=16 (exact)\n",
      "   Parabacteroides                          ID=NCBITaxon_375288 depth=16 score=1.00\n",
      "\n",
      "Query: Prevotella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Verrucomicrobia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_747582\n",
      "\n",
      "Query: Lactobacillus fermentum PS150\n",
      "   Limosilactobacillus fermentum  UI=D052197  Tree=B03.353.750.450.672  scoreMESH=0.79\n",
      "Query: hypnotic psychobiotic strain\n",
      "Query: Psychobiotics\n",
      "   Synechocystis                  UI=D046939  Tree=B03.280.750   scoreMESH=0.62\n",
      "Query: probiotics\n",
      "Query: psychobiotic strain\n",
      "Query: Lactobacillus fermentum PS150\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D052197\n",
      "\n",
      "Query: PS150\n",
      "Query: L  fermentum strain\n",
      "Query: GR1009\n",
      "Query: PS150\n",
      "Query: GR1009\n",
      "Query: PS150\n",
      "Query: PS150\n",
      "Query: GR1009\n",
      "Query: heat-killed PS150\n",
      "Query: GR1009\n",
      "Query: PS150\n",
      "Query: PS150\n",
      "Query: GR1009\n",
      "Query: PS150\n",
      "Query: GR1009\n",
      "Query: GR1009\n",
      "Query: PS150\n",
      "Query: PS150\n",
      "Query: PS150\n",
      "Query: GR1009\n",
      "Query: EC Ruminococcus torques\n",
      "   [Ruminococcus] torques                   ID=NCBITaxon_33039 depth=14 score=0.83\n",
      "\n",
      "Query: Coprococcus comes\n",
      "   Coprococcus comes                        ID=NCBITaxon_410072 depth=14 (exact)\n",
      "   Coprococcus comes                        ID=NCBITaxon_410072 depth=14 score=1.00\n",
      "\n",
      "Query: Megasphaera elsdenii\n",
      "   Megasphaera elsdenii                     ID=NCBITaxon_907   depth=14 (exact)\n",
      "   Megasphaera elsdenii                     ID=NCBITaxon_907   depth=14 score=1.00\n",
      "\n",
      "Query: Acidaminococcus intestini\n",
      "   Acidaminococcus intestini                ID=NCBITaxon_187327 depth=14 (exact)\n",
      "   Acidaminococcus intestini                ID=NCBITaxon_187327 depth=14 score=1.00\n",
      "\n",
      "Query: Barnesiella viscericola\n",
      "   Barnesiella viscericola                  ID=NCBITaxon_397865 depth=18 (exact)\n",
      "   Barnesiella viscericola                  ID=NCBITaxon_397865 depth=18 score=1.00\n",
      "\n",
      "Query: Bacteroides eggerthii\n",
      "   Bacteroides eggerthii                    ID=NCBITaxon_28111 depth=18 (exact)\n",
      "   Bacteroides eggerthii                    ID=NCBITaxon_28111 depth=18 score=1.00\n",
      "\n",
      "Query: Gut Bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Gut bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: mucus-associated species\n",
      "Query: Desulfovibrionaceae\n",
      "   Desulfovibrionaceae                      ID=NCBITaxon_194924 depth=10 (exact)\n",
      "   Desulfovibrionaceae                      ID=NCBITaxon_194924 depth=10 score=1.00\n",
      "\n",
      "Query: Enterobacteriaceae\n",
      "   Enterobacteriaceae                       ID=NCBITaxon_543   depth=10 (exact)\n",
      "   Enterobacteriaceae                       ID=NCBITaxon_543   depth=10 score=1.00\n",
      "\n",
      "Query: Turicibacter\n",
      "   Turicibacter                             ID=NCBITaxon_191303 depth=12 (exact)\n",
      "   Turicibacter                             ID=NCBITaxon_191303 depth=12 score=1.00\n",
      "\n",
      "Query: Coprococcus 1\n",
      "   Coprococcus                              ID=NCBITaxon_33042 depth=12 score=1.00\n",
      "\n",
      "Query: Lachnospiraceae FCS020 group\n",
      "   bacterium FCS020                         ID=NCBITaxon_1451755 depth=4  score=0.78\n",
      "\n",
      "Query: Eggerthellaceae\n",
      "   Eggerthellaceae                          ID=NCBITaxon_1643826 depth=10 (exact)\n",
      "   Eggerthellaceae                          ID=NCBITaxon_1643826 depth=10 score=1.00\n",
      "\n",
      "Query: Turicibacter\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_191303\n",
      "\n",
      "Query: Coprococcus 1\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33042\n",
      "\n",
      "Query: Eggerthellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1643826\n",
      "\n",
      "Query: Clostridium sp  BR31\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1485\n",
      "\n",
      "Query: Turicibacter\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_191303\n",
      "\n",
      "Query: Coprococcus 1\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33042\n",
      "\n",
      "Query: Eggerthellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1643826\n",
      "\n",
      "Query: Clostridium sp  BR31\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1485\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Actinobacteria\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_2013654\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Lactobacillus reuteri NK33\n",
      "   Limosilactobacillus reuteri    UI=D052200  Tree=B03.353.750.450.705  scoreMESH=0.79\n",
      "Query: Bifidobacterium adolescentis NK98\n",
      "   Bifidobacterium adolescentis             ID=NCBITaxon_1680  depth=14 score=1.00\n",
      "\n",
      "Query: Bifidobacteriaceae\n",
      "   Bifidobacteriaceae                       ID=NCBITaxon_31953 depth=10 (exact)\n",
      "   Bifidobacteriaceae                       ID=NCBITaxon_31953 depth=10 score=1.00\n",
      "\n",
      "Query: Lactobacillacea\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_33958\n",
      "\n",
      "Query: Enterobacteriaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_543\n",
      "\n",
      "Query: Lachnospiraceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_186803\n",
      "\n",
      "Query: Odoribacter\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_283168\n",
      "\n",
      "Query: cancer-related bacteria\n",
      "   Gram-Negative Bacteria         UI=D006090  Tree=B03.440       scoreMESH=0.67\n",
      "Query: Helicobacter\n",
      "   Helicobacter                             ID=NCBITaxon_209   depth=12 (exact)\n",
      "   Helicobacter                             ID=NCBITaxon_209   depth=12 score=1.00\n",
      "\n",
      "Query: Lactococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1357\n",
      "\n",
      "Query: Alloprevotella\n",
      "   Alloprevotella                           ID=NCBITaxon_1283313 depth=16 (exact)\n",
      "   Alloprevotella                           ID=NCBITaxon_1283313 depth=16 score=1.00\n",
      "\n",
      "Query: imbalance-inducing bacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D016999\n",
      "\n",
      "Query: Escherichia-Shigella\n",
      "   Escherichia/Shigella sp.                 ID=NCBITaxon_1940338 depth=14 score=0.99\n",
      "\n",
      "Query: Parabacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_375288\n",
      "\n",
      "Query: butyrate-producing bacteria\n",
      "   butyrate-producing bacterium             ID=NCBITaxon_2811711 depth=4  score=0.89\n",
      "\n",
      "Query: Ruminococcaceae UCG-014\n",
      "   Micrococcaceae                 UI=D008835  Tree=B03.510.024.850  scoreMESH=0.65\n",
      "Query: Lactobacillus reuteri ATG-F4\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D052200\n",
      "\n",
      "Query: Lactobacillus reuteri ATG-F4\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D052200\n",
      "\n",
      "Query: F4\n",
      "   Bacillus sp. F4-1                        ID=NCBITaxon_489169 depth=16 score=0.92\n",
      "\n",
      "Query: F4\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_489169\n",
      "\n",
      "Query: F4\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_489169\n",
      "\n",
      "Query: F4\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_489169\n",
      "\n",
      "Query: short-chain fatty acid producing bacteria\n",
      "Query: gut-resident bacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D016999\n",
      "\n",
      "Query: dysregulated bacteria\n",
      "   Bacteria                                 ID=NCBITaxon_2     depth=0  score=1.00\n",
      "\n",
      "Query: Fusobacterium nucleatum\n",
      "   Fusobacterium nucleatum                  ID=NCBITaxon_851   depth=14 (exact)\n",
      "   Fusobacterium nucleatum                  ID=NCBITaxon_851   depth=14 score=1.00\n",
      "\n",
      "Query: bifidobacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Actinobacteria\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_2013654\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Lachnospiraceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_186803\n",
      "\n",
      "Query: Faecalibacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_216851\n",
      "\n",
      "Query: Intestinal bacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: short-chain fatty acid  SCFA -producing bacteria\n",
      "Query: Lachnospira\n",
      "   Lachnospira                              ID=NCBITaxon_28050 depth=12 (exact)\n",
      "   Lachnospira                              ID=NCBITaxon_28050 depth=12 score=1.00\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Faecalibacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_216851\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Acute chronic insomnia-related signature bacteria\n",
      "Query: Negativicutes\n",
      "   Negativicutes                            ID=NCBITaxon_909932 depth=6  (exact)\n",
      "   Negativicutes                            ID=NCBITaxon_909932 depth=6  score=1.00\n",
      "\n",
      "Query: Clostridiuminnocuumgroup\n",
      "   Clostridium septicum           UI=D052557  Tree=B03.300.390.400.200.690  scoreMESH=0.68\n",
      "Query: Dorea\n",
      "   Dorea                                    ID=NCBITaxon_189330 depth=12 (exact)\n",
      "   Dorea                                    ID=NCBITaxon_189330 depth=12 score=1.00\n",
      "\n",
      "Query: Lachnoclostridium\n",
      "   Lachnoclostridium                        ID=NCBITaxon_1506553 depth=12 (exact)\n",
      "   Lachnoclostridium                        ID=NCBITaxon_1506553 depth=12 score=1.00\n",
      "\n",
      "Query: Prevotella7\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: Selenomonadalesare\n",
      "   Selenomonas                    UI=D020357  Tree=B03.440.425.410.800.877  scoreMESH=0.76\n",
      "Query: gut bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Prevotella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Megamonas\n",
      "   Megamonas                                ID=NCBITaxon_158846 depth=12 (exact)\n",
      "   Megamonas                                ID=NCBITaxon_158846 depth=12 score=1.00\n",
      "\n",
      "Query: Faecalibacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_216851\n",
      "\n",
      "Query: Synergistetes\n",
      "   Synergistetes bacterium HGW-Synergistetes-1 ID=NCBITaxon_2013842 depth=8  score=0.91\n",
      "\n",
      "Query: Proteobacteria\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D020560\n",
      "\n",
      "Query: olidextribacter\n",
      "   Liberibacter                   UI=D000086202 Tree=B03.660.050.662.347  scoreMESH=0.81\n",
      "Query: Intestinibacter\n",
      "   Intestinibacter                          ID=NCBITaxon_1505657 depth=12 (exact)\n",
      "   Intestinibacter                          ID=NCBITaxon_1505657 depth=12 score=1.00\n",
      "\n",
      "Query: Kineothrix\n",
      "   Kineothrix                               ID=NCBITaxon_2163168 depth=12 (exact)\n",
      "   Kineothrix                               ID=NCBITaxon_2163168 depth=12 score=1.00\n",
      "\n",
      "Query: Agathobaculum\n",
      "   Agathobaculum                            ID=NCBITaxon_2048137 depth=12 (exact)\n",
      "   Agathobaculum                            ID=NCBITaxon_2048137 depth=12 score=1.00\n",
      "\n",
      "Query: Roseburia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_841\n",
      "\n",
      "Query: Prevotella species\n",
      "   Prevotella nigrescens          UI=D045242  Tree=B03.440.080.094.640.550  scoreMESH=0.72\n",
      "Query: Actinobacteriota\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Verrucomicrobiota\n",
      "   Verrucomicrobiota                        ID=NCBITaxon_74201 depth=6  (exact)\n",
      "   Verrucomicrobiota                        ID=NCBITaxon_74201 depth=6  score=1.00\n",
      "\n",
      "Query: Akkermansia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Enterococcus\n",
      "   Enterococcus                             ID=NCBITaxon_1350  depth=12 (exact)\n",
      "   Enterococcus                             ID=NCBITaxon_1350  depth=12 score=1.00\n",
      "\n",
      "Query: Hungatella\n",
      "   Hungatella                               ID=NCBITaxon_1649459 depth=12 (exact)\n",
      "   Hungatella                               ID=NCBITaxon_1649459 depth=12 score=1.00\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Oscillospira\n",
      "   Oscillospira                             ID=NCBITaxon_119852 depth=12 (exact)\n",
      "   Oscillospira                             ID=NCBITaxon_119852 depth=12 score=1.00\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1776336\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Proteobacteria\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D020560\n",
      "\n",
      "Query: Lachnospiraceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_186803\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Coprococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33042\n",
      "\n",
      "Query: Roseburia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_841\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Verrucomicrobiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_74201\n",
      "\n",
      "Query: Akkermansia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Akkermansia muciniphila\n",
      "   Akkermansia muciniphila                  ID=NCBITaxon_239935 depth=16 (exact)\n",
      "   Akkermansia muciniphila                  ID=NCBITaxon_239935 depth=16 score=1.00\n",
      "\n",
      "Query: Verrucomicrobiaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_203557\n",
      "\n",
      "Query: Verrucomicrobiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_74201\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Roseburia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_841\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Lactobacillaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33958\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Barnesiellaceae\n",
      "   Barnesiellaceae                          ID=NCBITaxon_2005519 depth=14 (exact)\n",
      "   Barnesiellaceae                          ID=NCBITaxon_2005519 depth=14 score=1.00\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Actinobacteriota\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: Bilophila wadsworthia\n",
      "   Bilophila wadsworthia                    ID=NCBITaxon_35833 depth=14 (exact)\n",
      "   Bilophila wadsworthia                    ID=NCBITaxon_35833 depth=14 score=1.00\n",
      "\n",
      "Query: Thermodesulfobacteriota\n",
      "   Thermodesulfobacteriota                  ID=NCBITaxon_200940 depth=4  (exact)\n",
      "   Thermodesulfobacteriota                  ID=NCBITaxon_200940 depth=4  score=1.00\n",
      "\n",
      "Query: Prevotella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Verrucomicrobiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_74201\n",
      "\n",
      "Query: Akkermansia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Bifidobacteriaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_31953\n",
      "\n",
      "Query: Coriobacteriaceae\n",
      "   Coriobacteriaceae                        ID=NCBITaxon_84107 depth=10 (exact)\n",
      "   Coriobacteriaceae                        ID=NCBITaxon_84107 depth=10 score=1.00\n",
      "\n",
      "Query: Actinobacteriota\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Christensenellaceae                      ID=NCBITaxon_990719 depth=10 (exact)\n",
      "   Christensenellaceae                      ID=NCBITaxon_990719 depth=10 score=1.00\n",
      "\n",
      "Query: Lactobacillaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33958\n",
      "\n",
      "Query: Enterococcaceae\n",
      "   Enterococcaceae                          ID=NCBITaxon_81852 depth=10 (exact)\n",
      "   Enterococcaceae                          ID=NCBITaxon_81852 depth=10 score=1.00\n",
      "\n",
      "Query: Enterococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1350\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Lactococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1357\n",
      "\n",
      "Query: Oscillospira\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_119852\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Proteobacteria\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D020560\n",
      "\n",
      "Query: Enterobacteriaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_543\n",
      "\n",
      "Query: Citrobacter\n",
      "   Citrobacter                              ID=NCBITaxon_544   depth=12 (exact)\n",
      "   Citrobacter                              ID=NCBITaxon_544   depth=12 score=1.00\n",
      "\n",
      "Query: Klebsiella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_570\n",
      "\n",
      "Query: Salmonella\n",
      "   Salmonella                               ID=NCBITaxon_590   depth=12 (exact)\n",
      "   Salmonella                               ID=NCBITaxon_590   depth=12 score=1.00\n",
      "\n",
      "Query: Shigella\n",
      "   Shigella                                 ID=NCBITaxon_620   depth=12 (exact)\n",
      "   Shigella                                 ID=NCBITaxon_620   depth=12 score=1.00\n",
      "\n",
      "Query: ParaBacteroides\n",
      "   Parabacteroides                          ID=NCBITaxon_375288 depth=16 (exact)\n",
      "   Parabacteroides                          ID=NCBITaxon_375288 depth=16 score=1.00\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Lachnospiraceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_186803\n",
      "\n",
      "Query: Roseburia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_841\n",
      "\n",
      "Query: Ruminococcus\n",
      "   Ruminococcus                             ID=NCBITaxon_1263  depth=12 (exact)\n",
      "   Ruminococcus                             ID=NCBITaxon_1263  depth=12 score=1.00\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Muribaculaceae\n",
      "   Muribaculaceae                           ID=NCBITaxon_2005473 depth=14 (exact)\n",
      "   Muribaculaceae                           ID=NCBITaxon_2005473 depth=14 score=1.00\n",
      "\n",
      "Query: Faecalibaculum\n",
      "   Faecalibaculum                           ID=NCBITaxon_1729679 depth=12 (exact)\n",
      "   Faecalibaculum                           ID=NCBITaxon_1729679 depth=12 score=1.00\n",
      "\n",
      "Query: Flavonifractor\n",
      "   Flavonifractor                           ID=NCBITaxon_946234 depth=12 (exact)\n",
      "   Flavonifractor                           ID=NCBITaxon_946234 depth=12 score=1.00\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_990719\n",
      "\n",
      "Query: gut Flavonifractor genus\n",
      "Query: gut Christensenellaceae family\n",
      "Query: Flavonifractor\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_946234\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_990719\n",
      "\n",
      "Query: Flavonifractor\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_946234\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_990719\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_990719\n",
      "\n",
      "Query: Flavonifractor\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_946234\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_990719\n",
      "\n",
      "Query: Flavonifractor\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_946234\n",
      "\n",
      "Query: Christensenellaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_990719\n",
      "\n",
      "Query: Flavonifractor\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_946234\n",
      "\n",
      "Query: genus Ruminococcaceae\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_81852\n",
      "\n",
      "Query: UCG009\n",
      "Query: genus Subdoligranulum\n",
      "   Subdoligranulum                          ID=NCBITaxon_292632 depth=12 score=0.86\n",
      "\n",
      "Query: Family Ruminococcaceae\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D008835\n",
      "\n",
      "Query: genus Coprococcus2\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1350\n",
      "\n",
      "Query: genus Eggerthella\n",
      "   Eggerthella                              ID=NCBITaxon_84111 depth=12 score=0.85\n",
      "\n",
      "Query: genus Eubacterium\n",
      "   Eubacterium                    UI=D005051  Tree=B03.353.625.750  scoreMESH=0.79\n",
      "Query: xylanophilum group\n",
      "   Eubacterium xylanophilum                 ID=NCBITaxon_39497 depth=14 score=0.79\n",
      "\n",
      "Query: predominant bacteria\n",
      "   Bacteria                                 ID=NCBITaxon_2     depth=0  score=1.00\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1776336\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1776336\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1776336\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1776336\n",
      "\n",
      "Query: Ruminococcaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1776336\n",
      "\n",
      "Query: gut bacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Lachnoclostridium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1506553\n",
      "\n",
      "Query: Lachnostridium\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1485\n",
      "\n",
      "Query: Lachnoclostridium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1506553\n",
      "\n",
      "Query: Parabacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_375288\n",
      "\n",
      "Query: Anaerostipes\n",
      "   Anaerostipes                             ID=NCBITaxon_207244 depth=12 (exact)\n",
      "   Anaerostipes                             ID=NCBITaxon_207244 depth=12 score=1.00\n",
      "\n",
      "Query: Ruminococcus-2\n",
      "   Ruminococcus                             ID=NCBITaxon_1263  depth=12 score=1.00\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Akkermansia muciniphila\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239935\n",
      "\n",
      "Query: Akkermansia muciniphila\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239935\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Listeria monocytogenes\n",
      "   Listeria monocytogenes                   ID=NCBITaxon_1639  depth=14 (exact)\n",
      "   Listeria monocytogenes                   ID=NCBITaxon_1639  depth=14 score=1.00\n",
      "\n",
      "Query: Haemophilus\n",
      "   Haemophilus                              ID=NCBITaxon_724   depth=12 (exact)\n",
      "   Haemophilus                              ID=NCBITaxon_724   depth=12 score=1.00\n",
      "\n",
      "Query: Haemophilu\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_724\n",
      "\n",
      "Query: Coprococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_33042\n",
      "\n",
      "Query: Haemophilus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_724\n",
      "\n",
      "Query: Roseburia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_841\n",
      "\n",
      "Query: Faecalibacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_216851\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Lachnospira\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_28050\n",
      "\n",
      "Query: Prevotella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: Streptococcus\n",
      "   Streptococcus                            ID=NCBITaxon_1301  depth=12 (exact)\n",
      "   Streptococcus                            ID=NCBITaxon_1301  depth=12 score=1.00\n",
      "\n",
      "Query: Bifidobacterium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Akkermansia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Desulfovibrio\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_872\n",
      "\n",
      "Query: psychobiotics\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D046939\n",
      "\n",
      "Query: psychobiotics\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D046939\n",
      "\n",
      "Query: Bacteroides uniformis\n",
      "   Bacteroides uniformis                    ID=NCBITaxon_820   depth=18 (exact)\n",
      "   Bacteroides uniformis                    ID=NCBITaxon_820   depth=18 score=1.00\n",
      "\n",
      "Query: Roseburia inulinivorans\n",
      "   Roseburia inulinivorans                  ID=NCBITaxon_360807 depth=14 (exact)\n",
      "   Roseburia inulinivorans                  ID=NCBITaxon_360807 depth=14 score=1.00\n",
      "\n",
      "Query: Eubacterium rectale\n",
      "   Treponema rectale                        ID=NCBITaxon_744512 depth=14 score=0.78\n",
      "\n",
      "Query: Faecalibacterium prausnitzii\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_853\n",
      "\n",
      "Query: psychobiotics\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D046939\n",
      "\n",
      "Query: phylum Proteobacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D020560\n",
      "\n",
      "Query: Enterobacteriaceae\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_543\n",
      "\n",
      "Query: coliform genera\n",
      "Query: Escherichia\n",
      "   Escherichia                              ID=NCBITaxon_561   depth=12 (exact)\n",
      "   Escherichia                              ID=NCBITaxon_561   depth=12 score=1.00\n",
      "\n",
      "Query: Klebsiella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_570\n",
      "\n",
      "Query: Klebsiella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_570\n",
      "\n",
      "Query: Clostridium\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1485\n",
      "\n",
      "Query: Bilophila\n",
      "   Bilophila                                ID=NCBITaxon_35832 depth=12 (exact)\n",
      "   Bilophila                                ID=NCBITaxon_35832 depth=12 score=1.00\n",
      "\n",
      "Query: Escherichia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_561\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Enterococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1350\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Lactobacillus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1578\n",
      "\n",
      "Query: Lachnospiraceae Incertae Sedis\n",
      "   Lachnospiraceae incertae sedis           ID=NCBITaxon_2840493 depth=12 (exact)\n",
      "   Lachnospiraceae incertae sedis           ID=NCBITaxon_2840493 depth=12 score=1.00\n",
      "\n",
      "Query: Barnesiella\n",
      "   Barnesiella                              ID=NCBITaxon_397864 depth=16 (exact)\n",
      "   Barnesiella                              ID=NCBITaxon_397864 depth=16 score=1.00\n",
      "\n",
      "Query: Romboutsia\n",
      "   Romboutsia                               ID=NCBITaxon_1501226 depth=12 (exact)\n",
      "   Romboutsia                               ID=NCBITaxon_1501226 depth=12 score=1.00\n",
      "\n",
      "Query: A  muciniphila\n",
      "   Akkermansia muciniphila                  ID=NCBITaxon_239935 depth=16 score=0.80\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Bifidobacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Prevotella\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Bifidobacteria\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1678\n",
      "\n",
      "Query: Lactobacilli\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_1591090\n",
      "\n",
      "Query: Bacteroides genus\n",
      "   Bacteroides                              ID=NCBITaxon_816   depth=16 score=0.77\n",
      "\n",
      "Query: Firmicutes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1386\n",
      "\n",
      "Query: Bacteroidetes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_226197\n",
      "\n",
      "Query: Bacteroides\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_816\n",
      "\n",
      "Query: Lachnospiraceae g\n",
      "   Lachnospiraceae                          ID=NCBITaxon_186803 depth=10 score=1.00\n",
      "\n",
      "Query: Ruminococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1263\n",
      "\n",
      "Query: Akkermansia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_239934\n",
      "\n",
      "Query: Megamonas\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_158846\n",
      "\n",
      "Query: Prevotellaceae NK3B31\n",
      "   Reusing existing URI by label: http://purl.obolibrary.org/obo/NCBITaxon_838\n",
      "\n",
      "Query: butyrate-producing bacteria\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_2811711\n",
      "\n",
      "Query: Lachnospira\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_28050\n",
      "\n",
      "Query: Subdoligranulum\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_292632\n",
      "\n",
      "Query: Blautia\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_572511\n",
      "\n",
      "Query: Dialister\n",
      "   Dialister                                ID=NCBITaxon_39948 depth=12 (exact)\n",
      "   Dialister                                ID=NCBITaxon_39948 depth=12 score=1.00\n",
      "\n",
      "Query: acetate-producing bacteria\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n",
      "Query: Streptococcus\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_1301\n",
      "\n",
      "Query: Parasutterella\n",
      "   Parasutterella                           ID=NCBITaxon_577310 depth=12 (exact)\n",
      "   Parasutterella                           ID=NCBITaxon_577310 depth=12 score=1.00\n",
      "\n",
      "Query: Dialister\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCBITaxon_39948\n",
      "\n",
      "Query: intestinal bacteria\n",
      "   Reusing existing URI by label: https://www.ncbi.nlm.nih.gov/mesh/D039903\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "import os\n",
    "import re\n",
    "import unicodedata\n",
    "import json\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 1. Setup paths and namespaces\n",
    "# -----------------------------------------------------------------------------\n",
    "path = str(Path(os.path.abspath(os.getcwd())).absolute())\n",
    "json_file = r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\data\\train_platinum\\train_platinum.json\"\n",
    "#json_file = os.path.join(path, \"train_gold.json\")\n",
    "#vertex_file = os.path.join(path, \"train_vertexsets.json\")\n",
    "tokenized_file = os.path.join(path, \"tokenized_sentences_with_entitiesv2.json\")\n",
    "save_path = os.path.join(path, \"rdf\")\n",
    "os.makedirs(save_path, exist_ok=True)\n",
    "\n",
    "GUTBRAIN = Namespace(\"https://hereditary.dei.unipd.it/ontology/gutbrain/resource/\")\n",
    "GUTPROP = Namespace(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/\")\n",
    "\n",
    "PAPER_CLASS       = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Paper\")\n",
    "MENTION_CLASS     = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Mention\")\n",
    "PAPER_ABSTRACT    = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/PaperAbstract\")\n",
    "PAPER_TITLE       = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/PaperTitle\")\n",
    "PAPER_COLLECTION  = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/PaperCollection\")\n",
    "PROJECT           = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Project\")\n",
    "SAMPLE            = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Sample\")\n",
    "SENTENCE          = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Sentence\")\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 2. Load the JSON paper data\n",
    "# -----------------------------------------------------------------------------\n",
    "with open(json_file, \"r\", encoding=\"utf-8\") as f:\n",
    "    data = json.load(f)\n",
    "# -----------------------------------------------------------------------------\n",
    "# 3. Mapping dictionaries (keys must be in Title case)\n",
    "# -----------------------------------------------------------------------------\n",
    "label_mapping = {\n",
    "    \"Anatomical Location\":   URIRef(\"https://w3id.org/brainteaser/ontology/schema/AnatomicalSite\"),\n",
    "    \"Animal\":                URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Animal\"),\n",
    "    \"Biomedical Technique\":  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/BiomedicalTechnique\"),\n",
    "    \"Bacteria\":              URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Species\"),\n",
    "    \"Chemical\":              URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Chemical\"),\n",
    "    \"Dietary Supplement\":    URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/DietarySupplement\"),\n",
    "    \"DDF\":                   URIRef(\"https://w3id.org/brainteaser/ontology/schema/DiseaseDisorderOrFinding\"),\n",
    "    \"Drug\":                  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Drug\"),\n",
    "    \"Food\":                  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Food\"),\n",
    "    \"Gene\":                  URIRef(\"https://w3id.org/brainteaser/ontology/schema/Gene\"),\n",
    "    \"Human\":                 URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Human\"),\n",
    "    \"Microbiome\":            URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Microbiome\"),\n",
    "    \"Statistical Technique\": URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/StatisticalTechnique\")\n",
    "}\n",
    "concept_scheme_mapping = {\n",
    "    \"Anatomical Location\":   URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/AnatomicSite\"),\n",
    "    \"Animal\":                URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Animal\"),\n",
    "    \"Human\":                 URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Human\"),\n",
    "    \"Drug\":                  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Drug\"),\n",
    "    \"Gene\":                  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Gene\"),\n",
    "    \"Dietary Supplement\":    URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/dietarySupplement\"),\n",
    "    \"DDF\":                   URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/diseaseDisorderOrFindingTaxonomy\"),\n",
    "    \"Metabolite\":            URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Metabolite\"),\n",
    "    \"Bacteria\":               URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Bacteria\"),\n",
    "    \"Food\":                  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Food\"),\n",
    "    \"Chemical\":              URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Chemical\"),\n",
    "    \"Biomedical Technique\":  URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/BiomedicalTechnique\"),\n",
    "    \"Microbiome\":            URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Microbiome\"),\n",
    "    \"Statistical Technique\": URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/StatisticalTechnique\")\n",
    "}\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 4. Initialize the RDF graph and bind namespaces\n",
    "# -----------------------------------------------------------------------------\n",
    "BACTERIA_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Species\")\n",
    "FAMILY_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Family\")\n",
    "OBO_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "MESH_BASE = \"https://www.ncbi.nlm.nih.gov/mesh/\"\n",
    "BACTERIA_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Bacteria\")\n",
    "\n",
    "g = Graph()\n",
    "g.bind(\"gutbrain\", GUTBRAIN)\n",
    "g.bind(\"rdfs\", RDFS)\n",
    "g.bind(\"xsd\", XSD)\n",
    "g.bind(\"skos\", SKOS)\n",
    "g.bind(\"owl\", OWL)\n",
    "g.bind(\"gutprop\", GUTPROP)\n",
    "\n",
    "g.add((SKOS.inScheme, RDF.type, OWL.ObjectProperty))\n",
    "g.add((SKOS.broaderTransitive, RDF.type, OWL.ObjectProperty))\n",
    "\n",
    "g.add((GUTPROP.partOf, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTPROP.partOf, RDFS.label, Literal(\"partOf\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.hasTitle, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTPROP.hasTitle, RDFS.label, Literal(\"hasTitle\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.hasAbstract, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTPROP.hasAbstract, RDFS.label, Literal(\"hasAbstract\", datatype=XSD.string)))\n",
    "\n",
    "g.add((GUTPROP.containedIn, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTPROP.containedIn, RDFS.label, Literal(\"containedIn\", datatype=XSD.string)))\n",
    "g.add((GUTBRAIN.contains, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTBRAIN.contains, RDFS.label, Literal(\"contains\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.composedOf, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTPROP.composedOf, RDFS.label, Literal(\"composedOf\", datatype=XSD.string)))\n",
    "\n",
    "g.add((GUTPROP.locatedIn, RDF.type, OWL.ObjectProperty))\n",
    "g.add((GUTPROP.locatedIn, RDFS.label, Literal(\"locatedIn\", datatype=XSD.string)))\n",
    "\n",
    "g.add((GUTPROP.paperId, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.paperId, RDFS.label, Literal(\"paperId\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.paperAnnotator, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.paperAnnotator, RDFS.label, Literal(\"paperAnnotator\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.paperYear, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.paperYear, RDFS.label, Literal(\"paperYear\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.paperJournal, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.paperJournal, RDFS.label, Literal(\"paperJournal\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.paperAuthor, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.paperAuthor, RDFS.label, Literal(\"paperAuthor\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.numberOfRunsFound, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.numberOfRunsFound, RDFS.label, Literal(\"numberOfRunsFound\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.NCBITaxonID, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.NCBITaxonID, RDFS.label, Literal(\"NCBITaxonID\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.sdRelativeAbundance, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.sdRelativeAbundance, RDFS.label, Literal(\"sdRelativeAbundance\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.medianRelativeAbundance, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.medianRelativeAbundance, RDFS.label, Literal(\"medianRelativeAbundance\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.meanRelativeAbundance, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.meanRelativeAbundance, RDFS.label, Literal(\"meanRelativeAbundance\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.scientificName, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.scientificName, RDFS.label, Literal(\"scientificName\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.hasMentionText, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.hasMentionText, RDFS.label, Literal(\"hasMentionText\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.hasSentenceText, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.hasSentenceText, RDFS.label, Literal(\"hasSentenceText\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.hasTitleText, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.hasTitleText, RDFS.label, Literal(\"hasTitleText\", datatype=XSD.string)))\n",
    "g.add((GUTPROP.hasAbstractText, RDF.type, OWL.DatatypeProperty))\n",
    "g.add((GUTPROP.hasAbstractText, RDFS.label, Literal(\"hasAbstractText\", datatype=XSD.string)))\n",
    "\n",
    "for scheme_uri in set(concept_scheme_mapping.values()):\n",
    "    keys = [k for k, v in concept_scheme_mapping.items() if v == scheme_uri]\n",
    "    label_text = \", \".join(k.title() for k in keys) + \" Concept Scheme\"\n",
    "    g.add((scheme_uri, RDF.type, SKOS.ConceptScheme))\n",
    "    g.add((scheme_uri, RDFS.label, Literal(label_text, datatype=XSD.string)))\n",
    "\n",
    "is_train_platinum = \"train_platinum\" in os.path.basename(json_file)\n",
    "#is_train_gold = \"train_gold\" in os.path.basename(json_file)\n",
    "\n",
    "#if is_train_gold:\n",
    " #   gold_collection_uri = URIRef(GUTBRAIN[\"goldCollection\"])\n",
    " #   label_text = \"goldCollection\"\n",
    "  #  g.add((gold_collection_uri, RDF.type, PAPER_COLLECTION))\n",
    "   # g.add((gold_collection_uri, RDFS.label, Literal(label_text, datatype=XSD.string)))\n",
    "    \n",
    "if is_train_platinum:\n",
    "    platinum_collection_uri = URIRef(GUTBRAIN[\"platinumCollection\"])\n",
    "    label_text = \"platinumCollection\"\n",
    "    g.add((platinum_collection_uri, RDF.type, PAPER_COLLECTION))\n",
    "    g.add((platinum_collection_uri, RDFS.label, Literal(label_text, datatype=XSD.string)))\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Utility functions\n",
    "# -----------------------------------------------------------------------------\n",
    "def create_uri_fragment(text):\n",
    "    cleaned = re.sub(r'<[^>]*>', '', text)\n",
    "    cleaned = normalize_text(cleaned)\n",
    "    # Allow word characters, digits, underscores, hyphens, and Greek characters (U+0370 to U+03FF)\n",
    "    cleaned = re.sub(r'[^\\w\\u0370-\\u03FF-]', '_', cleaned)\n",
    "    return cleaned\n",
    "\n",
    "def to_camel_case(s):\n",
    "    s = re.sub(r'[^\\w\\s]', '', s)\n",
    "    parts = re.split(r'\\s+', s.strip())\n",
    "    if not parts:\n",
    "        return \"\"\n",
    "    return parts[0].lower() + ''.join(word.title() for word in parts[1:])\n",
    "\n",
    "def normalize_text(text):\n",
    "    return unicodedata.normalize('NFC', text)\n",
    "\n",
    "def preprocess(term):\n",
    "    return term.replace('_', ' ').strip()\n",
    "\n",
    "def normalize_to_ascii(s: str) -> str:\n",
    "    # 1) Decomponi i caratteri accentati in base + segni diacritici\n",
    "    nfkd = unicodedata.normalize('NFKD', s)\n",
    "    # 2) Filtra solo i caratteri ASCII\n",
    "    ascii_bytes = nfkd.encode('ascii', 'ignore')\n",
    "    # 3) Ritorna la stringa decodificata\n",
    "    return ascii_bytes.decode('ascii')\n",
    "\n",
    "mention_counter   = 1\n",
    "tokenized_mentions = {}\n",
    "missing_mentions = []\n",
    "created = {}\n",
    "label2uri = {}\n",
    "created[\"Bacteria\"] = URIRef(f\"{OBO_BASE}NCBITaxon_2\")\n",
    "g.add((created[\"Bacteria\"], RDF.type, FAMILY_CLASS))\n",
    "g.add((created[\"Bacteria\"], RDF.type, SKOS.Concept))\n",
    "g.add((created[\"Bacteria\"], RDFS.label, Literal(\"Bacteria\", datatype=XSD.string)))\n",
    "g.add((created[\"Bacteria\"], SKOS.inScheme, BACTERIA_CONCEPT_SCHEME))\n",
    "#created[\"Bifidobacterium\"] = URIRef(f\"{MESH_BASE}D001644\")\n",
    "#g.add((created[\"Bifidobacterium\"], RDF.type, BACTERIA_CLASS))\n",
    "#g.add((created[\"Bifidobacterium\"], RDF.type, SKOS.Concept))\n",
    "#g.add((created[\"Bifidobacterium\"], RDFS.label, Literal(\"Bifidobacterium\", datatype=XSD.string)))\n",
    "#g.add((created[\"Bifidobacterium\"], SKOS.inScheme, BACTERIA_CONCEPT_SCHEME))\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 5. Process each paper (each key in the JSON represents a paper)\n",
    "# -----------------------------------------------------------------------------\n",
    "for paper_id, paper_data in data.items():\n",
    "    paper_uri = URIRef(GUTBRAIN[f\"paper_{paper_id}\"])\n",
    "    g.add((paper_uri, RDF.type, PAPER_CLASS))\n",
    "    \n",
    "    if is_train_platinum:\n",
    "        g.add((paper_uri, GUTPROP.partOf, platinum_collection_uri))\n",
    "        g.add((platinum_collection_uri, GUTBRAIN.contains, paper_uri))\n",
    "\n",
    "    #if is_train_gold:\n",
    "    #    g.add((paper_uri, GUTPROP.partOf, gold_collection_uri))\n",
    "    #    g.add((gold_collection_uri, GUTBRAIN.contains, paper_uri))\n",
    "    \n",
    "    # Each paper gets its own mention node\n",
    "    #paper_mention = URIRef(GUTBRAIN[f\"mention_{paper_id}\"])\n",
    "    #g.add((paper_mention, RDF.type, MENTION_CLASS))\n",
    "    #g.add((paper_uri, GUTPROP.hasMention, paper_mention))\n",
    "    \n",
    "    metadata = paper_data.get(\"metadata\", {})\n",
    "    full_title = metadata.get(\"title\", None)\n",
    "    full_abstract = metadata.get(\"abstract\", None)\n",
    "    try:\n",
    "        paper_id_val = int(paper_id)\n",
    "    except ValueError:\n",
    "        paper_id_val = paper_id\n",
    "    paper_annotator = metadata.get(\"annotator\", None)\n",
    "    paper_year = metadata.get(\"year\", None)\n",
    "    paper_journal = metadata.get(\"journal\", None)\n",
    "    paper_author = metadata.get(\"author\", None)\n",
    "    \n",
    "    g.add((paper_uri, GUTPROP.paperId, Literal(paper_id_val, datatype=XSD.integer)))\n",
    "    if paper_annotator is not None:\n",
    "        g.add((paper_uri, GUTPROP.paperAnnotator, Literal(paper_annotator, datatype=XSD.string)))\n",
    "    if paper_year is not None:\n",
    "        g.add((paper_uri, GUTPROP.paperYear, Literal(paper_year, datatype=XSD.gYear)))\n",
    "    if paper_journal is not None:\n",
    "        g.add((paper_uri, GUTPROP.paperJournal, Literal(paper_journal, datatype=XSD.string)))\n",
    "    if paper_author is not None:\n",
    "        g.add((paper_uri, GUTPROP.paperAuthor, Literal(paper_author, datatype=XSD.string)))\n",
    "    \n",
    "    title_texts = []\n",
    "    abstract_texts = []\n",
    "    mention_created = False\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        label_title = raw_label.title() if raw_label.lower() != \"ddf\" else \"DDF\"\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        cleaned_text = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "        \n",
    "        if label_title == \"Bacteria\":\n",
    "            if cleaned_text_span.lower().endswith(\" bacteria\"):\n",
    "                term_raw = \"Bacteria\"\n",
    "            elif cleaned_text_span.lower().endswith(\" Bifidobacterium\"):\n",
    "                term_raw = \"Bifidobacterium\"\n",
    "            else:\n",
    "                term_raw = cleaned_text_span\n",
    "                \n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "                \n",
    "            hits = genus_abbrev_lookup(term)\n",
    "            if hits:\n",
    "                for l,t,d,s in hits: \n",
    "                    print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score={s:.2f} (abbr)\")\n",
    "                    entity_uri = URIRef(f\"{OBO_BASE}{t}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    label2uri[label_name.lower()] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, BACTERIA_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(label_name, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, BACTERIA_CONCEPT_SCHEME))\n",
    "                    g.add((entity_uri, SKOS.broaderTransitive, created[\"Bacteria\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_bacteria_{term_raw}\", datatype=XSD.string)))\n",
    "                    #mention_counter += 1\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "                    continue\n",
    "\n",
    "            ex = exact_ix.get(term.lower(), [])\n",
    "            if ex:\n",
    "                for label_name, taxon_id, depth in ex:\n",
    "                    print(f\"   {label_name:40s} ID={taxon_id:15s} depth={depth:<2d} (exact)\")\n",
    "                    entity_uri = URIRef(f\"{OBO_BASE}{taxon_id}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    label2uri[label_name.lower()] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, BACTERIA_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(label_name, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, BACTERIA_CONCEPT_SCHEME))\n",
    "                    g.add((entity_uri, SKOS.broaderTransitive, created[\"Bacteria\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_bacteria_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            if not cos:\n",
    "                matches = find_mesh_match(term, name_index)\n",
    "                if matches: \n",
    "                    for name, ui, tree, score in matches[:1]:\n",
    "                        name_key = name.lower()\n",
    "                        reused = False\n",
    "                        if name_key in label2uri:\n",
    "                            entity_uri = label2uri[name_key]\n",
    "                            print(f\"   Reusing existing URI by label: {entity_uri}\\n\")\n",
    "                            reused = True\n",
    "                            continue\n",
    "                        if reused:\n",
    "                            break\n",
    "                            \n",
    "                        print(f\"   {name:30s} UI={ui:8s} Tree={tree:12s}  scoreMESH={score:.2f}\")\n",
    "                        #mesh_key = label_name.lower() #DA FIXARE\n",
    "                        name_uri = URIRef(f\"{MESH_BASE}{ui}\")\n",
    "                        created[term_raw] = name_uri #created[MESH_raw] = name_uri\n",
    "                        label2uri[name_key] = name_uri #label2uri[MESH_key] = name_uri\n",
    "                        g.add((name_uri, RDF.type, BACTERIA_CLASS))\n",
    "                        g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                        g.add((name_uri, RDFS.label, Literal(name, datatype=XSD.string)))\n",
    "                        g.add((name_uri, SKOS.inScheme, BACTERIA_CONCEPT_SCHEME))\n",
    "                        g.add((name_uri, SKOS.broaderTransitive, created[\"Bacteria\"]))\n",
    "                        mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                        g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                        g.add((mention_uri, RDFS.label, Literal(f\"mention_bacteria_{term_raw}\", datatype=XSD.string)))\n",
    "                        #mention_counter += 1\n",
    "                        g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                        g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                        tokenized_mentions[term_raw] = mention_uri\n",
    "                        continue\n",
    "                else:\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_bacteria_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((created[\"Bacteria\"], GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    \n",
    "            else:\n",
    "                for l,t,d,s in cos[:1]:\n",
    "                    print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score={s:.2f}\")\n",
    "                    name_uri = URIRef(f\"{OBO_BASE}{t}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    label2uri[label_name.lower()] = name_uri\n",
    "                    g.add((name_uri, RDF.type, BACTERIA_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(l, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, BACTERIA_CONCEPT_SCHEME))\n",
    "                    g.add((name_uri, SKOS.broaderTransitive, created[\"Bacteria\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_bacteria_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                print()\n",
    "        \n",
    "        location_lower = entity.get(\"location\", \"\").strip().lower()\n",
    "        if location_lower == \"title\":\n",
    "            title_texts.append(cleaned_text_span)\n",
    "        elif location_lower == \"abstract\":\n",
    "            abstract_texts.append(cleaned_text_span)\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "            #entity_node = URIRef(GUTBRAIN[f\"entity_{paper_id}_{i}\"])\n",
    "            #g.add((entity_node, RDF.type, MENTION_CLASS))\n",
    "            #g.add((entity_node, RDF.type, SKOS.Concept))\n",
    "            #g.add((entity_node, RDFS.label, Literal(label_title, datatype=XSD.string)))\n",
    "            #g.add((entity_node, GUTPROP.annotator, Literal(metadata.get(\"annotator\", \"unknown\"), datatype=XSD.string)))\n",
    "            #g.add((entity_node, GUTPROP.location, Literal(entity.get(\"location\", \"\"), datatype=XSD.string)))\n",
    "            #g.add((entity_node, GUTPROP.text_span, Literal(cleaned_text_span, datatype=XSD.string)))\n",
    "    \n",
    "    \n",
    "    if full_title is None and title_texts:\n",
    "        full_title = \" \".join(title_texts)\n",
    "    if full_abstract is None and abstract_texts:\n",
    "        full_abstract = \" \".join(abstract_texts)\n",
    "    \n",
    "    if full_title:\n",
    "        title_uri = URIRef(GUTBRAIN[f\"title_{paper_id}\"])\n",
    "        g.add((title_uri, RDF.type, PAPER_TITLE))\n",
    "        g.add((title_uri, GUTPROP.hasTitleText, Literal(full_title, datatype=XSD.string)))\n",
    "        g.add((paper_uri, GUTPROP.hasTitle, title_uri))\n",
    "    \n",
    "    if full_abstract:\n",
    "        abstract_uri = URIRef(GUTBRAIN[f\"abstract_{paper_id}\"])\n",
    "        g.add((abstract_uri, RDF.type, PAPER_ABSTRACT))\n",
    "        g.add((abstract_uri, GUTPROP.hasAbstractText, Literal(full_abstract, datatype=XSD.string)))\n",
    "        g.add((paper_uri, GUTPROP.hasAbstract, abstract_uri))\n",
    "    \n",
    "    # Process relations\n",
    "    #relations = paper_data.get(\"relations\", [])\n",
    "    #for relation in relations:\n",
    "        #subj_text = relation.get(\"subject_text_span\", \"\").strip()\n",
    "        #cleaned_subj_text = re.sub(r'<[^>]*>', '', subj_text).strip()\n",
    "        #subj_fragment = create_uri_fragment(cleaned_subj_text)\n",
    "        #subj_uri = URIRef(GUTBRAIN[subj_fragment])\n",
    "        #if not list(g.triples((subj_uri, None, None))):\n",
    "           # print(f\"Warning: Subject not recognized: {subj_uri}. Info: '{cleaned_subj_text}'\")\n",
    "        \n",
    "        #obj_text = relation.get(\"object_text_span\", \"\").strip()\n",
    "        #cleaned_obj_text = re.sub(r'<[^>]*>', '', obj_text).strip()\n",
    "        #obj_fragment = create_uri_fragment(cleaned_obj_text)\n",
    "        #obj_uri = URIRef(GUTBRAIN[obj_fragment])\n",
    "        #if not list(g.triples((obj_uri, None, None))):\n",
    "            #print(f\"Warning: Object not recognized: {obj_uri}. Info: '{cleaned_obj_text}'\")\n",
    "        \n",
    "        #pred_text = relation.get(\"predicate\", \"\").strip()\n",
    "        #pred_text_clean = to_camel_case(pred_text)\n",
    "        #pred_uri = URIRef(GUTPROP[pred_text_clean])\n",
    "        #print(f\"Predicate: '{pred_text}' -> '{pred_text_clean}'\")\n",
    "        #g.add((pred_uri, RDF.type, OWL.ObjectProperty))\n",
    "        #g.add((pred_uri, RDFS.label, Literal(pred_text_clean, datatype=XSD.string)))\n",
    "       # g.add((subj_uri, pred_uri, obj_uri))\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue # o None, se ti serve\n",
    "\n",
    "        if label != \"bacteria\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical)\n",
    "        #print(cleaned_text_span)\n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_bacteria_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))\n",
    "\n",
    "        #entity_uri = URIRef(GUTBRAIN[canonical])\n",
    "        #g.add((entity_uri, GUTPROP.containedIn, mention_node))\n",
    "        #lit = Literal(ent_text, datatype=XSD.string)\n",
    "        #if (mention_node, GUTPROP.hasMentionText, lit) not in g:\n",
    "            #g.add((mention_node, GUTPROP.hasMentionText, lit))\n",
    "#print(\"\\nDettaglio di tokenized_mentions:\")\n",
    "#for canonical, mention_uri in tokenized_mentions.items():\n",
    "    #print(f\"{canonical!r}  {mention_uri}\")\n",
    "# -----------------------------------------------------------------------------\n",
    "# 7. Serialize and print the graph in Turtle format\n",
    "# -----------------------------------------------------------------------------\n",
    "#output_file = os.path.join(save_path, \"gutbrain_entities.ttl\")\n",
    "#ttl_output = g.serialize(format=\"turtle\")\n",
    "#with open(output_file, \"w\", encoding=\"utf-8\") as f_out:\n",
    "    #f_out.write(ttl_output)\n",
    "\n",
    "#print(f\"The RDF graph has been saved in {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "739388c3-48c0-4160-8cfc-ddbeb52de272",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: simple sugars\n",
      "Query: sfa\n",
      "Query: monounsaturated fatty acids\n",
      "   monounsaturated fatty acid               URI=http://purl.obolibrary.org/obo/CHEBI_25413 score=0.78\n",
      "\n",
      "Query: metabolite acetate\n",
      "   metabolite                               URI=http://purl.obolibrary.org/obo/CHEBI_25212 score=0.81\n",
      "\n",
      "Query: fat and sugar content\n",
      "Query: tnf-a\n",
      "Query: il-6\n",
      "Query: il-17\n",
      "Query: serum hormone\n",
      "Query: insulin\n",
      "   insulin                                  URI=http://purl.obolibrary.org/obo/CHEBI_145810\n",
      "   insulin                                  URI=http://purl.obolibrary.org/obo/CHEBI_145810 score=1.00\n",
      "\n",
      "Query: testosterone\n",
      "   testosterone                             URI=http://purl.obolibrary.org/obo/CHEBI_17347\n",
      "   testosterone                             URI=http://purl.obolibrary.org/obo/CHEBI_17347 score=1.00\n",
      "\n",
      "Query: skin lipids\n",
      "Query: triglycerides\n",
      "Query: non-esterified free fatty acids\n",
      "Query: skin triglycerides\n",
      "Query: hormones\n",
      "Query: skin lipids\n",
      "Query: melanin\n",
      "   Melanin                                  URI=http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "   Melanin                                  URI=http://purl.obolibrary.org/obo/CHEBI_89634 score=1.00\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: reactive oxygen species\n",
      "   reactive oxygen species                  URI=http://purl.obolibrary.org/obo/CHEBI_26523\n",
      "   reactive oxygen species                  URI=http://purl.obolibrary.org/obo/CHEBI_26523 score=1.00\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: neuromelanin\n",
      "Query: herpesvirus\n",
      "Query: melanosomes\n",
      "Query: herpesvirus\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: atp\n",
      "   ATP                                      URI=http://purl.obolibrary.org/obo/CHEBI_15422\n",
      "   ATP(3-)                                  URI=http://purl.obolibrary.org/obo/CHEBI_57299 score=1.00\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: atp\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_57299\n",
      "\n",
      "Query: atp\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_57299\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: melanin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89634\n",
      "\n",
      "Query: flavonoids\n",
      "   flavonoids                               URI=http://purl.obolibrary.org/obo/CHEBI_72544\n",
      "   flavonoids                               URI=http://purl.obolibrary.org/obo/CHEBI_72544 score=1.00\n",
      "\n",
      "Query: isorhamnetin\n",
      "   isorhamnetin                             URI=http://purl.obolibrary.org/obo/CHEBI_6052\n",
      "   isorhamnetin                             URI=http://purl.obolibrary.org/obo/CHEBI_6052 score=1.00\n",
      "\n",
      "Query: quercetin\n",
      "   quercetin                                URI=http://purl.obolibrary.org/obo/CHEBI_16243\n",
      "   quercetin                                URI=http://purl.obolibrary.org/obo/CHEBI_16243 score=1.00\n",
      "\n",
      "Query: kaempferol\n",
      "   kaempferol                               URI=http://purl.obolibrary.org/obo/CHEBI_28499\n",
      "   kaempferol                               URI=http://purl.obolibrary.org/obo/CHEBI_28499 score=1.00\n",
      "\n",
      "Query: neurotrophins\n",
      "Query: neurotransmitters\n",
      "Query: stress-related hormones\n",
      "Query: inflammation-related cytokines\n",
      "Query: neuroactive compounds\n",
      "Query: neuroactive compounds\n",
      "Query: neuroactive compounds\n",
      "Query: gamma-aminobutyric acid\n",
      "   gamma-aminobutyric acid                  URI=http://purl.obolibrary.org/obo/CHEBI_16865\n",
      "   gamma-aminobutyric acid                  URI=http://purl.obolibrary.org/obo/CHEBI_16865 score=1.00\n",
      "\n",
      "Query: gaba\n",
      "   N-Oleoyl GABA                            URI=http://purl.obolibrary.org/obo/CHEBI_165543 score=0.82\n",
      "\n",
      "Query: serotonin\n",
      "   serotonin                                URI=http://purl.obolibrary.org/obo/CHEBI_28790\n",
      "   serotonin(1+)                            URI=http://purl.obolibrary.org/obo/CHEBI_350546 score=1.00\n",
      "\n",
      "Query: melatonin\n",
      "   melatonin                                URI=http://purl.obolibrary.org/obo/CHEBI_16796\n",
      "   melatonin                                URI=http://purl.obolibrary.org/obo/CHEBI_16796 score=1.00\n",
      "\n",
      "Query: kynurenine\n",
      "   kynurenine                               URI=http://purl.obolibrary.org/obo/CHEBI_28683\n",
      "   D-kynurenine                             URI=http://purl.obolibrary.org/obo/CHEBI_86262 score=1.00\n",
      "\n",
      "Query: kynurenic acid\n",
      "   kynurenic acid                           URI=http://purl.obolibrary.org/obo/CHEBI_18344\n",
      "   kynurenic acid                           URI=http://purl.obolibrary.org/obo/CHEBI_18344 score=1.00\n",
      "\n",
      "Query: dopamine\n",
      "   dopamine                                 URI=http://purl.obolibrary.org/obo/CHEBI_18243\n",
      "   dopamine                                 URI=http://purl.obolibrary.org/obo/CHEBI_18243 score=1.00\n",
      "\n",
      "Query: norepinephrine\n",
      "   Norepinephrine sulfate                   URI=http://purl.obolibrary.org/obo/CHEBI_90022 score=0.89\n",
      "\n",
      "Query: histamine\n",
      "   histamine                                URI=http://purl.obolibrary.org/obo/CHEBI_18295\n",
      "   histamine                                URI=http://purl.obolibrary.org/obo/CHEBI_18295 score=1.00\n",
      "\n",
      "Query: tryptamine\n",
      "   tryptamine                               URI=http://purl.obolibrary.org/obo/CHEBI_16765\n",
      "   tryptamine                               URI=http://purl.obolibrary.org/obo/CHEBI_16765 score=1.00\n",
      "\n",
      "Query: tyramine\n",
      "   tyramine                                 URI=http://purl.obolibrary.org/obo/CHEBI_15760\n",
      "   m-tyramine                               URI=http://purl.obolibrary.org/obo/CHEBI_89626 score=1.00\n",
      "\n",
      "Query: -phenylethylamine\n",
      "   phenylethylamine                         URI=http://purl.obolibrary.org/obo/CHEBI_50048 score=1.00\n",
      "\n",
      "Query: neuroactive compounds\n",
      "Query: plant-origin and microbially-formed neuroactive compounds\n",
      "Query: neuroactive compounds\n",
      "Query: neuroactive compounds\n",
      "Query: microbial metabolites\n",
      "Query: neuroactive compounds\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: neuroactive compounds\n",
      "Query: -aminobutyric acid\n",
      "   L-3-Aminobutyric acid                    URI=http://purl.obolibrary.org/obo/CHEBI_229980 score=1.00\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: acetylcholine\n",
      "   acetylcholine                            URI=http://purl.obolibrary.org/obo/CHEBI_15355\n",
      "   acetylcholine                            URI=http://purl.obolibrary.org/obo/CHEBI_15355 score=1.00\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: acetylcholine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_15355\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: short chain fatty acids\n",
      "   short-chain fatty acid                   URI=http://purl.obolibrary.org/obo/CHEBI_26666 score=0.83\n",
      "\n",
      "Query: scfa\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: acetylcholine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_15355\n",
      "\n",
      "Query: glutamate\n",
      "   L-glutamate(1-)                          URI=http://purl.obolibrary.org/obo/CHEBI_29985 score=1.00\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: short-chain fatty acids\n",
      "   short-chain fatty acid                   URI=http://purl.obolibrary.org/obo/CHEBI_26666 score=0.83\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: vitamin b6\n",
      "   vitamin B6                               URI=http://purl.obolibrary.org/obo/CHEBI_27306\n",
      "   vitamin B6                               URI=http://purl.obolibrary.org/obo/CHEBI_27306 score=1.00\n",
      "\n",
      "Query: fatty acid\n",
      "   fatty acid                               URI=http://purl.obolibrary.org/obo/CHEBI_35366\n",
      "   fatty acid                               URI=http://purl.obolibrary.org/obo/CHEBI_35366 score=1.00\n",
      "\n",
      "Query: microbial neuroactive metabolites\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: gamma-aminobutyric acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16865\n",
      "\n",
      "Query: arachidonic acid\n",
      "   arachidonic acid                         URI=http://purl.obolibrary.org/obo/CHEBI_15843\n",
      "   arachidonic acid                         URI=http://purl.obolibrary.org/obo/CHEBI_15843 score=1.00\n",
      "\n",
      "Query: sphingomyelin\n",
      "   sphingomyelin                            URI=http://purl.obolibrary.org/obo/CHEBI_64583\n",
      "   sphingomyelin                            URI=http://purl.obolibrary.org/obo/CHEBI_64583 score=1.00\n",
      "\n",
      "Query: dextran sodium sulphate\n",
      "Query: dextran sodium sulphate\n",
      "Query: dss\n",
      "Query: lipopolysaccharide\n",
      "   lipopolysaccharide                       URI=http://purl.obolibrary.org/obo/CHEBI_16412\n",
      "   lipopolysaccharide                       URI=http://purl.obolibrary.org/obo/CHEBI_16412 score=1.00\n",
      "\n",
      "Query: lps\n",
      "Query: dss\n",
      "Query: dss\n",
      "Query: dss\n",
      "Query: lps\n",
      "Query: paraquat\n",
      "   paraquat                                 URI=http://purl.obolibrary.org/obo/CHEBI_34905\n",
      "   paraquat                                 URI=http://purl.obolibrary.org/obo/CHEBI_34905 score=1.00\n",
      "\n",
      "Query: circulating lipocalin-2\n",
      "Query: il-6\n",
      "Query: dss\n",
      "Query: xingpijieyu formula\n",
      "Query: xingpijieyu formula\n",
      "Query: xpjyf\n",
      "Query: xpjyf\n",
      "Query: serum cytokines\n",
      "Query: glial fibrillary acidic protein\n",
      "Query: gfap\n",
      "Query: xpjyf\n",
      "Query: lipopolysaccharide\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16412\n",
      "\n",
      "Query: lps\n",
      "Query: cytokines\n",
      "Query: 2-deoxy-d-glucose\n",
      "   2-deoxy-D-glucose                        URI=http://purl.obolibrary.org/obo/CHEBI_15866\n",
      "   2-deoxy-D-glucose                        URI=http://purl.obolibrary.org/obo/CHEBI_15866 score=1.00\n",
      "\n",
      "Query: 2-dg\n",
      "   dG+                                      URI=http://purl.obolibrary.org/obo/CHEBI_190142 score=1.00\n",
      "\n",
      "Query: adenosine triphosphate\n",
      "   adenosine 3'-diphosphate,5'-triphosphate(6-) URI=http://purl.obolibrary.org/obo/CHEBI_145540 score=0.86\n",
      "\n",
      "Query: atp\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_57299\n",
      "\n",
      "Query: glucose-1-phosphate\n",
      "   D-glucose 3-phosphate                    URI=http://purl.obolibrary.org/obo/CHEBI_50416 score=1.00\n",
      "\n",
      "Query: g1p\n",
      "Query: xpjyf\n",
      "Query: glucose1pmetab-pwy\n",
      "Query: xpjyf\n",
      "Query: xpjyf\n",
      "Query: atp\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_57299\n",
      "\n",
      "Query: glucose-1-phosphate\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_50416\n",
      "\n",
      "Query: xpjyf\n",
      "Query: interferon-\n",
      "   interferon                               URI=http://purl.obolibrary.org/obo/CHEBI_52999 score=1.00\n",
      "\n",
      "Query: tumor necrosis factor-\n",
      "   Recombinant Tumor Necrosis Factor-alpha (human) URI=http://purl.obolibrary.org/obo/CHEBI_197440 score=0.77\n",
      "\n",
      "Query: indoleamine 2 3-dioxygenase-1\n",
      "Query: maltotriose\n",
      "   maltotriose                              URI=http://purl.obolibrary.org/obo/CHEBI_61993\n",
      "   maltotriose                              URI=http://purl.obolibrary.org/obo/CHEBI_61993 score=1.00\n",
      "\n",
      "Query: isomaltose\n",
      "   isomaltose                               URI=http://purl.obolibrary.org/obo/CHEBI_28189\n",
      "   isomaltose                               URI=http://purl.obolibrary.org/obo/CHEBI_28189 score=1.00\n",
      "\n",
      "Query: anandamide\n",
      "   anandamide                               URI=http://purl.obolibrary.org/obo/CHEBI_2700\n",
      "   anandamide                               URI=http://purl.obolibrary.org/obo/CHEBI_2700 score=1.00\n",
      "\n",
      "Query: associated synaptic proteins\n",
      "Query: gap43\n",
      "Query: sv2c\n",
      "Query: short-chain fatty acid\n",
      "   short-chain fatty acid                   URI=http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "   short-chain fatty acid                   URI=http://purl.obolibrary.org/obo/CHEBI_26666 score=1.00\n",
      "\n",
      "Query: scfa\n",
      "Query: scfa\n",
      "Query: claudin-2\n",
      "Query: epinephrine\n",
      "   Epinephrine sulfate                      URI=http://purl.obolibrary.org/obo/CHEBI_89878 score=0.88\n",
      "\n",
      "Query: norepinephrine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_90022\n",
      "\n",
      "Query: helicobacter pylori-related inflammatory mediators\n",
      "Query: defensins\n",
      "Query: vitamin d\n",
      "   vitamin D                                URI=http://purl.obolibrary.org/obo/CHEBI_27300\n",
      "   vitamin A                                URI=http://purl.obolibrary.org/obo/CHEBI_12777 score=1.00\n",
      "\n",
      "Query: microbial tryptophan\n",
      "Query: wheat gluten\n",
      "Query: bovine milk casein\n",
      "Query: food antigens\n",
      "   food                                     URI=http://purl.obolibrary.org/obo/CHEBI_33290 score=1.00\n",
      "\n",
      "Query: complement c1q\n",
      "Query: serum metabolome\n",
      "Query: gamma-aminobutyric acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16865\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: scfas\n",
      "Query: scfa\n",
      "Query: scfas\n",
      "Query: tight junction proteins\n",
      "Query: zo-1\n",
      "Query: claudin\n",
      "Query: occludin\n",
      "Query: inflammatory cytokines\n",
      "Query: il-1\n",
      "Query: il-6\n",
      "Query: butyrate\n",
      "   butyrate                                 URI=http://purl.obolibrary.org/obo/CHEBI_17968\n",
      "   butyrate                                 URI=http://purl.obolibrary.org/obo/CHEBI_17968 score=1.00\n",
      "\n",
      "Query: trichloroethylene\n",
      "Query: pesticides\n",
      "Query: metals\n",
      "Query: organic solvents\n",
      "   organic acid                             URI=http://purl.obolibrary.org/obo/CHEBI_64709 score=0.93\n",
      "\n",
      "Query: trichloroethylene\n",
      "Query: tce\n",
      "Query: tce\n",
      "Query: tce\n",
      "Query: tce\n",
      "Query: short-chain fatty acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: lactic-acid\n",
      "   (S)-lactic acid                          URI=http://purl.obolibrary.org/obo/CHEBI_422 score=1.00\n",
      "\n",
      "Query: il-1\n",
      "Query: il-2\n",
      "Query: il-4\n",
      "Query: il-6\n",
      "Query: il-13\n",
      "Query: il-18\n",
      "Query: gm-csf\n",
      "Query: ifn\n",
      "Query: tnf\n",
      "Query: ifn\n",
      "Query: tnf\n",
      "Query: tnf\n",
      "Query: ifn\n",
      "Query: tnf\n",
      "Query: ifn\n",
      "Query: exopolysaccharide\n",
      "   exopolysaccharide                        URI=http://purl.obolibrary.org/obo/CHEBI_72813\n",
      "   exopolysaccharide                        URI=http://purl.obolibrary.org/obo/CHEBI_72813 score=1.00\n",
      "\n",
      "Query: exopolysaccharide\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_72813\n",
      "\n",
      "Query: eps\n",
      "Query: eps\n",
      "Query: eps\n",
      "Query: c-eps\n",
      "Query: c-eps\n",
      "Query: eps\n",
      "Query: hydrogen sulfide\n",
      "   hydrogen sulfide                         URI=http://purl.obolibrary.org/obo/CHEBI_16136\n",
      "   hydrogen sulfide                         URI=http://purl.obolibrary.org/obo/CHEBI_16136 score=1.00\n",
      "\n",
      "Query: hydrogen sulfide\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16136\n",
      "\n",
      "Query: h2s\n",
      "Query: h2s\n",
      "Query: h2s\n",
      "Query: h2s\n",
      "Query: cytochrome c protein\n",
      "   cytochrome f                             URI=http://purl.obolibrary.org/obo/CHEBI_38556 score=0.77\n",
      "\n",
      "Query: reactive oxygen species\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26523\n",
      "\n",
      "Query: alpha-synuclein oligomers\n",
      "   N(alpha),N(alpha),N(alpha)-trimethyl-L-histidine URI=http://purl.obolibrary.org/obo/CHEBI_15781 score=0.80\n",
      "\n",
      "Query: alpha-synuclein protein\n",
      "   Protein C                                URI=http://purl.obolibrary.org/obo/CHEBI_82131 score=0.88\n",
      "\n",
      "Query: h2s\n",
      "Query: blood erythrocytes and lymphocytes\n",
      "Query: gut microbiota-derived metabolites\n",
      "Query: trp\n",
      "   Trp-P-1                                  URI=http://purl.obolibrary.org/obo/CHEBI_82376 score=1.00\n",
      "\n",
      "Query: k01817\n",
      "Query: k11358\n",
      "Query: k01626\n",
      "Query: k01667\n",
      "Query: k01626\n",
      "Query: fe\n",
      "Query: rhepo\n",
      "Query: fe\n",
      "Query: rhepo\n",
      "Query: fe\n",
      "Query: rhepo\n",
      "Query: rhepo\n",
      "Query: fe\n",
      "Query: serotonin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_350546\n",
      "\n",
      "Query: 5-ht\n",
      "   HT-2 toxin                               URI=http://purl.obolibrary.org/obo/CHEBI_138861 score=0.80\n",
      "\n",
      "Query: 5-ht receptor\n",
      "Query: long-chain fatty acids\n",
      "   long-chain fatty acid                    URI=http://purl.obolibrary.org/obo/CHEBI_15904 score=0.80\n",
      "\n",
      "Query: bile acids\n",
      "   bile acids                               URI=http://purl.obolibrary.org/obo/CHEBI_138366\n",
      "   bile acids                               URI=http://purl.obolibrary.org/obo/CHEBI_138366 score=1.00\n",
      "\n",
      "Query: resistant starch\n",
      "   starch                                   URI=http://purl.obolibrary.org/obo/CHEBI_28017 score=1.00\n",
      "\n",
      "Query: -glucuronidase\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: bile acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138366\n",
      "\n",
      "Query: dna methylation\n",
      "   DNA 5'-phosphate                         URI=http://purl.obolibrary.org/obo/CHEBI_4294 score=0.85\n",
      "\n",
      "Query: enteric glial cells network\n",
      "Query: intestinal gut enteric neuropeptides\n",
      "Query: vagal and spinal afferent neurons\n",
      "Query: gut hormones\n",
      "Query: cytokines\n",
      "Query: microbially-derived neurotransmitters\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: dietary microbial metabolites\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: indoles\n",
      "   indoles                                  URI=http://purl.obolibrary.org/obo/CHEBI_24828\n",
      "   indoles                                  URI=http://purl.obolibrary.org/obo/CHEBI_24828 score=1.00\n",
      "\n",
      "Query: bile acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138366\n",
      "\n",
      "Query: choline metabolites\n",
      "   choline                                  URI=http://purl.obolibrary.org/obo/CHEBI_15354 score=1.00\n",
      "\n",
      "Query: lactate\n",
      "   lactate                                  URI=http://purl.obolibrary.org/obo/CHEBI_24996\n",
      "   lactate                                  URI=http://purl.obolibrary.org/obo/CHEBI_24996 score=1.00\n",
      "\n",
      "Query: vitamins\n",
      "   D3 vitamins                              URI=http://purl.obolibrary.org/obo/CHEBI_73558 score=0.85\n",
      "\n",
      "Query: atp\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_57299\n",
      "\n",
      "Query: adp\n",
      "   ADP                                      URI=http://purl.obolibrary.org/obo/CHEBI_16761\n",
      "   ADP(3-)                                  URI=http://purl.obolibrary.org/obo/CHEBI_456216 score=1.00\n",
      "\n",
      "Query: krebs cycle intermediates\n",
      "Query: oxygen\n",
      "   oxygen-16(6+)                            URI=http://purl.obolibrary.org/obo/CHEBI_149692 score=0.90\n",
      "\n",
      "Query: reactive oxygen species\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26523\n",
      "\n",
      "Query: ros\n",
      "Query: n-acylphosphatidyl ethanolamines\n",
      "   ethanolamines                            URI=http://purl.obolibrary.org/obo/CHEBI_23981 score=1.00\n",
      "\n",
      "Query: napes\n",
      "Query: circulating lipids\n",
      "Query: n-acyl-phosphatidylethanolamines\n",
      "   N-acyl-L-amino acid                      URI=http://purl.obolibrary.org/obo/CHEBI_21644 score=0.79\n",
      "\n",
      "Query: napes\n",
      "Query: napes\n",
      "Query: napes\n",
      "Query: napes\n",
      "Query: napes\n",
      "Query: lipids\n",
      "Query: pro-inflammatory cytokines\n",
      "Query: tight junction proteins\n",
      "Query: porphyra haitanensis polysaccharides\n",
      "Query: blood lipid\n",
      "Query: porphyra haitanensis polysaccharide\n",
      "Query: php\n",
      "Query: blood lipids\n",
      "Query: php\n",
      "Query: serum total cholesterol\n",
      "Query: triglycerides\n",
      "Query: low-density lipoprotein cholesterol\n",
      "   low-density lipoprotein cholesterol      URI=http://purl.obolibrary.org/obo/CHEBI_47774\n",
      "   low-density lipoprotein cholesterol      URI=http://purl.obolibrary.org/obo/CHEBI_47774 score=1.00\n",
      "\n",
      "Query: php\n",
      "Query: butyric acid\n",
      "   butyric acid                             URI=http://purl.obolibrary.org/obo/CHEBI_30772\n",
      "   butyric acid                             URI=http://purl.obolibrary.org/obo/CHEBI_30772 score=1.00\n",
      "\n",
      "Query: high-dose php\n",
      "Query: hphp\n",
      "Query: low-dose php\n",
      "Query: lphp\n",
      "Query: phosphatidylcholine metabolites\n",
      "   phosphatidylcholine(1+)                  URI=http://purl.obolibrary.org/obo/CHEBI_49183 score=1.00\n",
      "\n",
      "Query: hphp\n",
      "Query: triglyceride metabolites\n",
      "   triglyceride                             URI=http://purl.obolibrary.org/obo/CHEBI_17855 score=1.00\n",
      "\n",
      "Query: cd36\n",
      "Query: acacb\n",
      "Query: php\n",
      "Query: intestinal metabolites\n",
      "Query: microbial metabolites\n",
      "Query: microflora metabolites\n",
      "Query: microbial metabolites\n",
      "Query: 3-dehydrocarnitine\n",
      "   3-dehydrocarnitine                       URI=http://purl.obolibrary.org/obo/CHEBI_57885\n",
      "   3-dehydrocarnitine                       URI=http://purl.obolibrary.org/obo/CHEBI_57885 score=1.00\n",
      "\n",
      "Query: epiandrosterone sulfate\n",
      "   epiandrosterone sulfate                  URI=http://purl.obolibrary.org/obo/CHEBI_83040\n",
      "   epiandrosterone sulfate(1-)              URI=http://purl.obolibrary.org/obo/CHEBI_133729 score=1.00\n",
      "\n",
      "Query: leucine\n",
      "   leucine                                  URI=http://purl.obolibrary.org/obo/CHEBI_25017\n",
      "   L-leucine                                URI=http://purl.obolibrary.org/obo/CHEBI_15603 score=1.00\n",
      "\n",
      "Query: short chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: scfas\n",
      "Query: kynurenine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_86262\n",
      "\n",
      "Query: norepinephrine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_90022\n",
      "\n",
      "Query: glutamate\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_29985\n",
      "\n",
      "Query: tryptophan\n",
      "   tryptophan                               URI=http://purl.obolibrary.org/obo/CHEBI_27897\n",
      "   tryptophan                               URI=http://purl.obolibrary.org/obo/CHEBI_27897 score=1.00\n",
      "\n",
      "Query: dopamine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_18243\n",
      "\n",
      "Query: epinephrine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_89878\n",
      "\n",
      "Query: -aminobutyric acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_229980\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: scfas\n",
      "Query: scfa\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: gaba\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_165543\n",
      "\n",
      "Query: cholesterol\n",
      "   cholesterol                              URI=http://purl.obolibrary.org/obo/CHEBI_16113\n",
      "   cholesterol                              URI=http://purl.obolibrary.org/obo/CHEBI_16113 score=1.00\n",
      "\n",
      "Query: 5-ht\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138861\n",
      "\n",
      "Query: 5-hydroxytryptamine\n",
      "   4-hydroxytryptamine                      URI=http://purl.obolibrary.org/obo/CHEBI_139217 score=1.00\n",
      "\n",
      "Query: 5-ht\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138861\n",
      "\n",
      "Query: ldl-c\n",
      "Query: tc\n",
      "   5'-d[TC]-3'                              URI=http://purl.obolibrary.org/obo/CHEBI_75157 score=1.00\n",
      "\n",
      "Query: tg\n",
      "   TG(8:0/8:0/10:0)                         URI=http://purl.obolibrary.org/obo/CHEBI_180791 score=0.84\n",
      "\n",
      "Query: il-1\n",
      "Query: mcp-1\n",
      "Query: il-17a\n",
      "Query: mcp-1\n",
      "Query: 5-ht\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138861\n",
      "\n",
      "Query: mcp-1\n",
      "Query: il-17a\n",
      "Query: 5-ht\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138861\n",
      "\n",
      "Query: nutraceuticals\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: trimethylamine n-oxide\n",
      "   trimethylamine N-oxide                   URI=http://purl.obolibrary.org/obo/CHEBI_15724\n",
      "   trimethylamine N-oxide                   URI=http://purl.obolibrary.org/obo/CHEBI_15724 score=1.00\n",
      "\n",
      "Query: gut microbiota-associated epitopes\n",
      "Query: potentially immunogenic epitopes\n",
      "Query: gut microbiota-associated epitopes  mes\n",
      "Query: potentially immunogenic epitopes\n",
      "Query: candidate mes\n",
      "   MES                                      URI=http://purl.obolibrary.org/obo/CHEBI_39010 score=1.00\n",
      "\n",
      "Query: mes\n",
      "   MES                                      URI=http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "   MES                                      URI=http://purl.obolibrary.org/obo/CHEBI_39010 score=1.00\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: gut iga level\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: peptides\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: gap junction alpha-1\n",
      "Query: gja1\n",
      "Query: paired box protein pax-3\n",
      "Query: pax3\n",
      "   PAX3'                                    URI=http://purl.obolibrary.org/obo/CHEBI_209474 score=1.00\n",
      "\n",
      "Query: eyes absent homolog 1 isoform 4\n",
      "Query: eya1\n",
      "Query: me\n",
      "Query: listeriolysin o peptide\n",
      "   peptide                                  URI=http://purl.obolibrary.org/obo/CHEBI_16670 score=1.00\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: gut iga levels\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: mes\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_39010\n",
      "\n",
      "Query: central monoamine neurotransmitters\n",
      "Query: amino acid transmitters\n",
      "   amino acid                               URI=http://purl.obolibrary.org/obo/CHEBI_33709 score=1.00\n",
      "\n",
      "Query: monosaccharide glucose\n",
      "   monosaccharide                           URI=http://purl.obolibrary.org/obo/CHEBI_35381 score=0.80\n",
      "\n",
      "Query: structure -4--glcp- 1\n",
      "   Structure II                             URI=http://purl.obolibrary.org/obo/CHEBI_224481 score=0.76\n",
      "\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: lipopolysaccharides\n",
      "Query: hydrogen sulfide\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16136\n",
      "\n",
      "Query: glutamate\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_29985\n",
      "\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: butyric\n",
      "   butyric acid                             URI=http://purl.obolibrary.org/obo/CHEBI_30772 score=0.95\n",
      "\n",
      "Query: propionic acids\n",
      "Query: aspartic acid\n",
      "   aspartic acid                            URI=http://purl.obolibrary.org/obo/CHEBI_22660\n",
      "   aspartic acid                            URI=http://purl.obolibrary.org/obo/CHEBI_22660 score=1.00\n",
      "\n",
      "Query: glutamic acid\n",
      "   glutamic acid                            URI=http://purl.obolibrary.org/obo/CHEBI_18237\n",
      "   D-glutamic acid                          URI=http://purl.obolibrary.org/obo/CHEBI_15966 score=1.00\n",
      "\n",
      "Query: tryptophan\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_27897\n",
      "\n",
      "Query: glutamic acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_15966\n",
      "\n",
      "Query: gamma-aminobutyric acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16865\n",
      "\n",
      "Query: tryptophan\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_27897\n",
      "\n",
      "Query: serotonin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_350546\n",
      "\n",
      "Query: kynurenic acid\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_18344\n",
      "\n",
      "Query: tryptamine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16765\n",
      "\n",
      "Query: 5-hydroxytryptamine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_139217\n",
      "\n",
      "Query: 5-ht\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_138861\n",
      "\n",
      "Query: brain-derived neurotrophic factor\n",
      "Query: bdnf\n",
      "Query: toll-like receptors\n",
      "   Toll-like receptor 2 agonist             URI=http://purl.obolibrary.org/obo/CHEBI_77161 score=0.81\n",
      "\n",
      "Query: toll-like receptors\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: toll-like receptor 2\n",
      "   Toll-like receptor 2 agonist             URI=http://purl.obolibrary.org/obo/CHEBI_77161 score=0.91\n",
      "\n",
      "Query: tlr2\n",
      "Query: toll-like receptor 4\n",
      "   Toll-like receptor 2 agonist             URI=http://purl.obolibrary.org/obo/CHEBI_77161 score=0.91\n",
      "\n",
      "Query: tlr4\n",
      "Query: toll-like receptor 2\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: toll-like receptor 4\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: toll-like receptor 2\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: toll-like receptor 4\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: toll-like receptor 2\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: toll-like receptor 4\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77161\n",
      "\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: macro- and micro-nutrients\n",
      "Query: micronutrients\n",
      "Query: omega-3 fatty acids\n",
      "Query: bioactive polyphenols\n",
      "Query: carbohydrates\n",
      "Query: carbohydrates\n",
      "Query: short-chain fatty acid butyrate\n",
      "   short-chain fatty acid                   URI=http://purl.obolibrary.org/obo/CHEBI_26666 score=0.87\n",
      "\n",
      "Query: microbial short-chain fatty acids\n",
      "Query: scfas\n",
      "Query: scfas\n",
      "Query: branched-chain scfas\n",
      "   branched-chain amino acid                URI=http://purl.obolibrary.org/obo/CHEBI_22918 score=0.92\n",
      "\n",
      "Query: cecal acetate\n",
      "   acetate                                  URI=http://purl.obolibrary.org/obo/CHEBI_30089 score=1.00\n",
      "\n",
      "Query: scfas\n",
      "Query: microbial metabolites\n",
      "Query: serotonin\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_350546\n",
      "\n",
      "Query: kynurenine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_86262\n",
      "\n",
      "Query: tryptamine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_16765\n",
      "\n",
      "Query: indole\n",
      "   indole                                   URI=http://purl.obolibrary.org/obo/CHEBI_35581\n",
      "   indole                                   URI=http://purl.obolibrary.org/obo/CHEBI_35581 score=1.00\n",
      "\n",
      "Query: hba1c\n",
      "Query: hba1c\n",
      "Query: hba1c\n",
      "Query: lipids\n",
      "Query: hunger hormones\n",
      "Query: gut hormones\n",
      "Query: hba1c\n",
      "Query: fertilizers\n",
      "Query: pesticides\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: short-chain fatty acids\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_26666\n",
      "\n",
      "Query: scfas\n",
      "Query: metabolite scfas\n",
      "   metabolite                               URI=http://purl.obolibrary.org/obo/CHEBI_25212 score=1.00\n",
      "\n",
      "Query: scfas\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import os\n",
    "import unicodedata\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "CHEBI_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "CHEMICAL_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Chemical\")\n",
    "CHEMICAL_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Chemical\")\n",
    "\n",
    "created[\"Chemical Entity\"] = URIRef(f\"{CHEBI_BASE}CHEBI_24431\")\n",
    "g.add((created[\"Chemical Entity\"], RDF.type, CHEMICAL_CLASS))\n",
    "g.add((created[\"Chemical Entity\"], RDF.type, SKOS.Concept))\n",
    "g.add((created[\"Chemical Entity\"], RDFS.label, Literal(\"Chemical Entity\", datatype=XSD.string)))\n",
    "g.add((created[\"Chemical Entity\"], SKOS.inScheme, CHEMICAL_CONCEPT_SCHEME))\n",
    "\n",
    "\n",
    "def load_chebi_labels(path):\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        next(fh) \n",
    "        for ln in fh:\n",
    "            uri, label = ln.rstrip(\"\\n\").split(\"\\t\", 1)\n",
    "            rows.append((label, uri))\n",
    "    return rows\n",
    "\n",
    "CHEBI_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\chebi_labels.txt\"\n",
    "chebi_rows = load_chebi_labels(CHEBI_LABELS_FILE)\n",
    "\n",
    "exact_ix = defaultdict(list)\n",
    "for lbl, uri in chebi_rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "labels_only = [lbl for lbl, _ in chebi_rows]\n",
    "vec = TfidfVectorizer(stop_words=\"english\")\n",
    "mat = vec.fit_transform(labels_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v  = vec.transform([term])\n",
    "    sc = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = chebi_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "greek_map = {\n",
    "    '': 'alpha',  '': 'alpha',\n",
    "    '': 'beta',   '': 'beta',\n",
    "    '': 'gamma',  '': 'gamma',\n",
    "    '': 'delta',  '': 'delta',\n",
    "    '': 'epsilon','': 'epsilon',\n",
    "    '': 'zeta',   '': 'zeta',\n",
    "    '': 'eta',    '': 'eta',\n",
    "    '': 'theta',  '': 'theta',\n",
    "    '': 'iota',   '': 'iota',\n",
    "    '': 'kappa',  '': 'kappa',\n",
    "    '': 'lambda', '': 'lambda',\n",
    "    '': 'mu',     '': 'mu',\n",
    "    '': 'nu',     '': 'nu',\n",
    "    '': 'xi',     '': 'xi',\n",
    "    '': 'omicron','': 'omicron',\n",
    "    '': 'pi',     '': 'pi',\n",
    "    '': 'rho',    '': 'rho',\n",
    "    '': 'sigma',  '': 'sigma',\n",
    "    '': 'tau',    '': 'tau',\n",
    "    '': 'upsilon','': 'upsilon',\n",
    "    '': 'phi',    '': 'phi',\n",
    "    '': 'chi',    '': 'chi',\n",
    "    '': 'psi',    '': 'psi',\n",
    "    '': 'omega',  '': 'omega',\n",
    "}\n",
    "\n",
    "def preprocess(term):\n",
    "    \"\"\"\n",
    "    1) replace Greek letters with names\n",
    "    2) turn underscores into spaces\n",
    "    3) strip whitespace\n",
    "    4) lowercase\n",
    "    \"\"\"\n",
    "    # 1) map Greek  name\n",
    "    for greek_char, name in greek_map.items():\n",
    "        if greek_char in term:\n",
    "            term = term.replace(greek_char, name)\n",
    "    # 2) replace underscores\n",
    "    term = term.replace('_', ' ')\n",
    "    # 3) strip\n",
    "    term = term.strip()\n",
    "    # 4) lowercase\n",
    "    return term.lower()\n",
    "    \n",
    "for paper_id, paper_data in data.items():\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"chemical\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            ex = exact_ix.get(term, [])\n",
    "            if ex:\n",
    "                for lbl, uri in ex:\n",
    "                    print(f\"   {lbl:40s} URI={uri}\")\n",
    "                    entity_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, CHEMICAL_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, CHEMICAL_CONCEPT_SCHEME))\n",
    "                    g.add((entity_uri, SKOS.broaderTransitive, created[\"Chemical Entity\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_chemical_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            if cos:\n",
    "                for lbl, uri, score in cos[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, CHEMICAL_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, CHEMICAL_CONCEPT_SCHEME))\n",
    "                    g.add((name_uri, SKOS.broaderTransitive, created[\"Chemical Entity\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_chemical_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_chemical_{term_raw}\", datatype=XSD.string)))\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                g.add((created[\"Chemical Entity\"], GUTPROP.containedIn, mention_uri))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"chemical\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_chemical_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "40e497af-7ea7-423d-8376-95b1c8785f4d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: High-fiber diet\n",
      "   high fiber food                          ID=FOODON_03510048 depth=0  score=0.76 (TAX2 cosine)\n",
      "   bread (high fiber)                       ID=FOODON_03303712 depth=0  score=0.71 (TAX2 cosine)\n",
      "   breakfast cereal (high fiber)            ID=FOODON_03310642 depth=0  score=0.61 (TAX2 cosine)\n",
      "   bread (high fiber, reduced-calorie)      ID=FOODON_03302702 depth=0  score=0.55 (TAX2 cosine)\n",
      "   wheat fiber                              ID=FOODON_03307422 depth=0  score=0.51 (TAX2 cosine)\n",
      "\n",
      "Query: Low-Caloric Diet\n",
      "   Low-fat Milk                             ID=NCIT_C178186    depth=8  score=0.66 (TAX1 cosine)\n",
      "\n",
      "Query: Origanum vulgare\n",
      "  ! no matches in either taxonomy\n",
      "\n",
      "Query: Thymus vulgaris\n",
      "   beef thymus                              ID=FOODON_02020917 depth=0  score=0.54 (TAX2 cosine)\n",
      "   animal thymus                            ID=FOODON_02010051 depth=0  score=0.54 (TAX2 cosine)\n",
      "   pork thymus                              ID=FOODON_02021671 depth=0  score=0.52 (TAX2 cosine)\n",
      "   lamb thymus                              ID=FOODON_02021155 depth=0  score=0.51 (TAX2 cosine)\n",
      "   beef thymus (raw)                        ID=FOODON_02020924 depth=0  score=0.51 (TAX2 cosine)\n",
      "\n",
      "Query: Wheat germ\n",
      "   Wheat Gluten                             ID=NCIT_C84242     depth=2  score=0.71 (TAX1 cosine)\n",
      "\n",
      "Query: animal to vegetal foods\n",
      "   CIAA other foods                         ID=FOODON_03400467 depth=0  score=0.40 (TAX2 cosine)\n",
      "   animal back                              ID=FOODON_02010129 depth=0  score=0.35 (TAX2 cosine)\n",
      "   animal                                   ID=FOODON_00003004 depth=0  score=0.35 (TAX2 cosine)\n",
      "   45150 - pollen (residues of vegetal pollens) (efsa foodex2) ID=FOODON_03544515 depth=0  score=0.33 (TAX2 cosine)\n",
      "   CCFAC composite foods                    ID=FOODON_03400641 depth=0  score=0.31 (TAX2 cosine)\n",
      "\n",
      "Query: dietary fibers\n",
      "   Dietary Fish                             ID=NCIT_C72065     depth=4  score=0.71 (TAX1 cosine)\n",
      "\n",
      "Query: eggs\n",
      "   50132500 - eggs/eggs substitutes (gs1 gpc) ID=FOODON_03401038 depth=0  score=0.72 (TAX2 cosine)\n",
      "   obsolete: 10000582 - eggs/eggs substitutes (frozen) (gs1 gpc) ID=FOODON_03401039 depth=0  score=0.70 (TAX2 cosine)\n",
      "   obsolete: 10000031 - eggs/eggs substitutes (perishable) (gs1 gpc) ID=FOODON_03401040 depth=0  score=0.68 (TAX2 cosine)\n",
      "   obsolete: 10000032 - eggs/eggs substitutes (shelf stable) (gs1 gpc) ID=FOODON_03401041 depth=0  score=0.65 (TAX2 cosine)\n",
      "   29570 - eggs from other eggs producer birds (efsa foodex2) ID=FOODON_03542957 depth=0  score=0.65 (TAX2 cosine)\n",
      "\n",
      "Query: high-fiber diet\n",
      "   high fiber food                          ID=FOODON_03510048 depth=0  score=0.76 (TAX2 cosine)\n",
      "   bread (high fiber)                       ID=FOODON_03303712 depth=0  score=0.71 (TAX2 cosine)\n",
      "   breakfast cereal (high fiber)            ID=FOODON_03310642 depth=0  score=0.61 (TAX2 cosine)\n",
      "   bread (high fiber, reduced-calorie)      ID=FOODON_03302702 depth=0  score=0.55 (TAX2 cosine)\n",
      "   wheat fiber                              ID=FOODON_03307422 depth=0  score=0.51 (TAX2 cosine)\n",
      "\n",
      "Query: meat\n",
      "   Meat                                     ID=NCIT_C126482    depth=2  score=1.00 (TAX1 exact)\n",
      "\n",
      "Query: oregano\n",
      "   obsolete: oregano plant                  ID=FOODON_03411542 depth=0  score=0.85 (TAX2 cosine)\n",
      "   oregano food product                     ID=FOODON_00002073 depth=0  score=0.84 (TAX2 cosine)\n",
      "   oregano leaf                             ID=FOODON_03301557 depth=0  score=0.84 (TAX2 cosine)\n",
      "   oregano (ground)                         ID=FOODON_03301482 depth=0  score=0.81 (TAX2 cosine)\n",
      "   oregano (whole or pieces)                ID=FOODON_03000226 depth=0  score=0.80 (TAX2 cosine)\n",
      "\n",
      "Query: pork\n",
      "   Pork                                     ID=NCIT_C71958     depth=4  score=1.00 (TAX1 exact)\n",
      "\n",
      "Query: poultry meat\n",
      "   Meat                                     ID=NCIT_C126482    depth=2  score=0.73 (TAX1 cosine)\n",
      "   Poultry                                  ID=NCIT_C53475     depth=4  score=0.69 (TAX1 cosine)\n",
      "   Processed Meat                           ID=NCIT_C126483    depth=4  score=0.49 (TAX1 cosine)\n",
      "   Egg, Poultry                             ID=NCIT_C72064     depth=2  score=0.49 (TAX1 cosine)\n",
      "   Turkey, Poultry                          ID=NCIT_C73855     depth=6  score=0.46 (TAX1 cosine)\n",
      "\n",
      "Query: thyme\n",
      "   thyme plant                              ID=FOODON_03411199 depth=0  score=0.90 (TAX2 cosine)\n",
      "   thyme (dried)                            ID=FOODON_03301215 depth=0  score=0.85 (TAX2 cosine)\n",
      "   thyme food product                       ID=FOODON_00002312 depth=0  score=0.83 (TAX2 cosine)\n",
      "   thyme leaf                               ID=FOODON_03301562 depth=0  score=0.83 (TAX2 cosine)\n",
      "   thyme (ground)                           ID=FOODON_03301489 depth=0  score=0.80 (TAX2 cosine)\n",
      "\n",
      "Query: vegetables\n",
      "   mixed vegetables                         ID=FOODON_00002683 depth=0  score=0.69 (TAX2 cosine)\n",
      "   mixed vegetables (frozen)                ID=FOODON_03310847 depth=0  score=0.61 (TAX2 cosine)\n",
      "   mixed vegetables (canned)                ID=FOODON_03303882 depth=0  score=0.61 (TAX2 cosine)\n",
      "   mixed vegetables or garden vegetables (strained or junior food in jar) ID=FOODON_03311136 depth=0  score=0.55 (TAX2 cosine)\n",
      "   chicken and noodles with vegetables      ID=FOODON_03304772 depth=0  score=0.54 (TAX2 cosine)\n",
      "\n",
      "Query: wheat germs\n",
      "   Wheat Gluten                             ID=NCIT_C84242     depth=2  score=0.71 (TAX1 cosine)\n",
      "\n",
      "Query: whole grain cereals\n",
      "   Whole Grain                              ID=NCIT_C178197    depth=2  score=1.00 (TAX1 cosine)\n",
      "   Whole Grain Bread                        ID=NCIT_C178204    depth=6  score=0.74 (TAX1 cosine)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re, json, numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "# ---------------------------------------------------------------------------\n",
    "# 0) FILES\n",
    "# ---------------------------------------------------------------------------\n",
    "TAX1_FILE = \"food_tree.txt\"   # your primary taxonomy (FoodOn)\n",
    "TAX2_FILE = \"foodon_terms.txt\"      # your secondary taxonomy (NCIT)\n",
    "\n",
    "# ---------------------------------------------------------------------------\n",
    "# 1) LOADING\n",
    "# ---------------------------------------------------------------------------\n",
    "def load_taxonomy_tree(path):\n",
    "    row_re = re.compile(r\"^\\s*(.*?)\\s+\\[([^\\]]+)\\]\\s*$\")\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        for ln in fh:\n",
    "            m = row_re.match(ln)\n",
    "            if m:\n",
    "                label, tid = m.groups()\n",
    "                depth = len(ln) - len(ln.lstrip())\n",
    "                rows.append((label, tid, depth))\n",
    "    return rows\n",
    "\n",
    "# load both sets of rows\n",
    "rows1 = load_taxonomy_tree(TAX1_FILE)\n",
    "rows2 = load_taxonomy_tree(TAX2_FILE)\n",
    "\n",
    "# ---------------------------------------------------------------------------\n",
    "# 2) EXACTMATCH INDEXES\n",
    "# ---------------------------------------------------------------------------\n",
    "exact1 = defaultdict(list)\n",
    "for lbl, tid, depth in rows1:\n",
    "    exact1[lbl.lower()].append((lbl, tid, depth))\n",
    "\n",
    "exact2 = defaultdict(list)\n",
    "for lbl, tid, depth in rows2:\n",
    "    exact2[lbl.lower()].append((lbl, tid, depth))\n",
    "\n",
    "# ---------------------------------------------------------------------------\n",
    "# 3) TFIDF SETUP\n",
    "# ---------------------------------------------------------------------------\n",
    "labels1 = [lbl for lbl,_,_ in rows1]\n",
    "vec1    = TfidfVectorizer(stop_words=\"english\")\n",
    "mat1    = vec1.fit_transform(labels1)\n",
    "\n",
    "labels2 = [lbl for lbl,_,_ in rows2]\n",
    "vec2    = TfidfVectorizer(stop_words=\"english\")\n",
    "mat2    = vec2.fit_transform(labels2)\n",
    "\n",
    "def top_cosine(rows, vec, mat, labels, term, k=5, thr=0.25):\n",
    "    v   = vec.transform([term])\n",
    "    sc  = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, tid, depth = rows[i]\n",
    "        out.append((lbl, tid, depth, sc[i]))\n",
    "    return out\n",
    "\n",
    "# ---------------------------------------------------------------------------\n",
    "# 4) QUERY LOADING (as before)\n",
    "# ---------------------------------------------------------------------------\n",
    "def strip_html(text):\n",
    "    return re.sub(r\"<[^>]+>\", \"\", text)\n",
    "\n",
    "with open(r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\data\\train_platinum\\train_platinum.json\",\n",
    "          encoding=\"utf-8\") as fh:\n",
    "    js = json.load(fh)\n",
    "\n",
    "queries = {\n",
    "    strip_html(e[\"text_span\"]).strip()\n",
    "    for rec in js.values()\n",
    "    for e in rec.get(\"entities\", [])\n",
    "    if e.get(\"label\",\"\").lower() == \"food\"\n",
    "}\n",
    "\n",
    "def preprocess(term):\n",
    "    return term.replace('_', ' ').strip()\n",
    "\n",
    "# ---------------------------------------------------------------------------\n",
    "# 5) MATCHING LOOP WITH FALLBACK TO TAX2\n",
    "# ---------------------------------------------------------------------------\n",
    "for term_raw in sorted(queries):\n",
    "    term = preprocess(term_raw)\n",
    "    print(f\"Query: {term_raw}\")\n",
    "\n",
    "    # 5-a exact on TAX1\n",
    "    ex1 = exact1.get(term.lower(), [])\n",
    "    if ex1:\n",
    "        for l,t,d in ex1:\n",
    "            print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score=1.00 (TAX1 exact)\")\n",
    "        print()\n",
    "        continue\n",
    "\n",
    "    # 5-b cosine on TAX1\n",
    "    cos1 = top_cosine(rows1, vec1, mat1, labels1, term, k=5, thr=0.25)\n",
    "    if cos1:\n",
    "        for l,t,d,s in cos1:\n",
    "            print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score={s:.2f} (TAX1 cosine)\")\n",
    "        print()\n",
    "        continue\n",
    "\n",
    "    # 5-c FALLBACK: cosine on TAX2\n",
    "    cos2 = top_cosine(rows2, vec2, mat2, labels2, term, k=5, thr=0.25)\n",
    "    if cos2:\n",
    "        for l,t,d,s in cos2:\n",
    "            print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score={s:.2f} (TAX2 cosine)\")\n",
    "    else:\n",
    "        print(\"  ! no matches in either taxonomy\")\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d89c5023-59a3-45e3-891c-d823a119cb41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: vegetables\n",
      "Query: whole grain cereals\n",
      "   Whole Grain                              ID=NCIT_C178197    depth=2  score=1.00 (TAX1 cosine)\n",
      "\n",
      "Query: high-fiber diet\n",
      "   high fiber food                          ID=FOODON_03510048 depth=0  score=0.76 (TAX2 cosine)\n",
      "Query: high-fiber diet\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/FOODON_03510048\n",
      "\n",
      "Query: high-fiber diet\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/FOODON_03510048\n",
      "\n",
      "Query: high-fiber diet\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/FOODON_03510048\n",
      "\n",
      "Query: high-fiber diet\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/FOODON_03510048\n",
      "\n",
      "Query: wheat germ\n",
      "   wheat germ                               ID=FOODON_03301595 depth=0  score=1.00 (TAX2 cosine)\n",
      "Query: wheat germs\n",
      "Query: dietary fibers\n",
      "   dietary claim                            ID=FOODON_03510023 depth=0  score=0.77 (TAX2 cosine)\n",
      "Query: thyme\n",
      "   thyme plant                              ID=FOODON_03411199 depth=0  score=0.90 (TAX2 cosine)\n",
      "Query: thymus vulgaris\n",
      "Query: oregano\n",
      "   obsolete: oregano plant                  ID=FOODON_03411542 depth=0  score=0.85 (TAX2 cosine)\n",
      "Query: origanum vulgare\n",
      "Query: low-caloric diet\n",
      "Query: animal to vegetal foods\n",
      "Query: poultry meat\n",
      "   poultry meat product                     ID=FOODON_03315174 depth=0  score=0.89 (TAX2 cosine)\n",
      "Query: meat\n",
      "   Meat                                     ID=NCIT_C126482    depth=2  score=1.00 (TAX1 exact)\n",
      "   Meat                                     ID=NCIT_C126482    depth=2  score=1.00 (TAX1 cosine)\n",
      "\n",
      "Query: eggs\n",
      "Query: pork\n",
      "   Pork                                     ID=NCIT_C71958     depth=4  score=1.00 (TAX1 exact)\n",
      "   Pork                                     ID=NCIT_C71958     depth=4  score=1.00 (TAX1 cosine)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "import os\n",
    "import re\n",
    "import unicodedata\n",
    "import json\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "FOODON_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "NCIT_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "FOOD_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Food\")\n",
    "FOOD_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Food\")\n",
    "\n",
    "created[\"Food\"] = URIRef(f\"{NCIT_BASE}NCIT_C62695\")\n",
    "g.add((created[\"Food\"], RDF.type, FOOD_CLASS))\n",
    "g.add((created[\"Food\"], RDF.type, SKOS.Concept))\n",
    "g.add((created[\"Food\"], RDFS.label, Literal(\"Food\", datatype=XSD.string)))\n",
    "g.add((created[\"Food\"], SKOS.inScheme, FOOD_CONCEPT_SCHEME))\n",
    "\n",
    "for paper_id, paper_data in data.items():\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"food\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            ex1 = exact1.get(term.lower(), [])\n",
    "            if ex1:\n",
    "                for l,t,d in ex1:\n",
    "                    print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score=1.00 (TAX1 exact)\")\n",
    "                    entity_uri = URIRef(f\"{NCIT_BASE}{t}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    #label2uri[label_name.lower()] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, FOOD_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(l, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, FOOD_CONCEPT_SCHEME))\n",
    "                    g.add((entity_uri, SKOS.broaderTransitive, created[\"Food\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_food_{term_raw}\", datatype=XSD.string)))\n",
    "                    #mention_counter += 1\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos1 = top_cosine(rows1, vec1, mat1, labels1, term, k=5, thr=0.75)\n",
    "            if cos1:\n",
    "                for l,t,d,s in cos1[:1]:\n",
    "                    print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score={s:.2f} (TAX1 cosine)\")\n",
    "                    name_uri = URIRef(f\"{NCIT_BASE}{t}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, FOOD_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(l, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, FOOD_CONCEPT_SCHEME))\n",
    "                    g.add((name_uri, SKOS.broaderTransitive, created[\"Food\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_food_{term_raw}\", datatype=XSD.string)))\n",
    "                    #mention_counter += 1\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "                continue\n",
    "\n",
    "            cos2 = top_cosine(rows2, vec2, mat2, labels2, term, k=5, thr=0.75)\n",
    "            if cos2:\n",
    "                for l,t,d,s in cos2[:1]:\n",
    "                    print(f\"   {l:40s} ID={t:15s} depth={d:<2d} score={s:.2f} (TAX2 cosine)\")\n",
    "                    name_uri = URIRef(f\"{FOODON_BASE}{t}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    #label2uri[label_name.lower()] = name_uri\n",
    "                    g.add((name_uri, RDF.type, FOOD_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(l, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, FOOD_CONCEPT_SCHEME))\n",
    "                    g.add((name_uri, SKOS.broaderTransitive, created[\"Food\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_food_{term_raw}\", datatype=XSD.string)))\n",
    "                    #mention_counter += 1\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_food_{term_raw}\", datatype=XSD.string)))\n",
    "                #mention_counter += 1\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                g.add((created[\"Food\"], GUTPROP.containedIn, mention_uri))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"food\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_food_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "33cf623b-b431-4d99-a7c2-149d7118ce42",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: patients\n",
      "   M01.643                                  ID=D010361         depth=Patients                                 score=1.00 (TAX1 exact)\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: people\n",
      "Query: bariatric patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: humans\n",
      "   B01.050.150.900.649.313.988.400.112.400.400 ID=D006801         depth=Humans                                   score=1.00 (TAX1 exact)\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: females\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: schizophrenia patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: schizophrenia patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: schizophrenia patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: health controls\n",
      "Query: schizophrenia cohort\n",
      "Query: control cohort\n",
      "Query: healthy controls\n",
      "Query: schizophrenia patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: schizophrenia patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: stressed adults\n",
      "Query: placebo group\n",
      "Query: probiotic group\n",
      "Query: placebo group\n",
      "Query: probiotic group\n",
      "Query: probiotic group\n",
      "Query: stressed adults\n",
      "Query: parkinson s disease  pd  patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: pd patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: individuals\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: individuals\n",
      "Query: comorbid patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: ibs-d patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: depression patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: mother-infant pairs\n",
      "Query: mother\n",
      "Query: vaginally delivered infants\n",
      "Query: vaginally delivered  fully breastfed infants\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: world s population\n",
      "Query: major depressive episode patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: hcs\n",
      "Query: hcs\n",
      "Query: hcs\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: individuals\n",
      "Query: offspring\n",
      "Query: child\n",
      "   M01.060.406                              ID=D002648         depth=Child                                    score=1.00 (TAX1 exact)\n",
      "Query: women\n",
      "   M01.975                                  ID=D014930         depth=Women                                    score=1.00 (TAX1 exact)\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: child\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D002648\n",
      "\n",
      "Query: child\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D002648\n",
      "\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: children\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: children\n",
      "Query: adolescents\n",
      "Query: children\n",
      "Query: adolescents\n",
      "Query: pediatric population\n",
      "Query: child and adolescent population  0-18 years old\n",
      "Query: youth\n",
      "Query: children\n",
      "Query: asd and healthy controls\n",
      "Query: children\n",
      "Query: adolescents\n",
      "Query: child and adolescent populations\n",
      "Query: type 2 diabetes patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: participants\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: military veterans\n",
      "Query: homeless individuals\n",
      "Query: veterans\n",
      "   M01.930                                  ID=D014728         depth=Veterans                                 score=1.00 (TAX1 exact)\n",
      "Query: individuals\n",
      "Query: veterans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014728\n",
      "\n",
      "Query: veterans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014728\n",
      "\n",
      "Query: veterans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014728\n",
      "\n",
      "Query: veterans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014728\n",
      "\n",
      "Query: individual patient\n",
      "Query: individuals\n",
      "Query: persons\n",
      "   M01                                      ID=D009272         depth=Persons                                  score=1.00 (TAX1 exact)\n",
      "Query: individuals\n",
      "Query: hospitalized older adults\n",
      "Query: geriatric populations\n",
      "Query: hospitalized acutely ill older adults\n",
      "Query: geriatric patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: older adults\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: inpatients\n",
      "   M01.643.470                              ID=D007297         depth=Inpatients                               score=1.00 (TAX1 exact)\n",
      "   Inpatients                               ID=D007297         depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: females\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: taiwanese patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: gender-matched controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: population\n",
      "   N01.600                                  ID=D011153         depth=Population                               score=1.00 (TAX1 exact)\n",
      "Query: pcos patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: pcos patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: depressed women\n",
      "Query: pcos-dp\n",
      "Query: mentally healthy women\n",
      "Query: healthy control\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pcos-dp\n",
      "Query: hc  mdd  and pcos groups\n",
      "Query: pcos-dp\n",
      "Query: pcos patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pcos-dp\n",
      "Query: pcos-dp\n",
      "Query: pcos-dp\n",
      "Query: pcos patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: typically developing  td  children\n",
      "Query: children\n",
      "Query: td children\n",
      "Query: asd children\n",
      "Query: asd children\n",
      "Query: children\n",
      "Query: major depressive disorder patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: mdd patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: mdd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: hcs\n",
      "Query: mdd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: hcs\n",
      "Query: mdd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: mdd group\n",
      "Query: mdd group\n",
      "Query: mdd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: hcs\n",
      "Query: healthy adults\n",
      "Query: healthy adults\n",
      "Query: nvp-1704 group\n",
      "Query: placebo group\n",
      "Query: edf group\n",
      "Query: p group\n",
      "Query: ed and edf groups\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: individuals\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: individuals\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: reproductive-aged women\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pcos patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pcos patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pcos patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: alcohol overconsumption group\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pd and healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: insomnia patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: acute insomnia patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: chronic insomnia patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: hcs\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: hcs\n",
      "Query: hcs\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: tibetan buddhist monks\n",
      "Query: neighbouring residents\n",
      "Query: monks\n",
      "   M01.526.799.750                          ID=D065830         depth=Monks                                    score=1.00 (TAX1 exact)\n",
      "Query: control subjects\n",
      "Query: monks\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D065830\n",
      "\n",
      "Query: females\n",
      "Query: controls\n",
      "Query: females\n",
      "Query: controls\n",
      "Query: participants\n",
      "Query: participants\n",
      "Query: participants\n",
      "Query: participants aged 18-40 years\n",
      "Query: participants\n",
      "Query: people with pd\n",
      "Query: pwp\n",
      "Query: pwp\n",
      "Query: healthy controls\n",
      "Query: pwp\n",
      "Query: control groups\n",
      "   E05.318.370.074                          ID=D035061         depth=Control Groups                           score=1.00 (TAX1 exact)\n",
      "Query: pwp\n",
      "Query: pwp\n",
      "Query: pwp\n",
      "Query: pd cohort\n",
      "Query: pwp\n",
      "Query: southern hemisphere pd population\n",
      "Query: parkinson s disease patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: human subjects\n",
      "Query: individuals\n",
      "Query: healthy controls\n",
      "Query: controls\n",
      "Query: female pd patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: pd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: occidental patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: populations\n",
      "Query: pd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: non-pd controls\n",
      "Query: pd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy people\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: men\n",
      "   M01.390                                  ID=D008571         depth=Men                                      score=1.00 (TAX1 exact)\n",
      "Query: healthy controls\n",
      "Query: acute patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: remission patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: gut microbial composition of patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: healthy controls\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: unaffected first-degree relatives\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: smokers\n",
      "   M01.808                                  ID=D000074285      depth=Smokers                                  score=1.00 (TAX1 exact)\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: smokers\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000074285\n",
      "\n",
      "Query: non-smoking patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: non-smoking healthy controls\n",
      "Query: smokers\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000074285\n",
      "\n",
      "Query: unaffected relatives\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: participants\n",
      "Query: controls\n",
      "Query: children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: vvs children\n",
      "Query: matched controls\n",
      "Query: vvs cases\n",
      "Query: controls\n",
      "Query: vvs and controls\n",
      "Query: vvs children\n",
      "Query: controls\n",
      "Query: vvs patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: vvs patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: controls\n",
      "Query: patients with schizophrenia\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: males\n",
      "Query: females\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: asd children\n",
      "Query: typically developing  td  children\n",
      "Query: asd children\n",
      "Query: typically developing  td  children\n",
      "Query: children\n",
      "Query: asd children\n",
      "Query: td children\n",
      "Query: children\n",
      "Query: asd children\n",
      "Query: td children\n",
      "Query: asd children\n",
      "Query: td children\n",
      "Query: children\n",
      "Query: children\n",
      "Query: asd children\n",
      "Query: participants\n",
      "Query: criteria-acute group  patients with acute schizophrenia\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: remission group  patients with schizophrenia in remission\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: control group  healthy controls\n",
      "Query: acute group\n",
      "Query: remission and control groups\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: people\n",
      "Query: crew members\n",
      "Query: older subjects\n",
      "Query: older subjects\n",
      "Query: at-risk groups\n",
      "Query: parkinson s disease patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: mdd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: humans\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D006801\n",
      "\n",
      "Query: healthy persons\n",
      "Query: healthy persons\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: men\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D008571\n",
      "\n",
      "Query: people\n",
      "Query: people\n",
      "Query: individuals\n",
      "Query: participants\n",
      "Query: participants\n",
      "Query: participants\n",
      "Query: probiotic group\n",
      "Query: placebo group\n",
      "Query: generation over 75\n",
      "Query: younger generations\n",
      "Query: younger generation\n",
      "Query: egyptian patients\n",
      "   No-Show Patients                         ID=D000067455      depth=4  score=1.00 (Patients cosine)\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy subjects\n",
      "Query: pd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: pd patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy control\n",
      "Query: parkinson s disease patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy controls\n",
      "Query: breast cancer survivors\n",
      "Query: cancer survivors\n",
      "   M01.860.350                              ID=D000073116      depth=Cancer Survivors                         score=1.00 (TAX1 exact)\n",
      "Query: cancer survivors\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000073116\n",
      "\n",
      "Query: women\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D014930\n",
      "\n",
      "Query: participants\n",
      "Query: participants\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy population\n",
      "Query: patients\n",
      "   Reusing existing URI: https://www.ncbi.nlm.nih.gov/mesh/D000067455\n",
      "\n",
      "Query: healthy volunteers\n",
      "   M01.774.500                              ID=D064368         depth=Healthy Volunteers                       score=1.00 (TAX1 exact)\n",
      "Query: mdd group\n",
      "Query: control  c  group\n",
      "Query: mdd group\n",
      "Query: c group\n",
      "Query: mdd group\n",
      "Query: c group\n",
      "Query: mdd group\n",
      "Query: c group\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import unicodedata\n",
    "import json\n",
    "from pathlib import Path\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# Constants / Namespaces\n",
    "# -----------------------------------------------------------------------------\n",
    "NCBI_BASE             = \"http://purl.obolibrary.org/obo/\"\n",
    "MESH_BASE             = \"https://www.ncbi.nlm.nih.gov/mesh/\"\n",
    "HUMAN_CLASS           = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Human\")\n",
    "HUMAN_CONCEPT_SCHEME  = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Human\")\n",
    "\n",
    "# add a Humans concept\n",
    "#created[\"Humans\"] = URIRef(f\"{NCBI_BASE}NCBITaxon_9606\")\n",
    "#g.add((created[\"Humans\"], RDF.type, HUMAN_CLASS))\n",
    "#g.add((created[\"Humans\"], RDF.type, SKOS.Concept))\n",
    "#g.add((created[\"Humans\"], RDFS.label, Literal(\"Humans\", datatype=XSD.string)))\n",
    "#g.add((created[\"Humans\"], SKOS.inScheme, HUMAN_CONCEPT_SCHEME))\n",
    "\n",
    "def parse_mesh_descriptors(xml_path):\n",
    "    descs = []\n",
    "    tree  = ET.parse(xml_path)\n",
    "    root  = tree.getroot()\n",
    "    for dr in root.findall('DescriptorRecord'):\n",
    "        ui   = dr.findtext('DescriptorUI')\n",
    "        name = dr.findtext('DescriptorName/String')\n",
    "        tns  = [tn.text for tn in dr.findall('TreeNumberList/TreeNumber') if tn.text]\n",
    "        if ui and name:\n",
    "            descs.append({'ui':ui,'name':name,'tree_numbers':tns})\n",
    "    return descs\n",
    "\n",
    "def build_name_index(descriptors):\n",
    "    idx = defaultdict(list)\n",
    "    for d in descriptors:\n",
    "        if not d['tree_numbers']: continue\n",
    "        tn = d['tree_numbers'][0]\n",
    "        idx[d['name'].lower()].append((tn, d['ui'], d['name']))\n",
    "    return idx\n",
    "\n",
    "MESH_XML     = 'desc2025.xml'\n",
    "mesh_descs   = parse_mesh_descriptors(MESH_XML)\n",
    "mesh_index   = build_name_index(mesh_descs)\n",
    "\n",
    "mesh_items   = [(tn,ui,name) for vs in mesh_index.values() for tn,ui,name in vs]\n",
    "mesh_labels  = [name for (_,_,name) in mesh_items]\n",
    "mesh_uids    = [ui   for (_,ui,_)   in mesh_items]\n",
    "\n",
    "mesh_vec     = TfidfVectorizer(stop_words=\"english\").fit(mesh_labels)\n",
    "mesh_mat     = mesh_vec.transform(mesh_labels)\n",
    "\n",
    "def load_taxonomy_tree(path):\n",
    "    row_re = re.compile(r\"^\\s*(.*?)\\s+\\[([^\\]]+)\\]\\s*$\")\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        for ln in fh:\n",
    "            m = row_re.match(ln)\n",
    "            if m:\n",
    "                label, tid = m.groups()\n",
    "                depth      = len(ln) - len(ln.lstrip())\n",
    "                rows.append((label, tid, depth))\n",
    "    return rows\n",
    "\n",
    "TAX_FILE = \"patients_output.txt\"\n",
    "rows     = load_taxonomy_tree(TAX_FILE)\n",
    "\n",
    "exact_ix  = defaultdict(list)\n",
    "for lbl, tid, depth in rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, tid, depth))\n",
    "\n",
    "labels_only  = [r[0] for r in rows]\n",
    "vec          = TfidfVectorizer(stop_words=\"english\")\n",
    "mat          = vec.fit_transform(labels_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v   = vec.transform([term])\n",
    "    sc  = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr: break\n",
    "        lbl, tid, d = rows[i]\n",
    "        out.append((lbl, tid, d, sc[i]))\n",
    "    return out\n",
    "    \n",
    "for paper_id, paper_data in data.items():\n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"human\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            mesh_ex = mesh_index.get(term, [])\n",
    "            if mesh_ex:\n",
    "                for t,u,n in mesh_ex:\n",
    "                    print(f\"   {t:40s} ID={u:15s} depth={n:40s} score=1.00 (TAX1 exact)\")\n",
    "                    entity_uri = URIRef(f\"{MESH_BASE}{u}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, HUMAN_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(n, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, HUMAN_CONCEPT_SCHEME))\n",
    "                    #g.add((entity_uri, SKOS.broaderTransitive, created[\"Humans\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_human_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            if cos:\n",
    "                for name, ui, depth, score in cos[:1]:\n",
    "                    print(f\"   {name:40s} ID={ui:15s} depth={depth:<2d} score={score:.2f} (Patients cosine)\")\n",
    "                    name_uri = URIRef(f\"{MESH_BASE}{ui}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, HUMAN_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(name, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, HUMAN_CONCEPT_SCHEME))\n",
    "                    #g.add((name_uri, SKOS.broaderTransitive, created[\"Humans\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_human_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "                continue\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_human_{term_raw}\", datatype=XSD.string)))\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                #g.add((created[\"Humans\"], GUTPROP.containedIn, mention_uri))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"human\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "            print(cleaned_text_span)\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_human_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "146037fb-8af4-485f-8d89-1ebb042564a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: antibiotics\n",
      "   amino acid derivative antibiotics        URI=http://purl.obolibrary.org/obo/CHEBI_22476 score=0.78\n",
      "\n",
      "Query: antibiotic growth promotants\n",
      "Query: agps\n",
      "Query: agps\n",
      "Query: agps\n",
      "Query: agps\n",
      "Query: agp treatments\n",
      "Query: agps\n",
      "Query: agps\n",
      "Query: agps\n",
      "Query: non-absorbable antibiotic vancomycin\n",
      "Query: olanzapine\n",
      "   olanzapine                               URI=http://purl.obolibrary.org/obo/CHEBI_7735\n",
      "   olanzapine                               URI=http://purl.obolibrary.org/obo/CHEBI_7735 score=1.00\n",
      "\n",
      "Query: olanzapine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_7735\n",
      "\n",
      "Query: olanzapine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_7735\n",
      "\n",
      "Query: olanzapine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_7735\n",
      "\n",
      "Query: placebo tablets\n",
      "Query: placebo\n",
      "Query: placebo\n",
      "Query: placebo\n",
      "Query: olanzapine\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_7735\n",
      "\n",
      "Query: antibiotic\n",
      "   piperidine antibiotic                    URI=http://purl.obolibrary.org/obo/CHEBI_49318 score=0.79\n",
      "\n",
      "Query: antibiotic\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_49318\n",
      "\n",
      "Query: antibiotic\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_49318\n",
      "\n",
      "Query: antibiotic\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_49318\n",
      "\n",
      "Query: antidepressants\n",
      "Query: escitalopram\n",
      "   escitalopram                             URI=http://purl.obolibrary.org/obo/CHEBI_36791\n",
      "   escitalopram(1+)                         URI=http://purl.obolibrary.org/obo/CHEBI_77406 score=1.00\n",
      "\n",
      "Query: escitalopram\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_77406\n",
      "\n",
      "Query: ct-26\n",
      "   26,26,26-trifluoro-25-hydroxyvitamin D3 / 26,26,26-trifluoro-25-hydroxycholecalciferol URI=http://purl.obolibrary.org/obo/CHEBI_168482 score=0.83\n",
      "\n",
      "Query: oxaliplatin\n",
      "   oxaliplatin                              URI=http://purl.obolibrary.org/obo/CHEBI_31941\n",
      "   oxaliplatin                              URI=http://purl.obolibrary.org/obo/CHEBI_31941 score=1.00\n",
      "\n",
      "Query: rhepo\n",
      "Query: cuprizone\n",
      "Query: cpz\n",
      "Query: cpz\n",
      "Query: antibiotics\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_22476\n",
      "\n",
      "Query: csf1r inhibitor\n",
      "   inhibitor                                URI=http://purl.obolibrary.org/obo/CHEBI_35222 score=1.00\n",
      "\n",
      "Query: antidepressants\n",
      "Query: antidepressant medications\n",
      "   antidepressant                           URI=http://purl.obolibrary.org/obo/CHEBI_35469 score=1.00\n",
      "\n",
      "Query: antidepressants\n",
      "Query: antidepressants\n",
      "Query: antidepressants\n",
      "Query: antidepressants\n",
      "Query: antidepressant pmx\n",
      "Query: antidepressants\n",
      "Query: antidepressants\n",
      "Query: psychotropic medication\n",
      "   psychotropic drug                        URI=http://purl.obolibrary.org/obo/CHEBI_35471 score=0.82\n",
      "\n",
      "Query: drugs\n",
      "Query: antibiotics\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/CHEBI_22476\n",
      "\n",
      "Query: antidepressant\n",
      "   antidepressant                           URI=http://purl.obolibrary.org/obo/CHEBI_35469\n",
      "   antidepressant                           URI=http://purl.obolibrary.org/obo/CHEBI_35469 score=1.00\n",
      "\n",
      "Query: anti-anxiety drugs\n",
      "Query: placebo\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import os\n",
    "import unicodedata\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "DRUG_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "DRUG_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Drug\")\n",
    "DRUG_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Drug\")\n",
    "\n",
    "created[\"Drug\"] = URIRef(f\"{CHEBI_BASE}CHEBI_23888\")\n",
    "g.add((created[\"Drug\"], RDF.type, DRUG_CLASS))\n",
    "g.add((created[\"Drug\"], RDF.type, SKOS.Concept))\n",
    "g.add((created[\"Drug\"], RDFS.label, Literal(\"Drug\", datatype=XSD.string)))\n",
    "g.add((created[\"Drug\"], SKOS.inScheme, DRUG_CONCEPT_SCHEME))\n",
    "\n",
    "\n",
    "def load_chebi_labels(path):\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        next(fh) \n",
    "        for ln in fh:\n",
    "            uri, label = ln.rstrip(\"\\n\").split(\"\\t\", 1)\n",
    "            rows.append((label, uri))\n",
    "    return rows\n",
    "\n",
    "CHEBI_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\chebi_labels.txt\"\n",
    "chebi_rows = load_chebi_labels(CHEBI_LABELS_FILE)\n",
    "\n",
    "exact_ix = defaultdict(list)\n",
    "for lbl, uri in chebi_rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "labels_only = [lbl for lbl, _ in chebi_rows]\n",
    "vec = TfidfVectorizer(stop_words=\"english\")\n",
    "mat = vec.fit_transform(labels_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v  = vec.transform([term])\n",
    "    sc = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = chebi_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "greek_map = {\n",
    "    '': 'alpha',  '': 'alpha',\n",
    "    '': 'beta',   '': 'beta',\n",
    "    '': 'gamma',  '': 'gamma',\n",
    "    '': 'delta',  '': 'delta',\n",
    "    '': 'epsilon','': 'epsilon',\n",
    "    '': 'zeta',   '': 'zeta',\n",
    "    '': 'eta',    '': 'eta',\n",
    "    '': 'theta',  '': 'theta',\n",
    "    '': 'iota',   '': 'iota',\n",
    "    '': 'kappa',  '': 'kappa',\n",
    "    '': 'lambda', '': 'lambda',\n",
    "    '': 'mu',     '': 'mu',\n",
    "    '': 'nu',     '': 'nu',\n",
    "    '': 'xi',     '': 'xi',\n",
    "    '': 'omicron','': 'omicron',\n",
    "    '': 'pi',     '': 'pi',\n",
    "    '': 'rho',    '': 'rho',\n",
    "    '': 'sigma',  '': 'sigma',\n",
    "    '': 'tau',    '': 'tau',\n",
    "    '': 'upsilon','': 'upsilon',\n",
    "    '': 'phi',    '': 'phi',\n",
    "    '': 'chi',    '': 'chi',\n",
    "    '': 'psi',    '': 'psi',\n",
    "    '': 'omega',  '': 'omega',\n",
    "}\n",
    "\n",
    "def preprocess(term):\n",
    "    \"\"\"\n",
    "    1) replace Greek letters with names\n",
    "    2) turn underscores into spaces\n",
    "    3) strip whitespace\n",
    "    4) lowercase\n",
    "    \"\"\"\n",
    "    # 1) map Greek  name\n",
    "    for greek_char, name in greek_map.items():\n",
    "        if greek_char in term:\n",
    "            term = term.replace(greek_char, name)\n",
    "    # 2) replace underscores\n",
    "    term = term.replace('_', ' ')\n",
    "    # 3) strip\n",
    "    term = term.strip()\n",
    "    # 4) lowercase\n",
    "    return term.lower()\n",
    "    \n",
    "for paper_id, paper_data in data.items():\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"drug\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            ex = exact_ix.get(term, [])\n",
    "            if ex:\n",
    "                for lbl, uri in ex:\n",
    "                    print(f\"   {lbl:40s} URI={uri}\")\n",
    "                    entity_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, DRUG_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, DRUG_CONCEPT_SCHEME))\n",
    "                    g.add((entity_uri, SKOS.broaderTransitive, created[\"Drug\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_drug_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            if cos:\n",
    "                for lbl, uri, score in cos[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, DRUG_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, DRUG_CONCEPT_SCHEME))\n",
    "                    g.add((name_uri, SKOS.broaderTransitive, created[\"Drug\"]))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_drug_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_drug_{term_raw}\", datatype=XSD.string)))\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                g.add((created[\"Drug\"], GUTPROP.containedIn, mention_uri))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"drug\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_drug_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4fe97fb5-d87b-48d0-a7fe-e64494c98219",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: chicken gut microbiome\n",
      "   human gut microbiome                     URI=http://purl.obolibrary.org/obo/OHMI_0000020 score=0.91\n",
      "\n",
      "Query: gut microbiome\n",
      "   human gut microbiome                     URI=http://purl.obolibrary.org/obo/OHMI_0000020 score=0.91\n",
      "\n",
      "Query: poultry gut microbiome\n",
      "   human gut microbiome                     URI=http://purl.obolibrary.org/obo/OHMI_0000020 score=0.91\n",
      "\n",
      "Query: chicken gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: oral and gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.82\n",
      "\n",
      "Query: gut and oral microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.82\n",
      "\n",
      "Query: intestinal microbiome\n",
      "   intestinal cancer                        URI=http://purl.obolibrary.org/obo/DOID_10155 score=0.81\n",
      "\n",
      "Query: oral and gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: oral and gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: host microbiome\n",
      "   host-microbiome interaction              URI=http://purl.obolibrary.org/obo/OHMI_0000001 score=0.85\n",
      "\n",
      "Query: microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: host microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.77\n",
      "\n",
      "Query: host microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: individual s gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: individual s specific gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbial ecosystem\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: maternal prenatal gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: maternal gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: prenatal faecal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: faecal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: maternal faecal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: maternal faecal microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: maternal prenatal gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: prenatal gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: human microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.96\n",
      "\n",
      "Query: probiotic-exposed microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: gut microbe\n",
      "   microbe                                  URI=http://purl.obolibrary.org/obo/OHMI_0000460 score=0.96\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: fecal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: human intestinal microbiota\n",
      "Query: gut metagenomes\n",
      "Query: fecal metagenomes\n",
      "Query: gut microbiomes\n",
      "Query: fecal metagenomes\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: human microbiome\n",
      "   human microbiome                         URI=http://purl.obolibrary.org/obo/OHMI_0000002\n",
      "   human microbiome                         URI=http://purl.obolibrary.org/obo/OHMI_0000002 score=1.00\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: human gut microbiome\n",
      "   human gut microbiome                     URI=http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "   human gut microbiome                     URI=http://purl.obolibrary.org/obo/OHMI_0000020 score=1.00\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: fecal microbiota\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: gut microbial community\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: canine intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: depression gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: maternal vaginal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: microbiota of neonatal gut and nose\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: infant microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: maternal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: microbiota of infant gut and nose\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: maternal vaginal microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: infant faecal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: maternal rectal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: infant microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: infant s nasal and gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: infant stool microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: maternal rectal microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: infant stool microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: microbiota of infants\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: infant gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: infant microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: infant microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: maternal microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: infant gut and nasal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: colonic microbiome\n",
      "   microbiome                               URI=http://purl.obolibrary.org/obo/OHMI_0000003 score=1.00\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: altered gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: dysbiotic gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: non-commensal gut microbiome\n",
      "   human gut microbiome                     URI=http://purl.obolibrary.org/obo/OHMI_0000020 score=0.91\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gastrointestinal microbiome\n",
      "   microbiome in human gastrointestinal system URI=http://purl.obolibrary.org/obo/OHMI_0000016 score=0.97\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal flora\n",
      "   intestinal cancer                        URI=http://purl.obolibrary.org/obo/DOID_10155 score=0.93\n",
      "\n",
      "Query: intestinal flora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: intestinal flora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: maternal gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbes\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gastrointestinal tract microbiome\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: human microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: human microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: fecal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: fecal microbiota\n",
      "Query: fecal microbiota\n",
      "Query: fecal microbiota\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal flora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: irif\n",
      "Query: irif\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: intestinal microbiota\n",
      "Query: human gut microbiomes\n",
      "Query: fecal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: duodenal microbiome\n",
      "   microbiome                               URI=http://purl.obolibrary.org/obo/OHMI_0000003 score=1.00\n",
      "\n",
      "Query: microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: human gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microorganisms\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut-microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.95\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: microbiome populations\n",
      "   microbiome                               URI=http://purl.obolibrary.org/obo/OHMI_0000003 score=1.00\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut flora\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal flora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gms\n",
      "Query: intestinal flora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: faecal microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: intestinal flora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: faecal microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: oral and gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut and oral microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbial\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: pharmacomicrobiomics\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: host microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: cecal microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal microbe\n",
      "Query: intestinal microflora\n",
      "   intestinal cancer                        URI=http://purl.obolibrary.org/obo/DOID_10155 score=0.93\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: human gut microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=0.91\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbe\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000460\n",
      "\n",
      "Query: colonic microbiota\n",
      "   microbiota                               URI=http://purl.obolibrary.org/obo/OHMI_0000463 score=1.00\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: inflammatory microbes and genes\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: positive mood-related gut microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: small bowel microbiome\n",
      "Query: small intestinal microbiome\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal barrier\n",
      "   intestinal cancer                        URI=http://purl.obolibrary.org/obo/DOID_10155 score=0.93\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: colon microbiota\n",
      "Query: colon microbiota\n",
      "Query: colon microbiota\n",
      "Query: colon microbiota\n",
      "Query: fecal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: fecal microbiota\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiome\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000020\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: intestinal microflora\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/DOID_10155\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n",
      "Query: gut microbiota\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OHMI_0000463\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import os\n",
    "import unicodedata\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "MICROBIOME_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "MICROBIOME_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/Microbiome\")\n",
    "MICROBIOME_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/Microbiome\")\n",
    "\n",
    "def load_ohmi_labels(path):\n",
    "    pattern = re.compile(r'^\\s*(.*?)\\s*\\[([A-Za-z0-9_]+)\\]')\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        for ln in fh:\n",
    "            m = pattern.match(ln)\n",
    "            if not m:\n",
    "                continue\n",
    "            label, term_id = m.group(1), m.group(2)\n",
    "            uri = MICROBIOME_BASE + term_id\n",
    "            rows.append((label, uri))\n",
    "    return rows\n",
    "\n",
    "OHMI_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\OHMI_full_taxonomy.txt\"\n",
    "ohmi_rows = load_ohmi_labels(OHMI_LABELS_FILE)\n",
    "\n",
    "exact_ix = defaultdict(list)\n",
    "for lbl, uri in ohmi_rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "labels_only = [lbl for lbl, _ in ohmi_rows]\n",
    "vec = TfidfVectorizer(stop_words=\"english\")\n",
    "mat = vec.fit_transform(labels_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v  = vec.transform([term])\n",
    "    sc = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = ohmi_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "def preprocess(term):\n",
    "    term = term.replace('_', ' ')\n",
    "    term = term.strip()\n",
    "    return term.lower()\n",
    "    \n",
    "for paper_id, paper_data in data.items():\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"microbiome\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            if re.search(r'\\bbacter(?:ia|ium)\\b', text_span.lower()):\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_microbiome_{term_raw}\", datatype=XSD.string)))\n",
    "                #g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                continue\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            ex = exact_ix.get(term, [])\n",
    "            if ex:\n",
    "                for lbl, uri in ex:\n",
    "                    print(f\"   {lbl:40s} URI={uri}\")\n",
    "                    entity_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, MICROBIOME_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, MICROBIOME_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_microbiome_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            if cos:\n",
    "                for lbl, uri, score in cos[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, MICROBIOME_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, MICROBIOME_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_microbiome_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_microbiome_{term_raw}\", datatype=XSD.string)))\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"microbiome\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_microbiome_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ab0bafb8-28c5-423b-afd6-089c5ebae08e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: random-effects meta-analyses\n",
      "   Meta-Analysis                            URI=http://purl.obolibrary.org/obo/NCIT_C17886 score=0.84\n",
      "\n",
      "Query: wald test\n",
      "   Wald test                                URI=http://purl.obolibrary.org/obo/STATO_0000559\n",
      "   Wald test                                URI=http://purl.obolibrary.org/obo/STATO_0000559 score=1.00\n",
      "\n",
      "Query: receiver operating characteristic curve analysis\n",
      "no matches\n",
      "Query: random forests\n",
      "   random variable                          URI=http://purl.obolibrary.org/obo/STATO_0000221 score=0.78\n",
      "\n",
      "Query: rf\n",
      "no matches\n",
      "Query: boruta algorithm\n",
      "   algorithm                                URI=http://purl.obolibrary.org/obo/IAO_0000064 score=1.00\n",
      "\n",
      "Query: rf\n",
      "no matches\n",
      "Query: chao1 index\n",
      "no matches\n",
      "Query: shannon\n",
      "no matches\n",
      "Query: inverse simpson\n",
      "no matches\n",
      "Query: bray curtis dissimilarities\n",
      "no matches\n",
      "Query: two sided mann whitney test\n",
      "   Mann-Whitney U-test                      URI=http://purl.obolibrary.org/obo/STATO_0000076 score=0.85\n",
      "\n",
      "Query: cochrane risk of bias\n",
      "   risk                                     URI=http://purl.obolibrary.org/obo/STATO_0000620 score=1.00\n",
      "\n",
      "Query: random forests  rf\n",
      "   random variable                          URI=http://purl.obolibrary.org/obo/STATO_0000221 score=0.78\n",
      "\n",
      "Query: boruta algorithm\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/IAO_0000064\n",
      "\n",
      "Query: mr analysis\n",
      "no matches\n",
      "Query: reverse mr analysis\n",
      "no matches\n",
      "Query: linear discriminant analysis effect size\n",
      "no matches\n",
      "Query: lefse\n",
      "no matches\n",
      "Query: lefse analysis\n",
      "no matches\n",
      "Query: random forest classification\n",
      "no matches\n",
      "Query: false discovery rate  fdr\n",
      "   false discovery rate correction method   URI=http://purl.obolibrary.org/obo/OBI_0200163 score=0.78\n",
      "\n",
      "Query: mendelian randomization  mr  analyses\n",
      "   group randomization                      URI=http://purl.obolibrary.org/obo/OBI_0302900 score=0.77\n",
      "\n",
      "Query: two-sample mr\n",
      "   one sample t-test                        URI=http://purl.obolibrary.org/obo/STATO_0000302 score=0.82\n",
      "\n",
      "Query: inverse variance weighted  ivw   mr-egger  and weighted median  wm  methods\n",
      "no matches\n",
      "Query: multivariable mr\n",
      "no matches\n",
      "Query: mvmr\n",
      "no matches\n",
      "Query: cochran s q test\n",
      "   F-Test                                   URI=http://purl.obolibrary.org/obo/NCIT_C53234 score=1.00\n",
      "\n",
      "Query: mr egger intercept test\n",
      "   intercept                                URI=http://purl.obolibrary.org/obo/STATO_0000657 score=0.86\n",
      "\n",
      "Query: mr steiger test\n",
      "   Z-test                                   URI=http://purl.obolibrary.org/obo/STATO_0000052 score=1.00\n",
      "\n",
      "Query: sensitivity analyses\n",
      "no matches\n",
      "Query: reverse mr analysis\n",
      "no matches\n",
      "Query: independent sample t test\n",
      "   One Sample t-Test                        URI=http://purl.obolibrary.org/obo/NCIT_C53240 score=1.00\n",
      "\n",
      "Query: mann-whitney u test\n",
      "   Mann-Whitney U-test                      URI=http://purl.obolibrary.org/obo/STATO_0000076 score=1.00\n",
      "\n",
      "Query: pearson or spearman correlation test\n",
      "   Pearson Correlation Test                 URI=http://purl.obolibrary.org/obo/NCIT_C53244 score=0.83\n",
      "\n",
      "Query: lefse analysis\n",
      "no matches\n",
      "Query: linear discriminant analysis\n",
      "no matches\n",
      "Query: lda\n",
      "no matches\n",
      "Query: statistical modeling with logistic regression\n",
      "no matches\n",
      "Query: correlation analysis\n",
      "no matches\n",
      "Query: pearson correlation analysis\n",
      "   Pearson Correlation Test                 URI=http://purl.obolibrary.org/obo/NCIT_C53244 score=0.79\n",
      "\n",
      "Query: pearson s correlation analysis\n",
      "   Pearson Correlation Test                 URI=http://purl.obolibrary.org/obo/NCIT_C53244 score=0.79\n",
      "\n",
      "Query: pearson s correlation analysis\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/NCIT_C53244\n",
      "\n",
      "Query: random forest models\n",
      "   random forest procedure                  URI=http://purl.obolibrary.org/obo/STATO_0000549 score=0.82\n",
      "\n",
      "Query: multiple regression analysis\n",
      "no matches\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import os\n",
    "import unicodedata\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "STATISTICALTECHNIQUE_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "STATISTICALTECHNIQUE_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/StatisticalTechnique\")\n",
    "STATISTICALTECHNIQUE_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/StatisticalTechnique\")\n",
    "\n",
    "def load_statistical_labels(path):\n",
    "    pattern = re.compile(r'^\\s*(.*?)\\s*[\\(\\[]([A-Za-z0-9_]+)[\\)\\]]')\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        for ln in fh:\n",
    "            m = pattern.match(ln)\n",
    "            if not m:\n",
    "                continue\n",
    "            label, term_id = m.group(1), m.group(2)\n",
    "            uri = STATISTICALTECHNIQUE_BASE + term_id\n",
    "            rows.append((label, uri))\n",
    "    return rows\n",
    "\n",
    "STATISTICAL_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\stato_full_taxonomy.txt\"\n",
    "STATO_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\subtree_C19044.txt\"\n",
    "stat_rows = load_statistical_labels(STATISTICAL_LABELS_FILE)\n",
    "stat1_rows = load_statistical_labels(STATO_LABELS_FILE)\n",
    "\n",
    "exact_ix = defaultdict(list)\n",
    "for lbl, uri in stat_rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "exact_ix1 = defaultdict(list)\n",
    "for lbl, uri in stat1_rows:\n",
    "    exact_ix1[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "labels_only = [lbl for lbl, _ in stat_rows]\n",
    "vec = TfidfVectorizer(stop_words=\"english\")\n",
    "mat = vec.fit_transform(labels_only)\n",
    "\n",
    "labels1_only = [lbl for lbl, _ in stat1_rows]\n",
    "vec1 = TfidfVectorizer(stop_words=\"english\")\n",
    "mat1 = vec1.fit_transform(labels1_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v  = vec.transform([term])\n",
    "    sc = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = stat_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "def top_cosine1(term, k=5, thr=0.75):\n",
    "    v   = vec1.transform([term])\n",
    "    sc  = cosine_similarity(v, mat1).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = stat1_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "def preprocess(term):\n",
    "    term = term.replace('_', ' ')\n",
    "    term = term.strip()\n",
    "    return term.lower()\n",
    "    \n",
    "for paper_id, paper_data in data.items():\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"statistical technique\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            ex = exact_ix.get(term, [])\n",
    "            if ex:\n",
    "                for lbl, uri in ex:\n",
    "                    print(f\"   {lbl:40s} URI={uri}\")\n",
    "                    entity_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, STATISTICALTECHNIQUE_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, STATISTICALTECHNIQUE_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_stattechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            cos1 = top_cosine1(term)\n",
    "            if cos:\n",
    "                for lbl, uri, score in cos[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, STATISTICALTECHNIQUE_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, STATISTICALTECHNIQUE_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_stattechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "                    \n",
    "            elif cos1:\n",
    "                for lbl, uri, score in cos1[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, STATISTICALTECHNIQUE_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, STATISTICALTECHNIQUE_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_stattechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_stattechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                print(\"no matches\")\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"statistical technique\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_stattechnique_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "99ea7da7-5254-43e6-bb44-3827c9b32e87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: dqi-i\n",
      "no matches\n",
      "Query: 16s rdna surveys\n",
      "no matches\n",
      "Query: 16s rdna surveys\n",
      "no matches\n",
      "Query: metatranscriptomics analyses\n",
      "no matches\n",
      "Query: metabolomics\n",
      "no matches\n",
      "Query: community-based metabolic modeling\n",
      "no matches\n",
      "Query: 16s rrna amplicon sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: childhood behaviour checklist\n",
      "no matches\n",
      "Query: lc-ms\n",
      "no matches\n",
      "Query: gc\n",
      "no matches\n",
      "Query: 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: 16s rrna sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: picrust analysis\n",
      "no matches\n",
      "Query: dual hit toxin model\n",
      "no matches\n",
      "Query: forced swimming test\n",
      "   Z-test                                   URI=http://purl.obolibrary.org/obo/STATO_0000052 score=1.00\n",
      "\n",
      "Query: open field test\n",
      "   Z-test                                   URI=http://purl.obolibrary.org/obo/STATO_0000052 score=1.00\n",
      "\n",
      "Query: sucrose preference test\n",
      "   Z-test                                   URI=http://purl.obolibrary.org/obo/STATO_0000052 score=1.00\n",
      "\n",
      "Query: body weight\n",
      "   contrast weight                          URI=http://purl.obolibrary.org/obo/STATO_0000322 score=0.75\n",
      "\n",
      "Query: enzyme linked immunosorbent assay\n",
      "   enzyme                                   URI=http://purl.obolibrary.org/obo/OBI_0000427 score=0.80\n",
      "\n",
      "Query: immunofluorescence\n",
      "no matches\n",
      "Query: 16 s rna sequencing\n",
      "no matches\n",
      "Query: western blotting\n",
      "no matches\n",
      "Query: multiomics study\n",
      "   study design                             URI=http://purl.obolibrary.org/obo/OBI_0500000 score=0.78\n",
      "\n",
      "Query: multiomics study\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0500000\n",
      "\n",
      "Query: fecal metagenome-wide sequencing\n",
      "no matches\n",
      "Query: serum metabolomics profiling\n",
      "no matches\n",
      "Query: metagenomic analysis\n",
      "no matches\n",
      "Query: 16s rrna sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: qpcr\n",
      "no matches\n",
      "Query: hippocampal proteomic analysis\n",
      "no matches\n",
      "Query: repeated social defeat stress\n",
      "no matches\n",
      "Query: repeated social defeat stress\n",
      "no matches\n",
      "Query: rsds\n",
      "no matches\n",
      "Query: metagenomics analysis of the microbiome\n",
      "no matches\n",
      "Query: gut permeability assay\n",
      "   assay                                    URI=http://purl.obolibrary.org/obo/OBI_0000070 score=1.00\n",
      "\n",
      "Query: ussing chamber\n",
      "no matches\n",
      "Query: immunoblotting\n",
      "no matches\n",
      "Query: immunohistochemical analyses\n",
      "no matches\n",
      "Query: mechanistic analysis\n",
      "no matches\n",
      "Query: shotgun metagenomics\n",
      "no matches\n",
      "Query: shotgun metagenomics sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: sms\n",
      "no matches\n",
      "Query: tryptophan concentration measurements\n",
      "   concentration                            URI=http://purl.obolibrary.org/obo/STATO_0000704 score=1.00\n",
      "\n",
      "Query: fpg\n",
      "no matches\n",
      "Query: fpi\n",
      "no matches\n",
      "Query: homa-ir\n",
      "   IR, Rat Strain                           URI=http://purl.obolibrary.org/obo/NCIT_C14405 score=0.81\n",
      "\n",
      "Query: ancom-bc2\n",
      "no matches\n",
      "Query: chronic unpredictable mild stress  cums  model\n",
      "no matches\n",
      "Query: genomic sequencing\n",
      "no matches\n",
      "Query: whole genome shotgun sequencing\n",
      "no matches\n",
      "Query: 16s rrna gene-sequenced gut microbiota data\n",
      "no matches\n",
      "Query: 16s rrna sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: multiplex immunoassay\n",
      "no matches\n",
      "Query: updrs part iii motor scores\n",
      "no matches\n",
      "Query: pentobarbital-induced sleep test\n",
      "   Z-test                                   URI=http://purl.obolibrary.org/obo/STATO_0000052 score=1.00\n",
      "\n",
      "Query: comparative genomics analysis\n",
      "no matches\n",
      "Query: phenol-sulfate assay\n",
      "   assay                                    URI=http://purl.obolibrary.org/obo/OBI_0000070 score=1.00\n",
      "\n",
      "Query: pentobarbital-induced sleep test\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/STATO_0000052\n",
      "\n",
      "Query: metagenomic sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: transcriptome sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: rna-seq\n",
      "   RNA-seq assay                            URI=http://purl.obolibrary.org/obo/OBI_0001271 score=0.88\n",
      "\n",
      "Query: rna-seq profiling\n",
      "no matches\n",
      "Query: double-blind  randomized  placebo-controlled trials\n",
      "no matches\n",
      "Query: multi-omics analyses\n",
      "no matches\n",
      "Query: metabolomics\n",
      "no matches\n",
      "Query: 16s rrna\n",
      "no matches\n",
      "Query: untargeted metabolomics\n",
      "no matches\n",
      "Query: shotgun metagenomics\n",
      "no matches\n",
      "Query: shotgun metagenomic sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: mitbamp analyses\n",
      "no matches\n",
      "Query: aav-shrna\n",
      "no matches\n",
      "Query: 16s ribosomal rna analysis\n",
      "no matches\n",
      "Query: untargeted metabolomic analysis\n",
      "no matches\n",
      "Query: burst testing\n",
      "no matches\n",
      "Query: 16s rrna sequencing technique\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: shotgun metagenomic sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: analysis of the fecal samples\n",
      "no matches\n",
      "Query: wet laboratory methods\n",
      "no matches\n",
      "Query: 13c-d-xylose breath tests\n",
      "no matches\n",
      "Query: 13c-d-xylose breath testing\n",
      "no matches\n",
      "Query: single-cell rna-seq analysis\n",
      "no matches\n",
      "Query: microbiota-targeted therapies\n",
      "no matches\n",
      "Query: 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: shotgun metagenomic sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: strain-level meta-analysis\n",
      "   meta analysis                            URI=http://purl.obolibrary.org/obo/STATO_0000155 score=0.78\n",
      "\n",
      "Query: 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: phylogenetic investigation of communities by reconstruction of unobserved states  picrust  analysis\n",
      "   investigation                            URI=http://purl.obolibrary.org/obo/OBI_0000066 score=0.83\n",
      "\n",
      "Query: 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: nuclear mr spectroscopy\n",
      "no matches\n",
      "Query: illumina sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: kegg analysis\n",
      "no matches\n",
      "Query: kegg analysis\n",
      "no matches\n",
      "Query: omics techniques\n",
      "no matches\n",
      "Query: lc-ms ms\n",
      "no matches\n",
      "Query: gastrointestinal and microbiome profiling\n",
      "no matches\n",
      "Query: v4 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: microbiome profiling\n",
      "no matches\n",
      "Query: voluntary wheel running\n",
      "   voluntary sampling                       URI=http://purl.obolibrary.org/obo/STATO_0000513 score=0.83\n",
      "\n",
      "Query: splash\n",
      "no matches\n",
      "Query: elevated plus maze\n",
      "no matches\n",
      "Query: tail suspension tests\n",
      "no matches\n",
      "Query: multiple integrated omics\n",
      "no matches\n",
      "Query: 16s ribosomal rna  16s rrna  gene sequence-based approach\n",
      "no matches\n",
      "Query: untargeted liquid chromatography-mass spectrometry-based metabolic profiling approach\n",
      "no matches\n",
      "Query: 16s rrna gene amplicon sequencing\n",
      "no matches\n",
      "Query: metabolomics\n",
      "no matches\n",
      "Query: 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: head-up tilt test\n",
      "   Z-test                                   URI=http://purl.obolibrary.org/obo/STATO_0000052 score=1.00\n",
      "\n",
      "Query: hutt\n",
      "no matches\n",
      "Query: 16s rrna sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: quantitative fluorescence in situ hybridization\n",
      "no matches\n",
      "Query: high-sensitivity c-reactive protein\n",
      "no matches\n",
      "Query: oft\n",
      "no matches\n",
      "Query: spt\n",
      "no matches\n",
      "Query: tst\n",
      "no matches\n",
      "Query: igtt\n",
      "no matches\n",
      "Query: liver lipid kit\n",
      "no matches\n",
      "Query: legend plex kits\n",
      "no matches\n",
      "Query: elisa\n",
      "no matches\n",
      "Query: rt-pcr\n",
      "   RT-PCR standard curve                    URI=http://purl.obolibrary.org/obo/STATO_0000219 score=0.78\n",
      "\n",
      "Query: western blot\n",
      "no matches\n",
      "Query: 16s rrna gene sequencing\n",
      "no matches\n",
      "Query: multi-omics approach\n",
      "no matches\n",
      "Query: multi-omics approach\n",
      "no matches\n",
      "Query: functional enrichment analysis\n",
      "no matches\n",
      "Query: blast shotgun metagenomic sequencing\n",
      "   sequencing assay                         URI=http://purl.obolibrary.org/obo/OBI_0600047 score=0.79\n",
      "\n",
      "Query: 16s rrna sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: shotgun metagenomics\n",
      "no matches\n",
      "Query: 16s rrna gene and shotgun metagenomic analysis\n",
      "no matches\n",
      "Query: duodenal aspirates\n",
      "no matches\n",
      "Query: 16s rrna sequencing\n",
      "   Reusing existing URI: http://purl.obolibrary.org/obo/OBI_0600047\n",
      "\n",
      "Query: bioinformatics analysis\n",
      "no matches\n",
      "Query: stool biobank\n",
      "no matches\n",
      "Query: single prolonged stress  sps  model\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: behavioral tests\n",
      "no matches\n",
      "Query: anxiety index\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: sps\n",
      "no matches\n",
      "Query: sps model\n",
      "no matches\n",
      "Query: multi-omics analysis\n",
      "no matches\n",
      "Query: taxon-function analysis\n",
      "   function                                 URI=http://purl.obolibrary.org/obo/BFO_0000034 score=0.75\n",
      "\n",
      "Query: randomized  double-blinded  two-arm feasibility study\n",
      "no matches\n",
      "Query: quantitative sybr green real-time pcr\n",
      "   real time quantitative pcr plot          URI=http://purl.obolibrary.org/obo/STATO_0000168 score=0.94\n",
      "\n",
      "Query: 16s rna method\n",
      "no matches\n",
      "Query: elisa kit\n",
      "no matches\n",
      "The RDF graph has been saved in C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\rdf\\gutbrain_entities.ttl\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import os\n",
    "import unicodedata\n",
    "from pathlib import Path\n",
    "from rdflib import Graph, Namespace, URIRef, Literal\n",
    "from rdflib.namespace import RDF, RDFS, XSD, SKOS, OWL\n",
    "from pprint import pprint\n",
    "\n",
    "BIOMEDICALTECHNIQUE_BASE = \"http://purl.obolibrary.org/obo/\"\n",
    "BIOMEDICALTECHNIQUE_CLASS = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/BiomedicalTechnique\")\n",
    "BIOMEDICALTECHNIQUE_CONCEPT_SCHEME = URIRef(\"https://hereditary.dei.unipd.it/ontology/gutbrain/schema/conceptScheme/BiomedicalTechnique\")\n",
    "\n",
    "def load_biomedical_labels(path):\n",
    "    pattern = re.compile(r'^\\s*(.*?)\\s*[\\(\\[]([A-Za-z0-9_]+)[\\)\\]]')\n",
    "    rows = []\n",
    "    with open(path, encoding=\"utf-8\") as fh:\n",
    "        for ln in fh:\n",
    "            m = pattern.match(ln)\n",
    "            if not m:\n",
    "                continue\n",
    "            label, term_id = m.group(1), m.group(2)\n",
    "            uri = BIOMEDICALTECHNIQUE_BASE + term_id\n",
    "            rows.append((label, uri))\n",
    "    return rows\n",
    "\n",
    "BIOMEDICAL_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\subtree_C14182.txt\"\n",
    "STATO_LABELS_FILE = r\"C:\\Users\\samue\\OneDrive\\Desktop\\ThesisPiron\\stato_full_taxonomy.txt\"\n",
    "biom_rows = load_biomedical_labels(BIOMEDICAL_LABELS_FILE)\n",
    "biom1_rows = load_biomedical_labels(STATO_LABELS_FILE)\n",
    "\n",
    "exact_ix = defaultdict(list)\n",
    "for lbl, uri in biom_rows:\n",
    "    exact_ix[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "exact_ix1 = defaultdict(list)\n",
    "for lbl, uri in biom1_rows:\n",
    "    exact_ix1[lbl.lower()].append((lbl, uri))\n",
    "\n",
    "labels_only = [lbl for lbl, _ in biom_rows]\n",
    "vec = TfidfVectorizer(stop_words=\"english\")\n",
    "mat = vec.fit_transform(labels_only)\n",
    "\n",
    "labels1_only = [lbl for lbl, _ in biom1_rows]\n",
    "vec1 = TfidfVectorizer(stop_words=\"english\")\n",
    "mat1 = vec1.fit_transform(labels1_only)\n",
    "\n",
    "def top_cosine(term, k=5, thr=0.75):\n",
    "    v  = vec.transform([term])\n",
    "    sc = cosine_similarity(v, mat).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = biom_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "def top_cosine1(term, k=5, thr=0.75):\n",
    "    v   = vec1.transform([term])\n",
    "    sc  = cosine_similarity(v, mat1).ravel()\n",
    "    idx = np.argsort(sc)[::-1]\n",
    "    out = []\n",
    "    for i in idx[:k]:\n",
    "        if sc[i] < thr:\n",
    "            break\n",
    "        lbl, uri = biom1_rows[i]\n",
    "        out.append((lbl, uri, sc[i]))\n",
    "    return out\n",
    "\n",
    "def preprocess(term):\n",
    "    term = term.replace('_', ' ')\n",
    "    term = term.strip()\n",
    "    return term.lower()\n",
    "    \n",
    "for paper_id, paper_data in data.items():\n",
    "    \n",
    "    entities = paper_data.get(\"entities\", [])\n",
    "    \n",
    "    for i, entity in enumerate(entities):\n",
    "        raw_label = entity.get(\"label\", \"\").strip()\n",
    "        \n",
    "        text_span = entity.get(\"text_span\", \"\").strip()\n",
    "        \n",
    "        if raw_label == \"biomedical technique\":\n",
    "            text_span = entity.get(\"text_span\", \"\").strip()\n",
    "            cleaned_text = create_uri_fragment(text_span)\n",
    "            cleaned_text_span = normalize_to_ascii(cleaned_text)\n",
    "            term_raw = cleaned_text_span.lower()\n",
    "            term = preprocess(term_raw)\n",
    "            \n",
    "            print(f\"Query: {term}\")\n",
    "            \n",
    "            if term_raw in created:\n",
    "                entity_uri = created[term_raw]\n",
    "                print(f\"   Reusing existing URI: {entity_uri}\\n\")\n",
    "                continue\n",
    "\n",
    "            ex = exact_ix.get(term, [])\n",
    "            if ex:\n",
    "                for lbl, uri in ex:\n",
    "                    print(f\"   {lbl:40s} URI={uri}\")\n",
    "                    entity_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = entity_uri\n",
    "                    g.add((entity_uri, RDF.type, BIOMEDICALTECHNIQUE_CLASS))\n",
    "                    g.add((entity_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((entity_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, SKOS.inScheme, BIOMEDICALTECHNIQUE_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_biomtechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((entity_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    continue\n",
    "                    print(); \n",
    "                    \n",
    "            cos = top_cosine(term)\n",
    "            cos1 = top_cosine1(term)\n",
    "            if cos:\n",
    "                for lbl, uri, score in cos[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, BIOMEDICALTECHNIQUE_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, BIOMEDICALTECHNIQUE_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_biomtechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "                    \n",
    "            elif cos1:\n",
    "                for lbl, uri, score in cos1[:1]:\n",
    "                    print(f\"   {lbl:40s} URI={uri:40s} score={score:.2f}\")\n",
    "                    name_uri = URIRef(f\"{uri}\")\n",
    "                    created[term_raw] = name_uri\n",
    "                    g.add((name_uri, RDF.type, BIOMEDICALTECHNIQUE_CLASS))\n",
    "                    g.add((name_uri, RDF.type, SKOS.Concept))\n",
    "                    g.add((name_uri, RDFS.label, Literal(lbl, datatype=XSD.string)))\n",
    "                    g.add((name_uri, SKOS.inScheme, BIOMEDICALTECHNIQUE_CONCEPT_SCHEME))\n",
    "                    mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                    g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                    g.add((mention_uri, RDFS.label, Literal(f\"mention_biomtechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                    g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                    g.add((name_uri, GUTPROP.containedIn, mention_uri))\n",
    "                    tokenized_mentions[term_raw] = mention_uri\n",
    "                    print()\n",
    "            else:\n",
    "                mention_uri = URIRef(GUTBRAIN[term_raw])\n",
    "                g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "                g.add((mention_uri, RDFS.label, Literal(f\"mention_biomtechnique_{term_raw}\", datatype=XSD.string)))\n",
    "                g.add((mention_uri, GUTPROP.hasMentionText, Literal(term_raw, datatype=XSD.string)))\n",
    "                tokenized_mentions[term_raw] = mention_uri\n",
    "                print(\"no matches\")\n",
    "                continue\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "with open(tokenized_file, \"r\", encoding=\"utf-8\") as f_sent:\n",
    "    tokenized_data = json.load(f_sent)\n",
    "\n",
    "for entry in tokenized_data:\n",
    "    pmid         = entry[\"pmid\"]\n",
    "    sent_id      = entry[\"sent_id\"]\n",
    "    sentence_txt = entry[\"sentence\"].strip()\n",
    "    entities     = entry[\"entities\"]\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "\n",
    "    sent_uri = URIRef(GUTBRAIN[f\"sentence_{pmid}_{sent_id}\"])\n",
    "    g.add((sent_uri, RDF.type, SENTENCE))\n",
    "    g.add((sent_uri, GUTPROP.hasSentenceText, Literal(sentence_txt, datatype=XSD.string)))\n",
    "\n",
    "    if sent_id == 0:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"title_{pmid}\"])\n",
    "    else:\n",
    "        parent_uri = URIRef(GUTBRAIN[f\"abstract_{pmid}\"])\n",
    "    g.add((sent_uri,      GUTPROP.partOf,  parent_uri))\n",
    "    g.add((parent_uri,    GUTPROP.composedOf, sent_uri))\n",
    "        \n",
    "    for ent in entities:\n",
    "        if isinstance(ent, dict):\n",
    "            text_span = ent.get(\"text_span\", \"\").strip()\n",
    "            label     = ent.get(\"label\",    \"\").strip().lower()\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        if label != \"biomedical technique\":\n",
    "            continue\n",
    "        canonical = create_uri_fragment(text_span)\n",
    "        cleaned_text_span = normalize_to_ascii(canonical).lower()\n",
    "        \n",
    "        if cleaned_text_span not in tokenized_mentions:\n",
    "            mention_uri = URIRef(GUTBRAIN[cleaned_text_span])\n",
    "            tokenized_mentions[cleaned_text_span] = mention_uri\n",
    "\n",
    "            g.add((mention_uri, RDF.type, MENTION_CLASS))\n",
    "            g.add((mention_uri, RDFS.label, Literal(f\"mention_biomtechnique_{cleaned_text_span}\", datatype=XSD.string)))\n",
    "            g.add((mention_uri, GUTPROP.hasMentionText, Literal(text_span, datatype=XSD.string)))\n",
    "            \n",
    "        else:\n",
    "            mention_uri = tokenized_mentions[cleaned_text_span]\n",
    "            \n",
    "        g.add((mention_uri, GUTPROP.locatedIn, sent_uri))\n",
    "\n",
    "output_file = os.path.join(save_path, \"gutbrain_entities.ttl\")\n",
    "ttl_output = g.serialize(format=\"turtle\")\n",
    "with open(output_file, \"w\", encoding=\"utf-8\") as f_out:\n",
    "    f_out.write(ttl_output)\n",
    "\n",
    "print(f\"The RDF graph has been saved in {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "201e14f3-ae66-4412-b5a1-b4603bafb987",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
